<!DOCTYPE html>












  


<html class="theme-next muse use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge"/>
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222"/>
<script type="text/javascript" src="http://tajs.qq.com/stats?sId=66189719" charset="UTF-8"></script>


  
  
    
  
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.css"/>





  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/Han/3.3.0/han.min.css"/>



  <meta http-equiv="Cache-Control" content="no-transform"/>
  <meta http-equiv="Cache-Control" content="no-siteapp"/>



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg"/>



  <meta name="msvalidate.01" content="74A687E04D4EF9914B7AA20824207409"/>









  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.css"/>




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link rel="stylesheet" href="https://fonts.loli.net/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext"/>
  






<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2"/>

<link rel="stylesheet" href="/css/main.css?v=7.0.1"/>


  <link rel="apple-touch-icon" sizes="180x180" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/small_favicon.png?v=7.0.1">


  <link rel="mask-icon" href="/images/favicon.svg?v=7.0.1" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '7.0.1',
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":true,"dimmer":false},
    back2top: true,
    back2top_sidebar: false,
    fancybox: true,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="hope bioinformatics blog">
<meta property="og:type" content="website">
<meta property="og:title" content="hope">
<meta property="og:url" content="http://tiramisutes.github.io/page/11/index.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="hope bioinformatics blog">
<meta property="og:locale" content="en">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="hope">
<meta name="twitter:description" content="hope bioinformatics blog">



  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml"/>




  <link rel="canonical" href="http://tiramisutes.github.io/page/11/"/>



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>hope – A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</title>
  






  <script>
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>







  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br/>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br/>Archives</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-code">

    
    
    
      
    

    

    <a href="/code/" rel="section"><i class="menu-item-icon fa fa-fw fa-code"></i> <br/>Code</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br/>Categories</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br/>Tags</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-footprint">

    
    
    
      
    

    

    <a href="/footprint/" rel="section"><i class="menu-item-icon fa fa-fw fa-image"></i> <br/>Footprint</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-qrt-pcr">

    
    
    
      
    

    

    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8="><i class="menu-item-icon fa fa-fw fa-cloud"></i> <br/>qRT-PCR</span>

  </li>
        
        
        
          
          <li class="menu-item menu-item-project">

    
    
    
      
    

    

    <a href="/project/" rel="section"><i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br/>Project</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-publish">

    
    
    
      
    

    

    <a href="/publish/" rel="section"><i class="menu-item-icon fa fa-fw fa-book"></i> <br/>Publish</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-sci-figure">

    
    
    
      
    

    

    <a href="/SCIFigure/" rel="section"><i class="menu-item-icon fa fa-fw fa-angellist"></i> <br/>SCI Figure</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br/>About</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-cv">

    
    
    
      
    

    

    <a href="/resume/" rel="section"><i class="menu-item-icon fa fa-fw fa-address-card"></i> <br/>CV</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br/>Search</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/sort-k.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/sort-k.html" class="post-title-link" itemprop="url">Linux常用命令之sort强悍参数“k”</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 14:33:06" itemprop="dateCreated datePublished" datetime="2015-11-29T14:33:06+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/sort-k.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/sort-k.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/sort-k.html" class="leancloud_visitors" data-flag-title="Linux常用命令之sort强悍参数“k”">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="准备素材">准备素材</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ cat facebook.txt</span><br><span class="line">google 110 5000</span><br><span class="line">baidu 100 5000</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br></pre></td></tr></table></figure>
<p>第一个域是公司名称，第二个域是公司人数，第三个域是员工平均工资。</p>
<h2 id="排序">排序</h2><p><i class="fa fa-gavel"></i>让这个文件按公司的字母顺序排序，也就是按第一个域进行排序：（这个facebook.txt文件有三个域）<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -t <span class="string">' '</span> -k 1 facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br></pre></td></tr></table></figure></p>
<p>看到了吧，就直接用-k 1设定就可以了。（其实此处并不严格，稍后你就会知道）<br><strong>若数据存在表头，而我们一般不需要对表头排序，但是又希望排序后表头依然存在，该如何操作呢？</strong><br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## if you have two header lines and want to keep both of them:</span></span><br><span class="line">(sed -n <span class="string">'1,2p'</span> your_file; cat your_file | sed <span class="string">'1,2d'</span> | sort) &gt; sort_header.txt</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-gavel"></i>让facebook.txt按照公司人数排序<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -t <span class="string">' '</span> -k 2 facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">baidu 100 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p>但是，此处出现了问题，那就是baidu和sohu的公司人数相同，都是100人，这个时候怎么办呢？按照默认规矩，是从第一个域开始进行升序排序，因此baidu排在了sohu前面。<br><i class="fa fa-gavel"></i>让facebook.txt按照公司人数排序 ，人数相同的按照员工平均工资升序排序：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -t <span class="string">' '</span> -k 2 -k 3 facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p>加了一个-k2 -k3就解决了问题,sort支持这种设定，就是说设定域排序的优先级，先以第2个域进行排序，如果相同，再以第3个域进行排序。（如果你愿意，可以一直这么写下去，设定很多个排序优先级）<br><i class="fa fa-gavel"></i>让facebook.txt按照员工工资降序排序，如果员工人数相同的，则按照公司人数升序排序：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -t <span class="string">' '</span> -k 3r -k 2 facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">guge 50 3000</span><br></pre></td></tr></table></figure></p>
<p>此处有使用了一些小技巧，你仔细看看，在-k 3后面偷偷加上了一个小写字母r,r和-r选项的作用是一样的，就是表示逆序。因为sort默认是按照升序排序的，所以此处需要加上r表示第三个域（员工平均工资）是按照降序排序。此处你还可以加上n，就表示对这个域进行排序时，要按照数值大小进行排序，举个例子吧：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -t <span class="string">' '</span> -k 3nr -k 2n facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">guge 50 3000</span><br></pre></td></tr></table></figure></p>
<p>看，我们去掉了最前面的-n选项，而是将它加入到了每一个-k选项中了。</p>
<p><i class="fa fa-gavel"></i> -k选项的具体语法格式</p>
<p>要继续往下深入的话，就不得不来点理论知识。你需要了解-k选项的语法格式，如下：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">[ FStart [ .CStart ] ] [ Modifier ] [ , [ FEnd [ .CEnd ] ][ Modifier ] ]</span><br></pre></td></tr></table></figure></p>
<p>这个语法格式可以被其中的逗号（”，”）分为两大部分，Start部分和End部分。</p>
<p>先给你灌输一个思想，那就是”如果不设定End部分，那么就认为End被设定为行尾”。这个概念很重要的，但往往你不会重视它。</p>
<p>Start部分也由三部分组成，其中的Modifier部分就是我们之前说过的类似n和r的选项部分。我们重点说说Start部分的FStart和C.Start。</p>
<p>C.Start也是可以省略的，省略的话就表示从本域的开头部分开始。之前例子中的-k 2和-k 3就是省略了C.Start的例子喽。</p>
<p>FStart.CStart，其中FStart就是表示使用的域，而CStart则表示在FStart域中从第几个字符开始算”排序首字符”。</p>
<p>同理，在End部分中，你可以设定FEnd.CEnd，如果你省略.CEnd，则表示结尾到”域尾”，即本域的最后一个字符。或者，如果你将CEnd设定为0(零)，也是表示结尾到”域尾”。</p>
<p><i class="fa fa-gavel"></i>从公司英文名称的第二个字母开始进行排序：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -t <span class="string">' '</span> -k 1.2 facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">google 110 5000</span><br><span class="line">guge 50 3000</span><br></pre></td></tr></table></figure></p>
<p>使用了-k 1.2，这就表示对第一个域的第二个字符开始到本域的最后一个字符为止的字符串进行排序。你会发现baidu因为第二个字母是a而名列榜首。sohu和google第二个字符都是o，但sohu的h在google的o前面，所以两者分别排在第二和第三。guge只能屈居第四了。</p>
<p><i class="fa fa-gavel"></i>只针对公司英文名称的第二个字母进行排序，如果相同的按照员工工资进行降序排序：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -t <span class="string">' '</span> -k 1.2,1.2 -k 3,3nr facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">guge 50 3000</span><br></pre></td></tr></table></figure></p>
<p>由于只对第二个字母进行排序，所以我们使用了-k 1.2,1.2的表示方式，表示我们”只”对第二个字母进行排序。（如果你问”我使用-k 1.2怎么不行？”，当然不行，因为你省略了End部分，这就意味着你将对从第二个字母起到本域最后一个字符为止的字符串进行排序）。对于员工工资进行排序，我们也使用了-k 3,3，这是最准确的表述，表示我们”只”对本域进行排序，因为如果你省略了后面的3，就变成了我们”对第3个域开始到最后一个域位置的内容进行排序”了。</p>
<p><i class="fa fa-gavel"></i>在modifier部分还可以用到哪些选项？</p>
<p>可以用到b、d、f、i、n 或 r。</p>
<p>其中n和r你肯定已经很熟悉了。</p>
<p>b表示忽略本域的签到空白符号。</p>
<p>d表示对本域按照字典顺序排序（即，只考虑空白和字母）。</p>
<p>f表示对本域忽略大小写进行排序。</p>
<p>i表示忽略”不可打印字符”，只针对可打印字符进行排序。（有些ASCII就是不可打印字符，比如\a是报警，\b是退格，\n是换行，\r是回车等等）</p>
<p><i class="fa fa-gavel"></i>思考关于-k和-u联合使用的例子：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ cat facebook.txt</span><br><span class="line">google 110 5000</span><br><span class="line">baidu 100 5000</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br></pre></td></tr></table></figure></p>
<p>这是最原始的facebook.txt文件。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -k 2 facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">baidu 100 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">google 110 5000</span><br><span class="line">$ sort -n -k 2 -u facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p>当设定以公司员工域进行数值排序，然后加-u后，sohu一行就被删除了！原来-u只识别用-k设定的域，发现相同，就将后续相同的行都删除。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ sort  -k 1 -u facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br><span class="line"></span><br><span class="line">$ sort  -k 1.1,1.1 -u facebook.txt</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br><span class="line">sohu 100 4500</span><br></pre></td></tr></table></figure></p>
<p>这个例子也同理，开头字符是g的guge就没有幸免于难。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -k 2 -k 3 -u facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p>咦！这里设置了两层排序优先级的情况下，使用-u就没有删除任何行。原来-u是会权衡所有-k选项，将都相同的才会删除，只要其中有一级不同都不会轻易删除的:)（不信，你可以自己加一行sina 100 4500试试看）</p>
<p><i class="fa fa-heartbeat"></i>最诡异的排序：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -k 2.2,3.1 facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">baidu 100 5000</span><br><span class="line">sohu 100 4500</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p>以第二个域的第二个字符开始到第三个域的第一个字符结束的部分进行排序。</p>
<p>第一行，会提取0 3，第二行提取00 5，第三行提取00 4，第四行提取10 5。</p>
<p>又因为sort认为0小于00小于000小于0000….</p>
<p>因此0 3肯定是在第一个。10 5肯定是在最后一个。但为什么00 5却在00 4前面呢？（你可以自己做实验思考一下。）</p>
<p>答案揭晓：原来”跨域的设定是个假象”，sort只会比较第二个域的第二个字符到第二个域的最后一个字符的部分，而不会把第三个域的开头字符纳入比较范围。当发现00和00相同时，sort就会自动比较第一个域去了。当然baidu在sohu前面了。用一个范例即可证实：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ sort -n -k 2.2,3.1 -k 1,1r facebook.txt</span><br><span class="line">guge 50 3000</span><br><span class="line">sohu 100 4500</span><br><span class="line">baidu 100 5000</span><br><span class="line">google 110 5000</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-times"></i>有时候在sort命令后会看到+1 -2这些符号，这是什么东东？</p>
<p>关于这种语法，最新的sort是这么进行解释的：</p>
<p>On older systems, <code>sort&#39; supports an obsolete origin-zero syntax</code>+POS1 [-POS2]’ for specifying sort keys.  POSIX 1003.1-2001 (*note Standards conformance::) does not allow this; use `-k’ instead.</p>
<p>原来，这种古老的表示方式已经被淘汰了，以后可以理直气壮的鄙视使用这种表示方法的脚本喽！</p>
<p>（为了防止古老脚本的存在，在这再说一下这种表示方法，加号表示Start部分，减号表示End部分。最最重要的一点是，这种方式方法是从0开始计数的，以前所说的第一个域，在此被表示为第0个域。以前的第2个字符，在此表示为第1个字符.）</p>
<h2 id="其他有用参数">其他有用参数</h2><h3 id="V(大写)：聪明的字母和数字排序">V(大写)：聪明的字母和数字排序</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">cat example2.bed</span><br><span class="line">chr2    15      19</span><br><span class="line">chr22   32      46</span><br><span class="line">chr10   31      47</span><br><span class="line">chr1    34      49</span><br><span class="line">chr11   6       16</span><br><span class="line">chr2    17      22</span><br><span class="line">chr2    27      46</span><br><span class="line">chr10   30      42</span><br><span class="line">$ sort -k1,1 -k2,2n example2.bed</span><br><span class="line">chr1    34      49</span><br><span class="line">chr10   30      42</span><br><span class="line">chr10   31      47</span><br><span class="line">chr11   6       16</span><br><span class="line">chr2    15      19</span><br><span class="line">chr2    17      22</span><br><span class="line">chr2    27      46</span><br><span class="line">chr22   32      46</span><br><span class="line">$ sort -k1,1V -k2,2n example2.bed</span><br><span class="line">chr1    34      49</span><br><span class="line">chr2    15      19</span><br><span class="line">chr2    17      22</span><br><span class="line">chr2    27      46</span><br><span class="line">chr10   30      42</span><br><span class="line">chr10   31      47</span><br><span class="line">chr11   6       16</span><br><span class="line">chr22   32      46</span><br></pre></td></tr></table></figure>
<p><strong>V is only for GNU sort which will sort chromsome in alpha-numeric order。</strong></p>
<h2 id="设置临时目录">设置临时目录</h2><figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">-T, <span class="attribute">--temporary-directory</span>=DIR</span><br><span class="line">             use DIR <span class="keyword">for</span> temporaries, <span class="keyword">not</span> <span class="variable">$TMPDIR</span> <span class="keyword">or</span> /tmp; multiple options specify multiple directories</span><br></pre></td></tr></table></figure>
<p>因为sort命令默认临时文件目录为”/tmp”，在sort执行过程中，如果产生的临时文件过大会导致”/tmp”目录被占满的。</p>
<p></p><p></p><br>Contribution from ：<br>本原创文章属于《Linux大棚》博客，博客地址为<span class="exturl" data-url="aHR0cDovL3JvY2xpbnV4LmNu44CC5paH56ug5L2c6ICF5Li6cm9jcm9ja2V044CC" title="http://roclinux.cn。文章作者为rocrocket。">http://roclinux.cn。文章作者为rocrocket。<i class="fa fa-external-link"></i></span><p></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/sed.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/sed.html" class="post-title-link" itemprop="url">sed：流编辑器(stream editor)简单总结</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 13:29:36" itemprop="dateCreated datePublished" datetime="2015-11-29T13:29:36+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/sed.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/sed.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/sed.html" class="leancloud_visitors" data-flag-title="sed：流编辑器(stream editor)简单总结">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sed.jpg" alt=""></p>
<h2 id="Sed简介">Sed简介</h2><p>sed 是一种在线编辑器，它一次处理一行内容。处理时，把当前处理的行存储在临时缓冲区中，称为”模式空间”（pattern space），接着用sed命令处理缓冲区中的内容，处理完成后，把缓冲区的内容送往屏幕。接着处理下一行，这样不断重复，直到文件末尾。文件内容并没有 改变，除非你使用重定向存储输出。</p>
<h2 id="sed命令形式">sed命令形式</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sed [options] <span class="string">'command'</span> file(s)    </span><br><span class="line">sed [options] -f scriptfile file(s)</span><br></pre></td></tr></table></figure>
<h2 id="sed_command">sed command</h2><p><i class="fa fa-bicycle"></i>command部分可以分为两部分，一部分是确定范围部分，一部分是处理方式部分。</p>
<h3 id="确定范围部分">确定范围部分</h3><p>1 指定行数：例如3,5表示第3、第4和第5行；5,$表示第5行至最后一行；<br>2 用模式匹配进行指定：例如/^[^dD]/表示匹配行首不是以d或D开头的行等；</p>
<h3 id="处理方式部分呢，有很多命令可用，介绍几个最常用的：">处理方式部分呢，有很多命令可用，介绍几个最常用的：</h3><p>d 表示删除行<br>p 打印该行<br>r 读取指定文件的内容<br>w 写入指定文件sed ‘/200[4-6]/w new.txt’ mysed.txt（w new.txt表示将来源于mysed.txt中含有2004、2005、2006的行写入到new.txt文件中）<br>a\ 在特定行”下面”插入特定内容sed ‘/2004/a\China’ mysed.txt<br>i\ 在特定行”上面”插入特定内容sed ‘/2004/i\China’ mysed.txt<br>y  就表示将第一栏的每个字符都替换为相对应的第二栏的字符sed ‘y/eijng/EIJNG/‘ mysed.txt<br>n; 对匹配行的下一行进行处理sed ‘/2004/{n;y/eijng/EIJNG/;}’ mysed.txt  (找含有2004的行，然后将它下面的一行中的eijng替换为大写的EIJNG。这里面的”n;”起到了”移至下一行”的作用。n背后的含义其实是将下一行放到模式空间中去)</p>
<h2 id="sed选项">sed选项</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">-e <span class="built_in">command</span>, --expression=<span class="built_in">command</span>  允许多台编辑，sed -n -e <span class="string">'1,2p'</span> -e <span class="string">'4p'</span> mysed.txt </span><br><span class="line">-h, --<span class="built_in">help</span>                        打印帮助，并显示bug列表的地址。  </span><br><span class="line">-n, --quiet, --silent             取消默认输出（默认为全输出）,-n之后只输出后面处理过的行。  </span><br><span class="line">-f, --filer=script-file           引导sed脚本文件名。  </span><br><span class="line">-V, --version                     打印版本和版权信息。</span><br><span class="line">-r, --regexp-extended             使用扩展的正则表达式，如果不用r参数就要在正则表达式里使用很多的\来进行强制转换，如果使用r了就可以直接写正则表达式，而不用写那么多\了</span><br></pre></td></tr></table></figure>
<h2 id="sed元字符集">sed元字符集</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">^  锚定行的开始 如：/^sed/匹配所有以sed开头的行。  </span><br><span class="line">$  锚定行的结束 如：/sed$/匹配所有以sed结尾的行。 </span><br><span class="line">.  匹配一个非换行符的字符 如：/s.d/匹配s后接一个任意字符，然后是d。</span><br><span class="line">*  匹配零或多个字符 如：/*sed/匹配所有模板是一个或多个空格后紧跟sed的行。</span><br><span class="line">[] 匹配一个指定范围内的字符，如/[Ss]ed/匹配sed和Sed。  </span><br><span class="line">[^]匹配一个不在指定范围内的字符，如：/[^A-RT-Z]ed/匹配不包含A-R和T-Z的一个字母开头，紧跟ed的行。  </span><br><span class="line">\(..\) 保存匹配的字符，如s/\(love\)able/\1rs，loveable被替换成lovers。  </span><br><span class="line">&amp;  保存搜索字符用来替换其他字符，如s/love/**&amp;**/，love变成**love**。   </span><br><span class="line">\&lt; 锚定单词的开始，如:/\&lt;love/匹配包含以love开头的单词的行。   </span><br><span class="line">\&gt; 锚定单词的结束，如/love\&gt;/匹配包含以love结尾的单词的行。   </span><br><span class="line">x\&#123;m\&#125;  重复字符x，m次，如：/0\&#123;5\&#125;/匹配包含5个o的行。   </span><br><span class="line">x\&#123;m,\&#125; 重复字符x,至少m次，如：/o\&#123;5,\&#125;/匹配至少有5个o的行。   </span><br><span class="line">x\&#123;m,n\&#125; 重复字符x，至少m次，不多于n次，如：/o\&#123;5,10\&#125;/匹配5--10个o的行。</span><br></pre></td></tr></table></figure>
<h2 id="sed实例">sed实例</h2><p>例一 显示test文件的第20到30行：sed -n ‘20,30p’ test<br>例二 将所有以d或D开头的行的所有小写x变为大写X：sed ‘/^[dD]/s/x/X/g’ test<br>例三 删除每行最后的两个字符：sed ‘s/..$//‘ test<br>例四 删除每一行的前两个字符：sed ‘s/..//‘ test<br>例五<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$cat</span> mysed.txt</span><br><span class="line">Beijing Beijing Beijing Beijing</span><br><span class="line"><span class="variable">$sed</span> <span class="string">'s/\(Beijing\)\(.*\)\(Beijing\)/\12008\2\32008/'</span> mysed.txt</span><br><span class="line">Beijing2008 Beijing Beijing Beijing2008</span><br><span class="line"><span class="comment">##这个命令稍显复杂，其中用到了一个技巧，就是预存储，即被\(和\)括起来的匹配内容会被按顺序存储起来，存储到\1、\2…里面。这样你就可以使用\加数字来调用这些内容了。这个例子就是使用了这个技巧，分别存储了三个内容，分别为匹配Beijing、匹配.*和匹配Beijing。</span></span><br></pre></td></tr></table></figure></p>
<p>任意字符:  sed -n ‘/.<em>ing/‘p temp.txt     注意是.</em>ing,而不是<em>ing<br>‘s/^[][]</em>//g’      删除行首空格<br>‘s/^$/d’           删除空行<br>‘s/COL/(…/)//g’  删除紧跟COL的后三个字母</p>
<p>more<i class="fa fa-database"></i><span class="exturl" data-url="aHR0cDovL3d3dy5ncnltb2lyZS5jb20vVW5peC9TZWQuaHRtbA==" title="http://www.grymoire.com/Unix/Sed.html">http://www.grymoire.com/Unix/Sed.html<i class="fa fa-external-link"></i></span><br>Contribution from ：<br><span class="exturl" data-url="aHR0cDovL3JvY2xpbnV4LmNuLz9wPTEzNjM=" title="http://roclinux.cn/?p=1363">http://roclinux.cn/?p=1363<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL3d3dy5pdGV5ZS5jb20vdG9waWMvNTg3Njcz" title="http://www.iteye.com/topic/587673">http://www.iteye.com/topic/587673<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL3d3dy5jbmJsb2dzLmNvbS9lbWFubGVlL2FyY2hpdmUvMjAxMy8wOS8wNy8zMzA3NjQyLmh0bWw=" title="http://www.cnblogs.com/emanlee/archive/2013/09/07/3307642.html">http://www.cnblogs.com/emanlee/archive/2013/09/07/3307642.html<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/27/python-install.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/27/python-install.html" class="post-title-link" itemprop="url">python模块安装--无root权限（easy_install和pip）</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-27 13:17:35" itemprop="dateCreated datePublished" datetime="2015-11-27T13:17:35+08:00">2015-11-27</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Python/" itemprop="url" rel="index"><span itemprop="name">Python</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/27/python-install.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/27/python-install.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/27/python-install.html" class="leancloud_visitors" data-flag-title="python模块安装--无root权限（easy_install和pip）">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>easy_install是由PEAK(Python Enterprise Application Kit)开发的setuptools包里带的一个命令，所以使用easy_install实际上是在调用setuptools来完成安装模块的工作。 Perl 用户比较熟悉 CPAN，而 Ruby 用户则比较熟悉 Gems；引导 setuptools 的ez_setup工具和随之而生的扩展后的easy_install 与”Cheeseshop”（Python Package Index，也称为 “PyPI”）一起工作来实现相同的功能。它可以很方便的让您自动下载，编译，安装和管理Python包。【百度百科】<br>easy_install和pip都是用来下载安装Python一个公共资源库PyPI的相关资源包的，pip类似RedHat里面的yum，安装Python包非常方便。本节详细介绍pip的安装、以及使用方法。</p>
<h2 id="首先安装setuptools">首先安装setuptools</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget <span class="string">"https://bitbucket.org/pypa/setuptools/get/default.tar.gz#egg=setuptools-dev"</span> --no-check-certificate</span><br><span class="line">tar -xzvf default.tar.gz</span><br><span class="line"><span class="built_in">cd</span> pypa-setuptools-eb92fc5071bf //依据你的解压目录名而定</span><br><span class="line">python setup.py install</span><br></pre></td></tr></table></figure>
<h2 id="安装easy_install">安装easy_install</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget https://pypi.python.org/pypi/ez_setup</span><br></pre></td></tr></table></figure>
<p>解压,安装.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python ez_setup.py</span><br></pre></td></tr></table></figure></p>
<h3 id="easy_install安装包">easy_install安装包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">easy_install 【要安装的模块】</span><br></pre></td></tr></table></figure>
<h2 id="pip下载安装">pip下载安装</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget <span class="string">"https://pypi.python.org/packages/source/p/pip/pip-1.5.4.tar.gz#md5=834b2904f92d46aaa333267fb1c922bb"</span> --no-check-certificate</span><br></pre></td></tr></table></figure>
<h3 id="pip安装">pip安装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">tar -xzvf pip-1.5.4.tar.gz</span><br><span class="line"><span class="built_in">cd</span> pip-1.5.4</span><br><span class="line">python setup.py install</span><br></pre></td></tr></table></figure>
<h2 id="pip使用详解">pip使用详解</h2><h3 id="pip安装包">pip安装包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pip install SomePackage</span><br><span class="line">[...]</span><br><span class="line">Successfully installed SomePackage</span><br></pre></td></tr></table></figure>
<h3 id="pip查看已安装的包">pip查看已安装的包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pip show --files SomePackage</span><br><span class="line">Name: SomePackage</span><br><span class="line">Version: 1.0</span><br><span class="line">Location: /my/env/lib/pythonx.x/site-packages</span><br><span class="line">Files:</span><br><span class="line">../somepackage/__init__.py</span><br><span class="line">[...]</span><br></pre></td></tr></table></figure>
<h3 id="pip检查哪些包需要更新">pip检查哪些包需要更新</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pip list --outdated</span><br><span class="line">SomePackage (Current: 1.0 Latest: 2.0)</span><br></pre></td></tr></table></figure>
<h3 id="pip升级包">pip升级包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pip install --upgrade SomePackage</span><br><span class="line">[...]</span><br><span class="line">Found existing installation: SomePackage 1.0</span><br><span class="line">Uninstalling SomePackage:</span><br><span class="line">Successfully uninstalled SomePackage</span><br><span class="line">Running setup.py install <span class="keyword">for</span> SomePackage</span><br><span class="line">Successfully installed SomePackage</span><br></pre></td></tr></table></figure>
<h3 id="pip卸载包">pip卸载包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">pip uninstall SomePackage</span><br><span class="line">Uninstalling SomePackage:</span><br><span class="line">/my/env/lib/pythonx.x/site-packages/somepackage</span><br><span class="line">Proceed (y/n)? y</span><br><span class="line">Successfully uninstalled SomePackage</span><br></pre></td></tr></table></figure>
<h2 id="常见错误">常见错误</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ImportError No module named setuptools</span><br></pre></td></tr></table></figure>
<p>解决办法：安装setuptools</p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/25/gene-isoforms.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/25/gene-isoforms.html" class="post-title-link" itemprop="url">Genes and Isoforms</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-25 19:44:00" itemprop="dateCreated datePublished" datetime="2015-11-25T19:44:00+08:00">2015-11-25</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/25/gene-isoforms.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/25/gene-isoforms.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/25/gene-isoforms.html" class="leancloud_visitors" data-flag-title="Genes and Isoforms">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Isoforms">Isoforms</h2><h3 id="Definition(s)">Definition(s)</h3><p>Different forms of a protein that may be produced from different genes, or from the same gene by alternative splicing.<br>Definition from:&nbsp;MeSH via <span class="exturl" data-url="aHR0cHM6Ly93d3cubmxtLm5paC5nb3YvcmVzZWFyY2gvdW1scy8=" title="https://www.nlm.nih.gov/research/umls/">Unified Medical Language <nobr>System </nobr><i class="fa fa-external-link"></i></span> at the National Library of Medicine<br>The protein products of different versions of messenger RNA created from the<br>same gene by employing different promoters, which causes transcription to<br>skip certain exons. Since the promoters are tissue-specific, different tissues<br>express different protein products of the same gene.<br>Definition from:&nbsp;<span class="exturl" data-url="aHR0cDovL3d3dy5uY2JpLm5sbS5uaWguZ292L2Jvb2tzL05CSzUxOTEv" title="http://www.ncbi.nlm.nih.gov/books/NBK5191/"><nobr>GeneReviews</nobr><i class="fa fa-external-link"></i></span>from the University of Washington and the National Center for Biotechnology Information</p>
<h2 id="Related_discussion_in_the_Handbook">Related discussion in the Handbook</h2><p><li> <a href="/handbook/traits/fingerprints">Are fingerprints determined by genetics?</a></li></p>
<h2 id="Gene">Gene</h2><h3 id="Definition(s)-1">Definition(s)</h3><p>The functional and physical unit of heredity passed from parent to offspring. Genes are pieces of DNA, and most genes contain the information for making a specific protein.<br>Definition from:&nbsp;Physician Data Query via <span class="exturl" data-url="aHR0cHM6Ly93d3cubmxtLm5paC5nb3YvcmVzZWFyY2gvdW1scy8=" title="https://www.nlm.nih.gov/research/umls/">Unified Medical Language <nobr>System</nobr><i class="fa fa-external-link"></i></span> at the National Library of Medicine<br>The basic unit of heredity, consisting of a segment of DNA arranged in a linear<br>manner along a chromosome, which codes for a specific protein or segment<br>of protein leading to a particular characteristic or function.<br>Definition from:&nbsp;<span class="exturl" data-url="aHR0cDovL3d3dy5uY2JpLm5sbS5uaWguZ292L2Jvb2tzL05CSzUxOTEv" title="http://www.ncbi.nlm.nih.gov/books/NBK5191/"><nobr>GeneReviews</nobr><i class="fa fa-external-link"></i></span>from the University of Washington and the National Center for Biotechnology.<br>The gene is the basic physical unit of inheritance. Genes are passed from parents to offspring and contain the information needed to specify traits. Genes are arranged, one after another, on structures called chromosomes. A chromosome contains a single, long DNA molecule, only a portion of which corresponds to a single gene. Humans have approximately 23,000 genes arranged on their chromosomes.<br>Definition from:&nbsp;<span class="exturl" data-url="aHR0cDovL3d3dy5nZW5vbWUuZ292L2dsb3NzYXJ5Lw==" title="http://www.genome.gov/glossary/">Talking Glossary of Genetic <nobr>Terms</nobr><i class="fa fa-external-link"></i></span> from the National Human Genome Research Institute<br>The fundamental physical and functional unit of heredity. A gene is an ordered sequence of nucleotides located in a particular position on a particular chromosome that encodes a specific functional product (i.e., a protein or RNA molecule).<br>Definition from:&nbsp;<span class="exturl" data-url="aHR0cDovL3dlYi5vcm5sLmdvdi9zY2kvdGVjaHJlc291cmNlcy9IdW1hbl9HZW5vbWUvZ2xvc3Nhcnkuc2h0bWw=" title="http://web.ornl.gov/sci/techresources/Human_Genome/glossary.shtml">Human Genome Project <nobr>Information</nobr><i class="fa fa-external-link"></i></span>at the U.S. Department of Energy</p>
<h2 id="Related_discussion_in_the_Handbook-1">Related discussion in the Handbook</h2><p><li> <a href="/handbook/basics/gene">What is a gene?</a>           </li></p>
<p><li> <a href="/handbook/howgeneswork/genelocation">How do geneticists indicate the location of a gene?</a>  </li></p>
<p><li> <a href="/handbook/therapy">Gene Therapy</a>           </li></p>
<p><li> <a href="/handbook/howgeneswork/genefamilies">What are gene families?</a>           </li></p>
<p>See also <a href="/Resources/medicalterminology">Understanding Medical Terminology</a>.</p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/14/read-resequence.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/14/read-resequence.html" class="post-title-link" itemprop="url">重测序文献精读</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-14 20:57:42" itemprop="dateCreated datePublished" datetime="2015-11-14T20:57:42+08:00">2015-11-14</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/14/read-resequence.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/14/read-resequence.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/14/read-resequence.html" class="leancloud_visitors" data-flag-title="重测序文献精读">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>了解一项新的工作或研究内容，文献精读是首选，而Nature系类文章技术含金量高，文章内容条理清晰，精读一定会收获颇丰。<br>全基因组重测序是对已知基因组序列的物种进行不同个体的基因组测序，并在此基础上对个体或群体进行差异性分析。全基因组重测序的个体，通过序列比对，可以找到大量的单核苷酸多态性位点（SNP），插入缺失位点（InDel，Insertion/Deletion）、结构变异位点（SV，Structure Variation）位点，在全基因组水平上扫描并检测与重要性状相关的基因序列差异和结构变异，实现遗传进化分析及重要性状候选基因预测。<br>基于其研究的重要性，<a href="http://tiramisutes.github.io/2015/07/28/google/" target="_blank">Google Scholar</a>关键词搜索resequence，检索到《Resequencing of 31 wild and cultivated soybean genomes identifies patterns of genetic diversity and selection》，《Resequencing 302 wild and cultivated accessions identifies genes related to domestication and improvement in soybean》，现选择第一篇精读。</p>
<h2 id="关键点总结">关键点总结</h2><h3 id="测序相关">测序相关</h3><ol>
<li>Approximately ×5 depth and &gt;90% coverage.</li>
<li>Previous reports have shown that the SNP calling accuracy from resequencing data is ~95–99% (Wang, J. et al. The diploid genome sequence of an Asian individual. Nature 456,60–65 (2008)；Xia, Q. et al. Complete resequencing of 40 genomes reveals domestication events and genes in silkworm (Bombyx). Science 326, 433–436 (2009)).</li>
<li>D-value (Tajima’s D) distribution was significantly higher that indicating a significant loss of rare SNPs, which may be due to reduced recombination within the LD blocks.</li>
<li>Divergence index (FST) value allowed us to identify genomic regions of large FST value, which signified areas having a high degree of diversification.Subregions that have very high FST values may provide an indication of the functional genes or alleles involved.</li>
<li>A genome-wide sequencing comparison to reveal haplotype sharing could provide a unique tool to identify introgression events in the history of these cultivars.</li>
<li>Previous studies have indicated that whole genome duplication (WGD) events can cause gene loss and rapid functional diversification. <h3 id="大豆研究相关">大豆研究相关</h3></li>
<li>They have exceptionally high linkage disequilibrium (LD) and a high ratio of average nonsynonymous versus synonymous nucleotide differences (Nonsyn/Syn). </li>
<li>There was a recent history of introgression from wild soybean.</li>
<li>Human selection probably had a strong impact on the genetic diversity in the cultivated soybeans.</li>
<li>Genome-wide analyses showed the opposite: we found that the low-frequency alleles were less abundant among the wild as compared to the cultivated accessions.</li>
<li>In comparison with other crops, SNP analysis showed that the cultivated soybean exhibited a lower diversity (cultivated soybean: 1.89 × 10−3; rice: 2.29 × 10−3; corn: 6.6 × 10−3).</li>
<li>The average distance over which LD decays to half of its maximum value in soybean was substantially longer than that of all plants analyzed<br>to date.</li>
<li>SNP analyses in the LD blocks showed that there was a lower SNP ratio in long LD blocks as compared to the whole genome in both wild and cultivated.</li>
<li>Allelic diversity in wild soybeans was higher than in cultivated soybeans across the entire genome.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence4.png" alt=""></li>
<li>Only ~3% of the total SNPs identified were present in coding regions. The remaining ~97% SNPs were in noncoding regions.</li>
<li>The presence of a higher Nonsyn/Syn value at the whole-genome level and more large-effect mutations suggested that the soybean genome had accumulated a higher ratio of deleterious mutations.</li>
<li>High LD(long LD blocks) would result in the lack of effective recombination; consequently, deleterious mutations could not be eliminated and would accumulate.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence5.png" alt=""></li>
<li>Selection signals during domestication and improvement.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence6.png" alt=""><h2 id="分析大致流程">分析大致流程</h2><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence.png" alt=""></li>
<li>进化相关工作，包括phylogenetic tree(<a href="http://tiramisutes.github.io/2015/08/05/bio-online/" target="_blank">iTO</a>)，principle component analysis(PCA)，population structure(Bayesian clustering analysis).</li>
<li>Whole-genome SNP analysis (using the parameter θπ) and the distribution of genome-wide diversity.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence3.png" alt=""></li>
<li>High linkage disequilibrium and genomewide patterns of nucleotide diversity(Selection and introgression).</li>
<li>Genome duplication（copy number variations (CNVs)） and Gene content variation.<h2 id="涉及软件">涉及软件</h2></li>
<li>STRUCTURE，Bayesian clustering program，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3ByaXRjaGFyZGxhYi5zdGFuZm9yZC5lZHUvc3RydWN0dXJlLmh0bWw=" title="http://pritchardlab.stanford.edu/structure.html">http://pritchardlab.stanford.edu/structure.html<i class="fa fa-external-link"></i></span><br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sequence2.png" alt=""></li>
<li>Haploview，LD analysis，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cHM6Ly93d3cuYnJvYWRpbnN0aXR1dGUub3JnL3NjaWVudGlmaWMtY29tbXVuaXR5L3NjaWVuY2UvcHJvZ3JhbXMvbWVkaWNhbC1hbmQtcG9wdWxhdGlvbi1nZW5ldGljcy9oYXBsb3ZpZXcvaGFwbG92aWV3" title="https://www.broadinstitute.org/scientific-community/science/programs/medical-and-population-genetics/haploview/haploview">https://www.broadinstitute.org/scientific-community/science/programs/medical-and-population-genetics/haploview/haploview<i class="fa fa-external-link"></i></span></li>
<li>AUGUSTUS ，基因注释，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL2Jpb2luZi51bmktZ3JlaWZzd2FsZC5kZS9hdWd1c3R1cy8=" title="http://bioinf.uni-greifswald.de/augustus/">http://bioinf.uni-greifswald.de/augustus/<i class="fa fa-external-link"></i></span></li>
<li>GeneWise and Genomewise，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5uY2JpLm5sbS5uaWguZ292L3BtYy9hcnRpY2xlcy9QTUM0NzkxMzAv" title="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC479130/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC479130/<i class="fa fa-external-link"></i></span>   ，<span class="exturl" data-url="aHR0cDovL3d3dy5lYmkuYWMudWsvVG9vbHMvcHNhL2dlbmV3aXNlLw==" title="http://www.ebi.ac.uk/Tools/psa/genewise/">http://www.ebi.ac.uk/Tools/psa/genewise/<i class="fa fa-external-link"></i></span><br>GeneWise, which predicts gene structure using similar protein sequences, and Genomewise, which provides a gene structure final parse across cDNA- and EST-defined spliced structure. Both algorithms are heavily used by the Ensembl annotation system. The GeneWise algorithm was developed from a principled combination of hidden Markov models (HMMs). Both algorithms are highly accurate and can provide both accurate and complete gene structures when used with the correct evidence.</li>
<li>SOAP and SOAPsnp,Short Oligonucleotide Alignment Program(45-bp or 76-bp), <i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3NvYXAuZ2Vub21pY3Mub3JnLmNuLw==" title="http://soap.genomics.org.cn/">http://soap.genomics.org.cn/<i class="fa fa-external-link"></i></span>;</li>
<li>BWA，Paired-end sequencing reads mapping，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3NvdXJjZWZvcmdlLm5ldC9wcm9qZWN0cy9iaW8tYndhL2ZpbGVzLw==" title="http://sourceforge.net/projects/bio-bwa/files/">http://sourceforge.net/projects/bio-bwa/files/<i class="fa fa-external-link"></i></span></li>
<li>SAMtools，SNP detection<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5odHNsaWIub3JnL++8jGh0dHA6Ly9iaW9iaXRzLm9yZy9zYW10b29sc19wcmltZXIuaHRtbA==" title="http://www.htslib.org/，http://biobits.org/samtools_primer.html">http://www.htslib.org/，http://biobits.org/samtools_primer.html<i class="fa fa-external-link"></i></span></li>
<li>Picard package，Duplicated reads filtered，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3BpY2FyZC5zb3VyY2Vmb3JnZS5uZXQv" title="http://picard.sourceforge.net/">http://picard.sourceforge.net/<i class="fa fa-external-link"></i></span></li>
<li>BEDtools，coverage of sequence alignments<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL2JlZHRvb2xzLnJlYWR0aGVkb2NzLm9yZy9lbi9sYXRlc3Qv" title="http://bedtools.readthedocs.org/en/latest/">http://bedtools.readthedocs.org/en/latest/<i class="fa fa-external-link"></i></span></li>
<li>Genome Analysis Toolkit (GATK)，SNP/Indel calling，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cHM6Ly93d3cuYnJvYWRpbnN0aXR1dGUub3JnL2dhdGsvZ2F0a2RvY3Mvb3JnX2Jyb2FkaW5zdGl0dXRlX2dhdGtfdG9vbHNfd2Fsa2Vyc19nZW5vdHlwZXJfVW5pZmllZEdlbm90eXBlci5waHA=" title="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_genotyper_UnifiedGenotyper.php">https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_genotyper_UnifiedGenotyper.php<i class="fa fa-external-link"></i></span></li>
<li>ANNOVAR，SNP annotation<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL2Fubm92YXIub3BlbmJpb2luZm9ybWF0aWNzLm9yZy9lbi9sYXRlc3Qv" title="http://annovar.openbioinformatics.org/en/latest/">http://annovar.openbioinformatics.org/en/latest/<i class="fa fa-external-link"></i></span></li>
<li>EIGENSOFT，Principal component analysis (PCA) of whole-genome SNPs<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL2dlbmV0aWNzLm1lZC5oYXJ2YXJkLmVkdS9yZWljaC9SZWljaF9MYWIvU29mdHdhcmUuaHRtbA==" title="http://genetics.med.harvard.edu/reich/Reich_Lab/Software.html">http://genetics.med.harvard.edu/reich/Reich_Lab/Software.html<i class="fa fa-external-link"></i></span></li>
<li>PLINK，Whole genome association analysis toolset，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3BuZ3UubWdoLmhhcnZhcmQuZWR1L35wdXJjZWxsL3BsaW5rLw==" title="http://pngu.mgh.harvard.edu/~purcell/plink/">http://pngu.mgh.harvard.edu/~purcell/plink/<i class="fa fa-external-link"></i></span></li>
<li>GAPIT，Genome Association and Prediction<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5tYWl6ZWdlbmV0aWNzLm5ldC8jIWdhcGl0L2Nta3Y=" title="http://www.maizegenetics.net/#!gapit/cmkv">http://www.maizegenetics.net/#!gapit/cmkv<i class="fa fa-external-link"></i></span></li>
<li>manhattan plot，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cHM6Ly9wb2RzLmlwbGFudGNvbGxhYm9yYXRpdmUub3JnL3dpa2kvZGlzcGxheS9lb3QvTWFrZSttYW5oYXR0YW4rcGxvdCt3aXRoK2dncGxvdDIrc2NyaXB077yMaHR0cDovL2Jsb2cuaG93LXRvLWNvZGUuaW5mby9yL01hbmhhdHRhbi1wbG90Lmh0bWw=" title="https://pods.iplantcollaborative.org/wiki/display/eot/Make+manhattan+plot+with+ggplot2+script，http://blog.how-to-code.info/r/Manhattan-plot.html">https://pods.iplantcollaborative.org/wiki/display/eot/Make+manhattan+plot+with+ggplot2+script，http://blog.how-to-code.info/r/Manhattan-plot.html<i class="fa fa-external-link"></i></span><br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/manhattanplot.png" alt=""><h2 id="内容补充链接">内容补充链接</h2></li>
<li>Tajima’s D，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cHM6Ly9lbi53aWtpcGVkaWEub3JnL3dpa2kvVGFqaW1hJTI3c19E77yb" title="https://en.wikipedia.org/wiki/Tajima%27s_D；">https://en.wikipedia.org/wiki/Tajima%27s_D；<i class="fa fa-external-link"></i></span><i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL2JhaWtlLmJhaWR1LmNvbS9saW5rP3VybD1oa1JQUWNVdEJWTVRWaE1sMnd6S0dMeTVRdERjck13b25VVjdDc3BxeHFkcGhrR3p0clNORlpMaVVZYXpxNm96NnJ4WnlWb3kxWWhIamV4aGk5T3A5Xw==" title="http://baike.baidu.com/link?url=hkRPQcUtBVMTVhMl2wzKGLy5QtDcrMwonUV7CspqxqdphkGztrSNFZLiUYazq6oz6rxZyVoy1YhHjexhi9Op9_">http://baike.baidu.com/link?url=hkRPQcUtBVMTVhMl2wzKGLy5QtDcrMwonUV7CspqxqdphkGztrSNFZLiUYazq6oz6rxZyVoy1YhHjexhi9Op9_<i class="fa fa-external-link"></i></span>.</li>
<li>Penn State University Center for Comparative Genomics and Bioinformatics，<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5ieC5wc3UuZWR1L21pbGxlcl9sYWIv" title="http://www.bx.psu.edu/miller_lab/">http://www.bx.psu.edu/miller_lab/<i class="fa fa-external-link"></i></span></li>
</ol>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/14/P-FDR.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/14/P-FDR.html" class="post-title-link" itemprop="url">What's the probability that a significant p-value indicates a true effect?</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-14 19:32:12" itemprop="dateCreated datePublished" datetime="2015-11-14T19:32:12+08:00">2015-11-14</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/14/P-FDR.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/14/P-FDR.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/14/P-FDR.html" class="leancloud_visitors" data-flag-title="What's the probability that a significant p-value indicates a true effect?">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>If the p-value is &lt; .05, then the probability of falsely rejecting the null hypothesis is  &lt;5%, right? That means, a maximum of 5% of all significant results is a false-positive (that’s what we control with the α rate).</p>
<p>Well, no. As you will see in a minute, the “false discovery rate” (aka. false-positive rate), which indicates the probability that a significant p-value actually is a false-positive, usually is much higher than 5%.</p>
<h2 id="A_common_misconception_about_p-values">A common misconception about p-values</h2><p>Oates (1986) asked the following question to students and senior scientists:<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">You have a p-value of .01. Is the following statement <span class="literal">true</span>, or <span class="literal">false</span>?</span><br><span class="line"></span><br><span class="line">You know, <span class="keyword">if</span> you decide to reject the null hypothesis, the probability that you are making the wrong decision.</span><br></pre></td></tr></table></figure></p>
<p>The answer is “false” (you will learn why it’s false below). But 86% of all professors and lecturers in the sample who were teaching statistics (!) answered this question erroneously with “true”. Gigerenzer, Kraus, and Vitouch replicated this result in 2000 in a German sample (here, the “statistics lecturer” category had 73% wrong). Hence, it is a wide-spread error to confuse the p-value with the false discovery rate.</p>
<h2 id="The_False_Discovery_Rate_(FDR)_and_the_Positive_Predictive_Value_(PPV)">The False Discovery Rate (FDR) and the Positive Predictive Value (PPV)</h2><p>To answer the question “What’s the probability that a significant p-value indicates a true effect?”, we have to look at the positive predictive value (PPV) of a significant p-value. The PPV indicates the proportion of significant p-values which indicate a real effect amongst all significant p-values. Put in other words: Given that a p-value is significant: What is the probability (in a frequentist sense) that it stems from a real effect?</p>
<p>(The false discovery rate simply is 1-PPV: the probability that a significant p-value stems from a population with null effect).</p>
<p>That is, we are interested in a conditional probability Prob(effect is real | p-value is significant).<br>Inspired by Colquhoun (2014) one can visualize this conditional probability in the form of a tree-diagram (see below). Let’s assume, we carry out 1000 experiments for 1000 different research questions. We now have to make a couple of prior assumptions (which you can make differently in the app we provide below). For now, we assume that 30% of all studies have a real effect and the statistical test used has a power of 35% with an α level set to 5%. That is of the 1000 experiments, 300 investigate a real effect, and 700 a null effect. Of the 300 true effects, 0.35<em>300 = 105 are detected, the remaining 195 effects are non-significant false-negatives. On the other branch of 700 null effects, 0.05</em>700 = 35 p-values are significant by chance (false positives) and 665 are non-significant (true negatives).<br>This path is visualized here (completely inspired by Colquhoun, 2014):<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/PPV_tree.jpeg" alt=""><br>Now we can compute the false discovery rate (FDR): 35 of (35+105) = 140 significant p-values actually come from a null effect. That means, 35/140 = 25% of all significant p-values do not indicate a real effect! That is much more than the alleged 5% level.</p>
<p></p><p></p><br><i class="fa fa-link"></i> Contribution from ：<span class="exturl" data-url="aHR0cDovL3d3dy5yLWJsb2dnZXJzLmNvbS93aGF0cy10aGUtcHJvYmFiaWxpdHktdGhhdC1hLXNpZ25pZmljYW50LXAtdmFsdWUtaW5kaWNhdGVzLWEtdHJ1ZS1lZmZlY3Qv" title="http://www.r-bloggers.com/whats-the-probability-that-a-significant-p-value-indicates-a-true-effect/">http://www.r-bloggers.com/whats-the-probability-that-a-significant-p-value-indicates-a-true-effect/<i class="fa fa-external-link"></i></span><p></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/11/pca-f-c.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/11/pca-f-c.html" class="post-title-link" itemprop="url">主成份分析、因子分析和聚类分析的异同点</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-11 21:33:50" itemprop="dateCreated datePublished" datetime="2015-11-11T21:33:50+08:00">2015-11-11</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/11/pca-f-c.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/11/pca-f-c.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/11/pca-f-c.html" class="leancloud_visitors" data-flag-title="主成份分析、因子分析和聚类分析的异同点">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="基本介绍">基本介绍</h2><p>主成分分析就是将多项指标转化为少数几项综合指标,用综合指标来解释多变量的方差- 协方差结构。综合指标即为主成分。所得出的少数几个主成分，要尽可能多地保留原始变量的信息，且彼此不相关。</p>
<p>因子分析是研究如何以最少的信息丢失，将众多原始变量浓缩成少数几个因子变量，以及如何使因子变量具有较强的可解释性的一种多元统计分析方法。</p>
<p>聚类分析是依据实验数据本身所具有的定性或定量的特征来对大量的数据进行分组归类以了解数据集的内在结构，并且对每一个数据集进行描述的过程。其主要依据是聚到同一个数据集中的样本应该彼此相似，而属于不同组的样本应该足够不相似。</p>
<p>三种分析方法既有区别也有联系，本文力图将三者的异同进行比较，并举例说明三者在实际应用中的联系,以期为更好地利用这些高级统计方法为研究所用有所裨益。</p>
<h2 id="基本思想的异同">基本思想的异同</h2><h3 id="共同点">共同点</h3><p>主成分分析法和因子分析法都是用少数的几个变量(因子) 来综合反映原始变量(因子) 的主要信息，变量虽然较原始变量少，但所包含的信息量却占原始信息的85 %以上，所以即使用少数的几个新变量，可信度也很高，也可以有效地解释问题。并且新的变量彼此间互不相关，消除了多重共线性。这两种分析法得出的新变量，并不是原始变量筛选后剩余的变量。在主成分分析中，最终确定的新变量是原始变量的线性组合，如原始变量为x1 ，x2 ，. . . ，x3 ，经过坐标变换，将原有的p个相关变量xi 作线性变换，每个主成分都是由原有p 个变量线性组合得到。在诸多主成分Zi 中，Z1 在方差中占的比重最大，说明它综合原有变量的能力最强，越往后主成分在方差中的比重也小，综合原信息的能力越弱。因子分析是要利用少数几个公共因子去解释较多个要观测变量中存在的复杂关系，它不是对原始变量的重新组合，而是对原始变量进行分解，分解为公共因子与特殊因子两部分。公共因子是由所有变量共同具有的少数几个因子；特殊因子是每个原始变量独自具有的因子。对新产生的主成分变量及因子变量计算其得分，就可以将主成分得分或因子得分代替原始变量进行进一步的分析，因为主成分变量及因子变量比原始变量少了许多，所以起到了降维的作用，为我们处理数据降低了难度。</p>
<p>聚类分析的基本思想是: 采用多变量的统计值，定量地确定相互之间的亲疏关系，考虑对象多因素的联系和主导作用，按它们亲疏差异程度，归入不同的分类中一元，使分类更具客观实际并能反映事物的内在必然联系。也就是说，聚类分析是把研究对象视作多维空间中的许多点，并合理地分成若干类，因此它是一种根据变量域之间的相似性而逐步归群成类的方法，它能客观地反映这些变量或区域之间的内在组合关系[3 ]。聚类分析是通过一个大的对称矩阵来探索相关关系的一种数学分析方法，是多元统计分析方法，分析的结果为群集。对向量聚类后，我们对数据的处理难度也自然降低，所以从某种意义上说，聚类分析也起到了降维的作用。</p>
<h3 id="不同之处">不同之处</h3><p>主成分分析是研究如何通过少数几个主成分来解释多变量的方差一协方差结构的分析方法，也就是求出少数几个主成分(变量) ，使它们尽可能多地保留原始变量的信息，且彼此不相关。它是一种数学变换方法，即把给定的一组变量通过线性变换，转换为一组不相关的变量(两两相关系数为0 ，或样本向量彼此相互垂直的随机变量) ，在这种变换中，保持变量的总方差(方差之和) 不变，同时具有最大方差，称为第一主成分；具有次大方差，称为第二主成分。依次类推。若共有p 个变量，实际应用中一般不是找p 个主成分，而是找出m (m &lt; p) 个主成分就够了，只要这m 个主成分能反映原来所有变量的绝大部分的方差。主成分分析可以作为因子分析的一种方法出现。</p>
<p>因子分析是寻找潜在的起支配作用的因子模型的方法。因子分析是根据相关性大小把变量分组，使得同组内的变量之间相关性较高，但不同的组的变量相关性较低，每组变量代表一个基本结构，这个基本结构称为公共因子。对于所研究的问题就可试图用最少个数的不可测的所谓公共因子的线性函数与特殊因子之和来描述原来观测的每一分量。通过因子分析得来的新变量是对每个原始变量进行内部剖析。因子分析不是对原始变量的重新组合，而是对原始变量进行分解，分解为公共因子和特殊因子两部分。具体地说，就是要找出某个问题中可直接测量的具有一定相关性的诸指标，如何受少数几个在专业中有意义、又不可直接测量到、且相对独立的因子支配的规律，从而可用各指标的测定来间接确定各因子的状态。因子分析只能解释部分变异，主成分分析能解释所有变异。       </p>
<p>聚类分析算法是给定m 维空间R 中的n 个向量，把每个向量归属到k 个聚类中的某一个，使得每一个向量与其聚类中心的距离最小。聚类可以理解为: 类内的相关性尽量大，类间相关性尽量小。聚类问题作为一种无指导的学习问题，目的在于通过把原来的对象集合分成相似的组或簇，来获得某种内在的数据规律。</p>
<p>从三类分析的基本思想可以看出，聚类分析中并没于产生新变量，但是主成分分析和因子分析都产生了新变量。</p>
<h2 id="数据标准化的比较">数据标准化的比较</h2><p>主成分分析中为了消除量纲和数量级，通常需要将原始数据进行标准化，将其转化为均值为0方差为1 的无量纲数据。而因子分析在这方面要求不是太高，因为在因子分析中可以通过主因子法、加权最小二乘法、不加权最小二乘法、重心法等很多解法来求因子变量，并且因子变量是每一个变量的内部影响变量，它的求解与原始变量是否同量纲关系并不太大，当然在采用主成分法求因子变量时，仍需标准化。不过在实际应用的过程中，为了尽量避免量纲或数量级的影响，建议在使用因子分析前还是要进行数据标准化。在构造因子变量时采用的是主成分分析方法，主要将指标值先进行标准化处理得到协方差矩阵，即相关矩阵和对应的特征值与特征向量，然后构造综合评价函数进行评价。</p>
<p>聚类分析中如果参与聚类的变量的量纲不同会导致错误的聚类结果。因此在聚类过程进行之前必须对变量值进行标准化，即消除量纲的影响。不同方法进行标准化，会导致不同的聚类结果要注意变量的分布。如果是正态分布应该采用z 分数法。</p>
<h2 id="应用中的优缺点比较">应用中的优缺点比较</h2><h3 id="主成分分析">主成分分析</h3><p><i class="fa fa-diamond"></i>优点：首先它利用降维技术用少数几个综合变量来代替原始多个变量，这些综合变量集中了原始变量的大部分信息。其次它通过计算综合主成分函数得分，对客观经济现象进行科学评价。再次它在应用上侧重于信息贡献影响力综合评价。<br><i class="fa fa-times"></i>缺点：当主成分的因子负荷的符号有正有负时，综合评价函数意义就不明确。命名清晰性低。</p>
<h3 id="因子分析">因子分析</h3><p><i class="fa fa-diamond"></i>优点：第一它不是对原有变量的取舍，而是根据原始变量的信息进行重新组合，找出影响变量的共同因子，化简数据；第二，它通过旋转使得因子变量更具有可解释性，命名清晰性高。<br><i class="fa fa-times"></i>缺点：在计算因子得分时，采用的是最小二乘法，此法有时可能会失效。</p>
<h3 id="聚类分析">聚类分析</h3><p><i class="fa fa-diamond"></i>优点：聚类分析模型的优点就是直观，结论形式简明。<br><i class="fa fa-times"></i>缺点：在样本量较大时，要获得聚类结论有一定困难。由于相似系数是根据被试的反映来建立反映被试间内在联系的指标，而实践中有时尽管从被试反映所得出的数据中发现他们之间有紧密的关系，但事物之间却无任何内在联系，此时，如果根据距离或相似系数得出聚类分析的结果，显然是不适当的，但是，聚类分析模型本身却无法识别这类错误。<br>Contribution from ：<span class="exturl" data-url="aHR0cDovL2Jsb2cuc2luYS5jb20uY24vcy9ibG9nXzY2ZDM2MmQ3MDEwMWZpdWouaHRtbA==" title="http://blog.sina.com.cn/s/blog_66d362d70101fiuj.html">http://blog.sina.com.cn/s/blog_66d362d70101fiuj.html<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/11/cluster.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/11/cluster.html" class="post-title-link" itemprop="url">R中cluster包进行聚类分析</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-11 21:05:33" itemprop="dateCreated datePublished" datetime="2015-11-11T21:05:33+08:00">2015-11-11</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/11/cluster.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/11/cluster.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/11/cluster.html" class="leancloud_visitors" data-flag-title="R中cluster包进行聚类分析">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Description">Description</h2><p>Methods for Cluster analysis. Much extended the original from Peter Rousseeuw, Anja Struyf and Mia Hubert, based on Kaufman and Rousseeuw (1990) ``Finding Groups in Data’’.<br>聚类分析：按照个体或样品(individuals, objects or subjects)的特征将它们分类，使同一类别内的个体具有尽可能高的同质性(homogeneity)，而类别之间则应具有尽可能高的异质性(heterogeneity)。<br>Cluster analysis or clustering is the task of grouping a set of objects in such a way that objects in the same group (called a cluster) are more similar (in some sense or another) to each other than to those in other groups (clusters). It is a main task of exploratory data mining, and a common technique for statistical data analysis, used in many fields, including machine learning, pattern recognition, image analysis, information retrieval, and bioinformatics.</p>
<h2 id="Typical_cluster_models">Typical cluster models</h2><p><ul></ul></p>
<p><li>Connectivity models: for example <a href="/wiki/Hierarchical_clustering" title="Hierarchical clustering">hierarchical clustering</a> builds models based on distance connectivity.</li></p>
<p><li>Centroid models: for example the <a href="/wiki/K-means_algorithm" title="K-means algorithm" class="mw-redirect">k-means algorithm</a> represents each cluster by a single mean vector.</li></p>
<p><li>Distribution models: clusters are modeled using statistical distributions, such as <a href="/wiki/Multivariate_normal_distribution" title="Multivariate normal distribution">multivariate normal distributions</a> used by the <a href="/wiki/Expectation-maximization_algorithm" title="Expectation-maximization algorithm" class="mw-redirect">Expectation-maximization algorithm</a>.</li></p>
<p><li>Density models: for example <a href="/wiki/DBSCAN" title="DBSCAN">DBSCAN</a> and <a href="/wiki/OPTICS" title="OPTICS" class="mw-redirect">OPTICS</a> defines clusters as connected dense regions in the data space.</li></p>
<p><li>Subspace models: in <a href="/wiki/Biclustering" title="Biclustering">Biclustering</a> (also known as Co-clustering or two-mode-clustering), clusters are modeled with both cluster members and relevant attributes.</li></p>
<p><li>Group models: some algorithms do not provide a refined model for their results and just provide the grouping information.</li></p>
<p><li>Graph-based models: a <a href="/wiki/Clique_(graph_theory)" title="Clique (graph theory)">clique</a>, i.e., a subset of nodes in a <a href="/wiki/Graph_(mathematics)" title="Graph (mathematics)">graph</a> such that every two nodes in the subset are connected by an edge can be considered as a prototypical form of cluster. Relaxations of the complete connectivity requirement (a fraction of the edges can be missing) are known as quasi-cliques, as in the <a href="/wiki/HCS_clustering_algorithm" title="HCS clustering algorithm">HCS clustering algorithm</a>.</li><br></p>
<h2 id="Algorithms">Algorithms</h2><h3 id="Connectivity_based_clustering_(hierarchical_clustering)">Connectivity based clustering (hierarchical clustering)</h3><p>Connectivity based clustering, also known as hierarchical clustering, is based on the core idea of objects being more related to nearby objects than to objects farther away. These algorithms connect “objects” to form “clusters” based on their distance. A cluster can be described largely by the maximum distance needed to connect parts of the cluster. At different distances, different clusters will form, which can be represented using a dendrogram, which explains where the common name “hierarchical clustering” comes from: these algorithms do not provide a single partitioning of the data set, but instead provide an extensive hierarchy of clusters that merge with each other at certain distances. In a dendrogram, the y-axis marks the distance at which the clusters merge, while the objects are placed along the x-axis such that the clusters don’t mix.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cluster1.png" alt=""></p>
<h3 id="Centroid-based_clustering">Centroid-based clustering</h3><p>In centroid-based clustering, clusters are represented by a central vector, which may not necessarily be a member of the data set. When the number of clusters is fixed to k, k-means clustering gives a formal definition as an optimization problem: find the k cluster centers and assign the objects to the nearest cluster center, such that the squared distances from the cluster are minimized.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cluster2.png" alt=""></p>
<h3 id="Distribution-based_clustering">Distribution-based clustering</h3><p>The clustering model most closely related to statistics is based on distribution models. Clusters can then easily be defined as objects belonging most likely to the same distribution. A convenient property of this approach is that this closely resembles the way artificial data sets are generated: by sampling random objects from a distribution.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cluster3.png" alt=""></p>
<h3 id="Density-based_clustering">Density-based clustering</h3><p>The most popular density based clustering method is DBSCAN. In contrast to many newer methods, it features a well-defined cluster model called “density-reachability”. Similar to linkage based clustering, it is based on connecting points within certain distance thresholds. However, it only connects points that satisfy a density criterion, in the original variant defined as a minimum number of other objects within this radius.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cluster4.png" alt=""><br>综上内容看出其和主成份分析(Principal component analysis,PCA)有较多相似性，所以开始前先对<a href="http://tiramisutes.github.io/2015/11/11/pca-f-c/" target="_blank">主成份分析、因子分析和聚类分析的异同点</a>进行比较分析，具体关于PCA分析可看本博博文<a href="http://tiramisutes.github.io/2015/07/29/PCA/" target="_blank">PCA分析</a>。</p>
<h2 id="R中进行聚类分析">R中进行聚类分析</h2><h3 id="cluster包">cluster包</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">&gt; library(cluster)</span><br><span class="line">&gt; <span class="comment">#载入所需数据</span></span><br><span class="line">&gt; data(votes.repub)</span><br><span class="line">&gt; votes.diss &lt;- daisy(votes.repub)</span><br><span class="line">&gt; pamv &lt;- pam(votes.diss, 2, diss = TRUE)</span><br><span class="line">&gt; clusplot(pamv, shade = TRUE)</span><br><span class="line">&gt; <span class="comment">## is the same as</span></span><br><span class="line">&gt; votes.clus &lt;- pamv<span class="variable">$clustering</span></span><br><span class="line">&gt; clusplot(votes.diss, votes.clus, diss = TRUE, shade = TRUE)</span><br><span class="line">&gt; <span class="comment">##Remove the dotted line</span></span><br><span class="line">&gt; clusplot(votes.diss, votes.clus, diss = TRUE)</span><br><span class="line">&gt; <span class="comment">## show label</span></span><br><span class="line">&gt; op &lt;- par(new=TRUE, cex = 0.6)</span><br><span class="line">&gt; clusplot(votes.diss, votes.clus, diss = TRUE,</span><br><span class="line">+          axes=FALSE,ann=FALSE, sub=<span class="string">""</span>, col.p=NA, col.txt=<span class="string">"dark green"</span>, labels=3)</span><br><span class="line">&gt; par(op)</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cluster5.png" alt=""><br><i class="fa fa-slideshare fa-2x"></i>参数解释：<br><i class="fa fa-cogs"></i>daisy:Dissimilarity Matrix(相异度矩阵)Calculation,compute all the pairwise dissimilarities (distances) between observations in the data set.<br>相异度矩阵：相异度矩阵是对象—对象结构的一种数据表达方式，多数聚类算法都是建立在相异度矩阵基础上，如果数据是以数据矩阵形式给出的，就要将数据矩阵转化为相异度矩阵。对象间的相似度或相异度是基于两个对象间的距离来计算的。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">daisy(x, metric = c(<span class="string">"euclidean"</span>, <span class="string">"manhattan"</span>, <span class="string">"gower"</span>),stand = FALSE, <span class="built_in">type</span> = list(), weights = rep.int(1, p))</span><br></pre></td></tr></table></figure></p>
<p>x—numeric matrix or data frame, of dimension n*p.<br>metric—“euclidean” (the default), “manhattan” and “gower”.Euclidean distances are root sum-of-squares of differences, and manhattan distances<br>are the sum of absolute differences.”Gower’s distance” is chosen by metric “gower” or automatically if some columns of x are not numeric.<br>stand—logical flag: if TRUE, then the measurements in x are standardized before calculating the dissimilarities. Measurements are standardized for each variable (column), by subtracting the variable’s mean value and dividing by the variable’s mean absolute deviation.<br><i class="fa fa-cogs"></i>pam:Partitioning (clustering) of the data into k clusters “around medoids”, a more robust version of K-means.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">pam(x, k, diss = inherits(x, <span class="string">"dist"</span>), metric = <span class="string">"euclidean"</span>,</span><br><span class="line">medoids = NULL, stand = FALSE, cluster.only = FALSE,</span><br><span class="line">do.swap = TRUE,</span><br><span class="line">keep.diss = !diss &amp;&amp; !cluster.only &amp;&amp; n &lt; 100,</span><br><span class="line">keep.data = !diss &amp;&amp; !cluster.only,</span><br><span class="line">pamonce = FALSE, trace.lev = 0)</span><br></pre></td></tr></table></figure></p>
<p>x—data matrix or data frame, or dissimilarity matrix or object.In case of a dissimilarity matrix, x is typically the output of daisy or dist.<br>k—positive integer specifying the number of clusters, less than the number of observations.<br>diss—logical flag: if TRUE (default for dist or dissimilarity objects), then x will be considered as a dissimilarity matrix. If FALSE, then x will be considered as a matrix of observations by variables.<br>metric—“euclidean” and “manhattan”,If x is already a dissimilarity matrix, then this argument will be ignored.<br><i class="fa fa-cogs"></i>clusplot:Bivariate Cluster Plot (clusplot) Default Method<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">clusplot(x, clus, diss = FALSE,</span><br><span class="line">s.x.2d = mkCheckX(x, diss), stand = FALSE,</span><br><span class="line">lines = 2, shade = FALSE, color = FALSE,</span><br><span class="line">labels= 0, plotchar = TRUE,</span><br><span class="line">col.p = <span class="string">"dark green"</span>, col.txt = col.p,</span><br><span class="line">col.clus = <span class="keyword">if</span>(color) c(2, 4, 6, 3) <span class="keyword">else</span> 5, cex = 1, cex.txt = cex,</span><br><span class="line">span = TRUE,</span><br><span class="line">add = FALSE,</span><br><span class="line">xlim = NULL, ylim = NULL,</span><br><span class="line">main = paste(<span class="string">"CLUSPLOT("</span>, deparse(substitute(x)),<span class="string">")"</span>),</span><br><span class="line">sub = paste(<span class="string">"These two components explain"</span>,</span><br><span class="line">round(100 * var.dec, digits = 2), <span class="string">"% of the point variability."</span>),</span><br><span class="line">xlab = <span class="string">"Component 1"</span>, ylab = <span class="string">"Component 2"</span>,</span><br><span class="line">verbose = getOption(<span class="string">"verbose"</span>),</span><br><span class="line">...)</span><br></pre></td></tr></table></figure></p>
<p>x—data matrix or data frame, or dissimilarity matrix or object.In case of a dissimilarity matrix, x is typically the output of daisy or dist.<br>clus—clus is often the clustering component of the output of pam.<br>diss—logical flag: if TRUE (default for dist or dissimilarity objects), then x will be considered as a dissimilarity matrix. If FALSE, then x will be considered as a matrix of observations by variables.<br>lines—lines = 0, no distance lines will appear on the plot;lines = 1, the line segment between m1 and m2 is drawn;lines = 2, a line segment between the boundaries of E1 and E2 is drawn (along the line connecting m1 and m2).<br>shade—logical flag: if TRUE, then the ellipses are shaded in relation to their density.<br>color—logical flag: if TRUE, then the ellipses are colored with respect to their density.labels—labels= 0, no labels are placed in the plot;<br>labels= 1, points and ellipses can be identified in the plot (see identify);labels= 2, all points and ellipses are labelled in the plot;labels= 3, only the points are labelled in the plot;labels= 4, only the ellipses are labelled in the plot.labels= 5, the ellipses are labelled in the plot, and points can be identified.<br>col.p—color code(s) used for the observation points.</p>
<h2 id="更多资源">更多资源</h2><p><span class="exturl" data-url="aHR0cHM6Ly9jcmFuLnItcHJvamVjdC5vcmcvd2ViL3ZpZXdzL0NsdXN0ZXIuaHRtbA==" title="https://cran.r-project.org/web/views/Cluster.html">https://cran.r-project.org/web/views/Cluster.html<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/08/fa-length.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/08/fa-length.html" class="post-title-link" itemprop="url">批量求fasta格式序列长度</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-08 22:14:22" itemprop="dateCreated datePublished" datetime="2015-11-08T22:14:22+08:00">2015-11-08</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/08/fa-length.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/08/fa-length.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/08/fa-length.html" class="leancloud_visitors" data-flag-title="批量求fasta格式序列长度">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>linux下用awk计算fasta序列的长度<br>fasta序列文件data.fa<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt;Gorai.004G111100.1</span><br><span class="line">ATGGGTACTGCTCCAACCCAGTGCCCTTCTGGAATCACTGCAAATTTCCACGCCAAATTTGATAACAGAACTGAGTTTTC</span><br><span class="line">&gt;Gorai.004G111100.2</span><br><span class="line">ATGTTTTTCATGCTCCGGTGGACAAGATACTCTGGGATGCCGGGGAACAGTTTTTCCTTTTCTTGGCAGACATATGCACATAAAATTCTT</span><br><span class="line">&gt;Gorai.004G111100.3</span><br><span class="line">ATGGGTACTGCTCCAACCCAGTGCCCTTCTGGAATCACTGCAAATTTCCAC</span><br><span class="line">&gt;Gorai.004G111100.4</span><br><span class="line">ATGGGAATGCATGAACTAGCAGCCAAAGTTGATGAGT</span><br></pre></td></tr></table></figure></p>
<p>首先将fasta序列转换成一行显示，命令如下：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk <span class="string">'/^&gt;/&amp;&amp;NR&gt;1&#123;print "";&#125;&#123; printf "%s",/^&gt;/ ? $0"%":$0 &#125;'</span>  data.fa &gt;data2.fa</span><br></pre></td></tr></table></figure></p>
<p>结果：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt;Gorai.004G111100.1%ATGGGTACTGCTCCAACCCAGTGCCCTTCTGGAATCACTGCAAATTTCCACGCCAAATTTGATAACAGAACTGAGTTTTC</span><br><span class="line">&gt;Gorai.004G111100.2%ATGTTTTTCATGCTCCGGTGGACAAGATACTCTGGGATGCCGGGGAACAGTTTTTCCTTTTCTTGGCAGACATATGCACATAAAATTCTT</span><br><span class="line">&gt;Gorai.004G111100.3%ATGGGTACTGCTCCAACCCAGTGCCCTTCTGGAATCACTGCAAATTTCCAC</span><br><span class="line">&gt;Gorai.004G111100.4%ATGGGAATGCATGAACTAGCAGCCAAAGTTGATGAGT</span><br></pre></td></tr></table></figure></p>
<p>长度计算：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -F<span class="string">"%"</span> <span class="string">'&#123;print $1"\t"length($2)&#125;'</span>  data2.fa &gt;data3.fa</span><br></pre></td></tr></table></figure></p>
<p>结果：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt;Gorai.004G111100.1 80</span><br><span class="line">&gt;Gorai.004G111100.2 90</span><br><span class="line">&gt;Gorai.004G111100.3 51</span><br><span class="line">&gt;Gorai.004G111100.4 37</span><br></pre></td></tr></table></figure></p>
<p>More： <span class="exturl" data-url="aHR0cHM6Ly93d3cuYmlvc3RhcnMub3JnL3AvOTI2Mi8jOTI2NA==" title="https://www.biostars.org/p/9262/#9264">Question: Multiline Fasta To Single Line Fasta<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/08/R-fun.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/08/R-fun.html" class="post-title-link" itemprop="url">R中的基础函数</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-08 16:53:00" itemprop="dateCreated datePublished" datetime="2015-11-08T16:53:00+08:00">2015-11-08</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/08/R-fun.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/08/R-fun.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/08/R-fun.html" class="leancloud_visitors" data-flag-title="R中的基础函数">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>在R中偶尔也需要对一些数据进行简短的处理，掌握一些基本函数是必须的，本文将持续收集那些短小精悍的R函数，正确的运用还是能起到四两拔千斤的效果，欢迎评论补充。</p>
<h3 id="aggregate">aggregate</h3><p>功能:aggregate(formula, data, FUN)<br>首先将数据进行分组（按行），然后对每一组数据进行函数统计，最后把结果组合成一个比较nice的表格返回.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(chickwts)</span><br><span class="line">  weight      feed</span><br><span class="line">1    179 horsebean</span><br><span class="line">2    160 horsebean</span><br><span class="line">3    136 horsebean</span><br><span class="line">4    227 horsebean</span><br><span class="line">5    217 horsebean</span><br><span class="line">6    168 horsebean</span><br><span class="line">&gt; unique(chickwts<span class="variable">$feed</span>)</span><br><span class="line">[1] horsebean linseed   soybean   sunflower meatmeal  casein   </span><br><span class="line">Levels: casein horsebean linseed meatmeal soybean sunflower</span><br><span class="line">&gt; <span class="comment">#aggregate(chickwts$weight, by=list(chickwts$feed), FUN=mean)</span></span><br><span class="line">&gt; aggregate(weight ~ feed, data = chickwts, mean)</span><br><span class="line">       feed   weight</span><br><span class="line">1    casein 323.5833</span><br><span class="line">2 horsebean 160.2000</span><br><span class="line">3   linseed 218.7500</span><br><span class="line">4  meatmeal 276.9091</span><br><span class="line">5   soybean 246.4286</span><br><span class="line">6 sunflower 328.9167</span><br><span class="line">&gt; head(iris)</span><br><span class="line">  Sepal.Length Sepal.Width Petal.Length Petal.Width Species</span><br><span class="line">1          5.1         3.5          1.4         0.2  setosa</span><br><span class="line">2          4.9         3.0          1.4         0.2  setosa</span><br><span class="line">3          4.7         3.2          1.3         0.2  setosa</span><br><span class="line">4          4.6         3.1          1.5         0.2  setosa</span><br><span class="line">5          5.0         3.6          1.4         0.2  setosa</span><br><span class="line">6          5.4         3.9          1.7         0.4  setosa</span><br><span class="line">&gt; unique(iris<span class="variable">$Species</span>)</span><br><span class="line">[1] setosa     versicolor virginica </span><br><span class="line">Levels: setosa versicolor virginica</span><br><span class="line">&gt; aggregate(. ~ Species, data = iris, mean)</span><br><span class="line">     Species Sepal.Length Sepal.Width Petal.Length Petal.Width</span><br><span class="line">1     setosa        5.006       3.428        1.462       0.246</span><br><span class="line">2 versicolor        5.936       2.770        4.260       1.326</span><br><span class="line">3  virginica        6.588       2.974        5.552       2.026</span><br></pre></td></tr></table></figure></p>
<h3 id="paste">paste</h3><p>功能：paste(…, sep = “ “, collapse = NULL)<br>字符串连接<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; paste(<span class="string">"CK"</span>, 1:6, sep = <span class="string">""</span>)</span><br><span class="line">[1] <span class="string">"CK1"</span> <span class="string">"CK2"</span> <span class="string">"CK3"</span> <span class="string">"CK4"</span> <span class="string">"CK5"</span> <span class="string">"CK6"</span></span><br><span class="line">&gt; <span class="comment">#设置collapse参数，连成一个字符串</span></span><br><span class="line">&gt; paste(<span class="string">"CK"</span>, 1:6, sep = <span class="string">""</span>, collapse = <span class="string">"-"</span>)</span><br><span class="line">[1] <span class="string">"CK1-CK2-CK3-CK4-CK5-CK6"</span></span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-binoculars" aria-hidden="true"></i>paste在不指定分割符的情况下，默认分割符是空格 ，paste0在不指定分割符的情况下，默认分割符是空。</p>
<h3 id="strsplit">strsplit</h3><p>功能：strsplit(x, split, fixed = FALSE, perl = FALSE, useBytes = FALSE)<br>字符串拆分，生成一个list<br>参数解释：<br>x为字串向量，每个元素都将单独进行拆分。<br>split为拆分位置的字串向量（分隔符），默认为正则表达式匹配（fixed=FALSE）。如果你没接触过正则表达式，设置fixed=TRUE，表示使用普通文本匹配或正则表达式的精确匹配。普通文本的运算速度快。<br>perl=TRUE/FALSE的设置和perl语言版本有关，如果正则表达式很长，正确设置表达式并且使用perl=TRUE可以提高运算速度。<br>useBytes设置是否逐个字节进行匹配，默认为FALSE，即按字符而不是字节进行匹配。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; text &lt;- <span class="string">"Hello Adam!\nHello Ava!"</span></span><br><span class="line">&gt; strsplit(text, <span class="string">"\\s"</span>)</span><br><span class="line">[[1]]</span><br><span class="line">[1] <span class="string">"Hello"</span> <span class="string">"Adam!"</span> <span class="string">"Hello"</span> <span class="string">"Ava!"</span></span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-bullhorn" aria-hidden="true"></i>如果要对一个向量使用该函数，需要注意。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#分割向量的每一个元素，并取分割后的第一个元素</span></span><br><span class="line">unlist(lapply(X = c(<span class="string">"abc"</span>, <span class="string">"bcd"</span>, <span class="string">"dfafadf"</span>), FUN = <span class="keyword">function</span>(x) &#123;<span class="built_in">return</span>(strsplit(x, split = <span class="string">""</span>)[[1]][1])&#125;))</span><br><span class="line">[1] <span class="string">"a"</span> <span class="string">"b"</span> <span class="string">"d"</span></span><br></pre></td></tr></table></figure></p>
<h3 id="grep/regexpr/gregexpr/regexec">grep/regexpr/gregexpr/regexec</h3><p>功能:grep(pattern, x, ignore.case = FALSE, perl = FALSE, value = FALSE, fixed = FALSE, useBytes = FALSE, invert = FALSE)<br>grep仅返回匹配项的下标.<br>regexpr、gregexpr和regexec返回的结果包含了匹配的具体位置和字符串长度信息.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line">&gt; text &lt;- c(<span class="string">"Hellow, Adam!"</span>, <span class="string">"Hi, Adam!"</span>, <span class="string">"How are you, Adam."</span>)</span><br><span class="line">&gt; text</span><br><span class="line">[1] <span class="string">"Hellow, Adam!"</span>      <span class="string">"Hi, Adam!"</span>          <span class="string">"How are you, Adam."</span></span><br><span class="line">&gt; grep(<span class="string">"Adam"</span>,text)</span><br><span class="line">[1] 1 2 3</span><br><span class="line">&gt; regexpr(<span class="string">"Adam"</span>, text)</span><br><span class="line">[1]  9  5 14</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4 4 4</span><br><span class="line">attr(,<span class="string">"useBytes"</span>)</span><br><span class="line">[1] TRUE</span><br><span class="line">&gt; gregexpr(<span class="string">"Adam"</span>, text)</span><br><span class="line">[[1]]</span><br><span class="line">[1] 9</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br><span class="line">attr(,<span class="string">"useBytes"</span>)</span><br><span class="line">[1] TRUE</span><br><span class="line"></span><br><span class="line">[[2]]</span><br><span class="line">[1] 5</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br><span class="line">attr(,<span class="string">"useBytes"</span>)</span><br><span class="line">[1] TRUE</span><br><span class="line"></span><br><span class="line">[[3]]</span><br><span class="line">[1] 14</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br><span class="line">attr(,<span class="string">"useBytes"</span>)</span><br><span class="line">[1] TRUE</span><br><span class="line"></span><br><span class="line">&gt; regexec(<span class="string">"Adam"</span>, text)</span><br><span class="line">[[1]]</span><br><span class="line">[1] 9</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br><span class="line"></span><br><span class="line">[[2]]</span><br><span class="line">[1] 5</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br><span class="line"></span><br><span class="line">[[3]]</span><br><span class="line">[1] 14</span><br><span class="line">attr(,<span class="string">"match.length"</span>)</span><br><span class="line">[1] 4</span><br></pre></td></tr></table></figure></p>
<h3 id="substr">substr</h3><p>功能：substr(x, start, stop)<br>字符串提取<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; substr(text,9,12)</span><br><span class="line">[1] <span class="string">"Adam"</span> <span class="string">"!"</span>    <span class="string">"you,"</span></span><br></pre></td></tr></table></figure></p>
<h3 id="strtrim">strtrim</h3><p>功能：strtrim(x, width)<br>将字符串修剪到特定的显示宽度.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(iris)</span><br><span class="line">  Sepal.Length Sepal.Width Petal.Length Petal.Width Species</span><br><span class="line">1            5         3.5          1.4         0.2  setosa</span><br><span class="line">2            4         3.0          1.4         0.2  setosa</span><br><span class="line">3            4         3.2          1.3         0.2  setosa</span><br><span class="line">4            4         3.1          1.5         0.2  setosa</span><br><span class="line">5            5         3.6          1.4         0.2  setosa</span><br><span class="line">6            5         3.9          1.7         0.4  setosa</span><br><span class="line">&gt; strtrim(head(iris)<span class="variable">$Sepal</span>.Length,1)</span><br><span class="line">[1] <span class="string">"5"</span> <span class="string">"4"</span> <span class="string">"4"</span> <span class="string">"4"</span> <span class="string">"5"</span> <span class="string">"5"</span></span><br><span class="line">&gt; iris<span class="variable">$Sepal</span>.Length&lt;-strtrim(head(iris)<span class="variable">$Sepal</span>.Length,1)</span><br><span class="line">&gt; head(iris)</span><br><span class="line">  Sepal.Length Sepal.Width Petal.Length Petal.Width Species</span><br><span class="line">1            5         3.5          1.4         0.2  setosa</span><br><span class="line">2            4         3.0          1.4         0.2  setosa</span><br><span class="line">3            4         3.2          1.3         0.2  setosa</span><br><span class="line">4            4         3.1          1.5         0.2  setosa</span><br><span class="line">5            5         3.6          1.4         0.2  setosa</span><br><span class="line">6            5         3.9          1.7         0.4  setosa</span><br></pre></td></tr></table></figure></p>
<h3 id="length/nchar">length/nchar</h3><p>nchar是向量元素的字符个数，而length是向量长度.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">&gt; length(<span class="string">"ATGGGAATGCATGAACTAGCAGCCAAAGTTGATGAGT"</span>)</span><br><span class="line">[1] 1</span><br><span class="line">&gt; car &lt;- c(<span class="string">'bmw'</span>,<span class="string">'ford'</span>,<span class="string">'mini'</span>,<span class="string">'bmw'</span>,<span class="string">'mini'</span>)</span><br><span class="line">&gt; length(car)</span><br><span class="line">[1] 5</span><br><span class="line">&gt; length(unique(car))</span><br><span class="line">[1] 3</span><br><span class="line">&gt; nchar(<span class="string">"ATGGGAATGCATGAACTAGCAGCCAAAGTTGATGAGT"</span>)</span><br><span class="line">[1] 37</span><br></pre></td></tr></table></figure></p>
<h3 id="round">round</h3><p>功能：round(x, digits = 0)<br>四舍五入<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; round(c(1.1254,0.1247844),3)</span><br><span class="line">[1] 1.125 0.125</span><br></pre></td></tr></table></figure></p>
<h3 id="axes/axis">axes/axis</h3><p>axes=FALSE       暂时禁止坐标轴的生成|以便使用axis()函数添加你自己定制的坐标轴。默认情况是axes=TRUE，即包含坐标轴。<br>axis(side, . . . )<br>在当前图形的指定边上添加坐标，在哪个边上由第一个参数指定（1到4，从底部按照顺时针顺序）。其他参数控制坐标的位置|在图形内或图形外，以及标记的位置和标签。适合在调用参数为axes=FALSE的函数plot()后添加定制的坐标轴。</p>
<h3 id="order()">order()</h3><p>A[order(A[,4],decreasing=T),] ＃按照第4列降序排序<br>data  #dataframe对象 含有v1,v2两列<br>data[sort(data$v1,index.return=TRUE)$ix,]　　#对data的数据按v1排列,v1须为numeric  as.numeric()</p>
<h3 id="%in%">%in%</h3><p>功能：在数据框中选取某一列只含特定字符的行<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">&gt; library(dplyr)</span><br><span class="line">&gt; library(tidyr)</span><br><span class="line">&gt; df &lt;- data_frame(</span><br><span class="line">+   group = c(1:2, 1),</span><br><span class="line">+   item_name = c(<span class="string">"a"</span>, <span class="string">"b"</span>, <span class="string">"b"</span>),</span><br><span class="line">+   value1 = 1:3,</span><br><span class="line">+   value2 = 4:6</span><br><span class="line">+ )</span><br><span class="line">&gt; df</span><br><span class="line">Source: <span class="built_in">local</span> data frame [3 x 4]</span><br><span class="line"></span><br><span class="line">  group item_name value1 value2</span><br><span class="line">  (dbl)     (chr)  (int)  (int)</span><br><span class="line">1     1         a      1      4</span><br><span class="line">2     2         b      2      5</span><br><span class="line">3     1         b      3      6</span><br><span class="line">&gt;<span class="comment">#选取item_name列中只含有a的行</span></span><br><span class="line">&gt; a&lt;-c(<span class="string">"a"</span>)</span><br><span class="line">&gt; df[df<span class="variable">$item_name</span> %<span class="keyword">in</span>% a,]</span><br><span class="line">Source: <span class="built_in">local</span> data frame [1 x 4]</span><br><span class="line"></span><br><span class="line">  group item_name value1 value2</span><br><span class="line">  (dbl)     (chr)  (int)  (int)</span><br><span class="line">1     1         a      1      4</span><br></pre></td></tr></table></figure></p>
<h3 id="table">table</h3><p>功能：统计数据的频数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; a&lt;-c(1,1,1,2,2,3)</span><br><span class="line">&gt; table(a)</span><br><span class="line">a</span><br><span class="line">1 2 3 </span><br><span class="line">3 2 1</span><br></pre></td></tr></table></figure></p>
<h3 id="str()">str()</h3><p>功能：查看数据结构<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">&gt; str(reps )</span><br><span class="line"><span class="string">'data.frame'</span>:	255956 obs. of  17 variables:</span><br><span class="line"> $ bin      : int  585 585 585 585 585 585 585 585 585 585 ...</span><br><span class="line"> $ swScore  : int  342 2271 3379 2704 180 380 1113 233 388 673 ...</span><br><span class="line"> $ milliDiv : int  0 159 80 108 0 131 168 302 256 182 ...</span><br><span class="line"> $ milliDel : int  0 37 4 31 0 0 19 0 0 7 ...</span><br><span class="line"> $ milliIns : int  0 25 0 10 0 14 19 68 14 90 ...</span><br><span class="line"> $ genoName : Factor w/ 1 level <span class="string">"chrX"</span>: 1 1 1 1 1 1 1 1 1 1 ...</span><br><span class="line"> $ genoStart: int  0 41 1799 2290 2797 2945 3015 3565 5012 5164 ...</span><br><span class="line"> $ genoEnd  : int  38 446 2272 2703 2817 3015 3221 3757 5164 5186 ...</span><br><span class="line"> $ genoLeft : int  -154913716 -154913308 -154911482 -154911051 -154910937 -154910739 -154910533 -154909997 -154908590 -154908568 ...</span><br><span class="line"> $ strand   : Factor w/ 2 levels <span class="string">"-"</span>,<span class="string">"+"</span>: 2 2 2 2 2 2 1 2 1 2 ...</span><br><span class="line"> $ repName  : Factor w/ 1105 levels <span class="string">"(A)n"</span>,<span class="string">"(AAATG)n"</span>,..: 90 519 656 579 90 183 248 284 940 240 ...</span><br><span class="line"> $ repClass : Factor w/ 15 levels <span class="string">"DNA"</span>,<span class="string">"LINE"</span>,<span class="string">"Low_complexity"</span>,..: 10 4 4 4 10 10 11 3 4 11 ...</span><br><span class="line"> $ repFamily: Factor w/ 35 levels <span class="string">"AcHobo"</span>,<span class="string">"Alu"</span>,..: 28 7 7 7 28 28 2 12 13 2 ...</span><br><span class="line"> $ repStart : int  3 741 1 1 2 1 -20 1 -200 1 ...</span><br><span class="line"> $ repEnd   : int  40 1150 475 422 21 69 292 179 228 22 ...</span><br><span class="line"> $ repLeft  : int  0 -104 -83 -79 0 0 87 0 37 -280 ...</span><br><span class="line"> $ id       : int  1 2 3 4 5 6 7 8 9 1 ...</span><br></pre></td></tr></table></figure></p>
<h3 id="ifelse">ifelse</h3><p>功能：条件判断<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">&gt; set.seed(123)</span><br><span class="line">&gt; col1 &lt;- runif (5, 0, 2)</span><br><span class="line">&gt; col2 &lt;- rnorm (5, 0, 2)</span><br><span class="line">&gt; col3 &lt;- rpois (5, 3)</span><br><span class="line">&gt; col4 &lt;- rchisq (5, 0.1)</span><br><span class="line">&gt; df &lt;- data.frame (col1, col2, col3, col4)</span><br><span class="line">&gt; df</span><br><span class="line">       col1       col2 col3         col4</span><br><span class="line">1 0.5751550 -3.3791113    5 2.771082e-01</span><br><span class="line">2 1.5766103  2.4789918    2 3.888853e-04</span><br><span class="line">3 0.8179538 -0.2179319    0 8.652702e-05</span><br><span class="line">4 1.7660348 -0.2344839    2 6.492406e-17</span><br><span class="line">5 1.8809346  0.3661652    6 2.963428e-02</span><br><span class="line">&gt; output &lt;- ifelse ((df<span class="variable">$col1</span>) &gt; 1 &amp; (df<span class="variable">$col3</span>) &gt; 2, <span class="string">"yes"</span>, <span class="string">"no"</span>)</span><br><span class="line">&gt; df<span class="variable">$output</span> &lt;- output</span><br><span class="line">&gt; df</span><br><span class="line">       col1       col2 col3         col4 output</span><br><span class="line">1 0.5751550 -3.3791113    5 2.771082e-01     no</span><br><span class="line">2 1.5766103  2.4789918    2 3.888853e-04     no</span><br><span class="line">3 0.8179538 -0.2179319    0 8.652702e-05     no</span><br><span class="line">4 1.7660348 -0.2344839    2 6.492406e-17     no</span><br><span class="line">5 1.8809346  0.3661652    6 2.963428e-02    yes</span><br></pre></td></tr></table></figure></p>
<h3 id="gsub和sub">gsub和sub</h3><p>字符串替换<br>gsub替换匹配到的全部<br>sub 替换匹配到的第一个</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 将b替换为B</span></span><br><span class="line">gsub(pattern = <span class="string">"b"</span>, replacement = <span class="string">"B"</span>, x = c(<span class="string">"abcb"</span>, <span class="string">"boy"</span>, <span class="string">"baby"</span>))</span><br><span class="line">[1] <span class="string">"aBcB"</span> <span class="string">"Boy"</span>  <span class="string">"BaBy</span></span><br><span class="line"><span class="string"># 只替换第一个b</span></span><br><span class="line"><span class="string">sub(pattern = "</span>b<span class="string">", replacement = "</span>B<span class="string">", x = c("</span>abcb<span class="string">", "</span>baby<span class="string">"))</span></span><br><span class="line"><span class="string">[1] "</span>aBcb<span class="string">" "</span>Baby<span class="string">"</span></span><br></pre></td></tr></table></figure>
<h3 id="字符串中字符统计">字符串中字符统计</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">s &lt;- <span class="string">"aababac"</span></span><br><span class="line">p &lt;- <span class="string">"a"</span></span><br><span class="line">countCharOccurrences &lt;- <span class="keyword">function</span>(char, s) &#123;</span><br><span class="line">    s2 &lt;- gsub(char,<span class="string">""</span>,s)</span><br><span class="line">    <span class="built_in">return</span> (nchar(s) - nchar(s2))</span><br><span class="line">&#125;</span><br><span class="line">countCharOccurrences(p,s)</span><br></pre></td></tr></table></figure>
          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/10/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/10/">10</a><span class="page-number current">11</span><a class="page-number" href="/page/12/">12</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/12/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      

      <div class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <a href="/" class="site-author-image" rel="start" style="border:none">
              <img class="site-author-image" itemprop="image"
                src="/images/author.jpg"
                alt="tiramisutes"/>
              </a>
            
              <p class="site-author-name" itemprop="name">tiramisutes</p>
              <div class="site-description motion-element" itemprop="description">hope bioinformatics blog</div>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">148</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  
                    
                      <a href="/categories/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">17</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                      <a href="/tags/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">112</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly9naXRodWIuY29tL3RpcmFtaXN1dGVz" title="GitHub &rarr; https://github.com/tiramisutes"><i class="fa fa-fw fa-github"></i>GitHub</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="bWFpbHRvOnRpcmFtaXN1dGVze2F0fTEyNi5jb20=" title="E-Mail &rarr; mailto:tiramisutes{at}126.com"><i class="fa fa-fw fa-envelope"></i>E-Mail</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cDovL3dlaWJvLmNvbS90aXJhbWlzdXRlc1Rlcw==" title="Weibo &rarr; http://weibo.com/tiramisutesTes"><i class="fa fa-fw fa-globe"></i>Weibo</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly90d2l0dGVyLmNvbS9ob3BldG9neQ==" title="Twitter &rarr; https://twitter.com/hopetogy"><i class="fa fa-fw fa-twitter"></i>Twitter</span>
                </span>
              
            </div>
          

          
             <div class="cc-license motion-element" itemprop="license">
              
              
                
              
              
              
              <span class="exturl cc-opacity" data-url="aHR0cHM6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LW5jLXNhLzQuMC8="><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"/></span>
             </div>
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-inline">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8=" title="https://ihope.shinyapps.io/qRT-PCR-Pipeline/">qRT-PCR-Pipeline</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly90aXJhbWlzdXRlcy5naXRib29rcy5pby9hd2svY29udGVudC8=" title="https://tiramisutes.gitbooks.io/awk/content/">GitBook《awk学习总结》</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2NvdHRvbi5oemF1LmVkdS5jbi8=" title="http://cotton.hzau.edu.cn/">棉花遗传改良团队</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW8taW5mby10cmFpbmVlLmNvbS8=" title="http://www.bio-info-trainee.com/">生信菜鸟团-深圳大学城</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW90cmFpbmVlLmNvbS9mb3J1bS5waHA=" title="http://www.biotrainee.com/forum.php">生信技能树</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5pd2hnYW8uY29tLw==" title="http://www.iwhgao.com/">rabbit gao's blog</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9odWFucy5naXRodWIuaW8v" title="https://huans.github.io/">Huans(代谢+GWAS)</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5zZXF1ZW5jaW5nLnVpby5uby8=" title="http://www.sequencing.uio.no/">NSC-sequenceing</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2Jsb2cuam9iYm9sZS5jb20v" title="http://blog.jobbole.com/">伯乐在线</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9saW51eC5jbi8=" title="https://linux.cn/">Linux开源中文社区</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5yLWJsb2dnZXJzLmNvbS8=" title="http://www.r-bloggers.com/">R-bloggers</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9jcmFuLnItcHJvamVjdC5vcmcvb3RoZXItZG9jcy5odG1s" title="https://cran.r-project.org/other-docs.html">R-resource</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3R1dG9yaWFscy5pcS5oYXJ2YXJkLmVkdS9SL1JncmFwaGljcy9SZ3JhcGhpY3MuaHRtbA==" title="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html">ggplot2</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2xpbGliZWkubmV0Lw==" title="http://lilibei.net/">pele' blog</span>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </div>

      

      

    </div>
  </aside>
  


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2014 – <span itemprop="copyrightYear">2023</span>
  <span class="with-love" id="animate">
    <i class="fa fa-heart"></i>
  </span>
  <a class="theme-link" target="_blank" href="http://tiramisutes.github.io/about/"><span class="author" itemprop="copyrightHolder">tiramisutes</span></a>

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Symbols count total"></span>
  

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Reading time total"></span>
  
</div>


  <div class="powered-by">Powered: <span class="exturl theme-link" data-url="aHR0cHM6Ly9oZXhvLmlv">Hexo</span></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">▽ – <span class="exturl theme-link" data-url="aHR0cHM6Ly9naXRodWIuY29tL25leHQtdGhlbWUvaGV4by10aGVtZS1uZXh0">NexT.Muse</span></div>




        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  

  

  
    <span class="post-meta-item-icon">
      <i class="fa fa-line-chart"></i>
    </span>
    <span class="site-pv" title="Total Views">
    <a style="line-height: 1.5;" href="https://tongji.baidu.com/web/welcome/login" target="_blank">The total visits
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>&nbsp;times</a>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>












  



  
    
    
      
    
  
  <script color='0,0,255' opacity='0.5' zIndex='-1' count='20' src="//cdn.jsdelivr.net/gh/theme-next/theme-next-canvas-nest@1/canvas-nest-nomobile.min.js"></script>













  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.js"></script>


  


  <script src="/js/src/utils.js?v=7.0.1"></script>

  <script src="/js/src/motion.js?v=7.0.1"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.1"></script>



  

  


  <script src="/js/src/next-boot.js?v=7.0.1"></script>


  

  
  <script src="/js/src/exturl.js?v=7.0.1"></script>


  
  
  
    
  

  
    
      
    
  

  
  
  
    <script src="//cdn.jsdelivr.net/npm/quicklink@1/dist/quicklink.umd.js"></script>
    <script>
      
        window.addEventListener('load', () => {
      
        quicklink({
          timeout: 3000,
          priority: true,
          ignores: [uri => uri.includes('#'),uri => uri == 'http://tiramisutes.github.io/page/11/',]
        });
      
        });
      
    </script>
  


  
  

  

<script src="//cdn.jsdelivr.net/npm/leancloud-storage@3/dist/av-min.js"></script>



<script src="//unpkg.com/valine/dist/Valine.min.js"></script>

<script>
  var GUEST = ['nick', 'mail', 'link'];
  var guest = 'nick,mail,link';
  guest = guest.split(',').filter(function(item) {
    return GUEST.indexOf(item) > -1;
  });
  new Valine({
    el: '#comments',
    verify: false,
    notify: true,
    appId: 'c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz',
    appKey: 'iVJ6ST0e1O5xaiIhVwOyhbf3',
    placeholder: 'Just go go',
    avatar: 'mm',
    meta: guest,
    pageSize: '10' || 10,
    visitor: true
  });
</script>





  
  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  

  
  
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  

  


  
<style>
.pdfobject-container {
  position: relative;
  overflow: auto;
  width: 100%;
}
</style>
<script>
if ($('body').find('div.pdf').length) {
  $.ajax({
    type: 'GET',
    url: '//cdn.jsdelivr.net/npm/pdfobject@2.1.1/pdfobject.min.js',
    dataType: 'script',
    cache: true,
    success: function() {
      $('body').find('div.pdf').each(function(i, o) {
        PDFObject.embed($(o).attr('target'), $(o), {
          pdfOpenParams: {
            navpanes: 0,
            toolbar: 0,
            statusbar: 0,
            pagemode: 'thumbs',
            view: 'FitH'
          },
          PDFJS_URL: '/lib/pdf/web/viewer.html',
          height: $(o).attr('height') || '500px'
        });
      });
    },
  });
}
</script>


  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      bp.src = (curProtocol === 'https') ? 'https://zz.bdstatic.com/linksubmit/push.js' : 'http://push.zhanzhang.baidu.com/push.js';
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  
  
  
    
  
  <script src="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.js"></script>
  <script>
    
      pbOptions = {};
      
        pbOptions.iconStyle = "box";
      
        pbOptions.boxForm = "horizontal";
      
        pbOptions.position = "bottomCenter";
      
        pbOptions.networks = "Weibo,Wechat,Douban,QQZone,Twitter,Evernote,Facebook";
      
      new needShareButton('#needsharebutton-postbottom', pbOptions);
    
    
  </script>


  

  

  

  

</body>
</html>
