<!DOCTYPE html>












  


<html class="theme-next muse use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge"/>
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222"/>
<script type="text/javascript" src="http://tajs.qq.com/stats?sId=66189719" charset="UTF-8"></script>


  
  
    
  
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.css"/>





  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/Han/3.3.0/han.min.css"/>



  <meta http-equiv="Cache-Control" content="no-transform"/>
  <meta http-equiv="Cache-Control" content="no-siteapp"/>



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg"/>



  <meta name="msvalidate.01" content="74A687E04D4EF9914B7AA20824207409"/>









  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.css"/>




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link rel="stylesheet" href="https://fonts.loli.net/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext"/>
  






<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2"/>

<link rel="stylesheet" href="/css/main.css?v=7.0.1"/>


  <link rel="apple-touch-icon" sizes="180x180" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/small_favicon.png?v=7.0.1">


  <link rel="mask-icon" href="/images/favicon.svg?v=7.0.1" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '7.0.1',
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":true,"dimmer":false},
    back2top: true,
    back2top_sidebar: false,
    fancybox: true,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="hope bioinformatics blog">
<meta property="og:type" content="website">
<meta property="og:title" content="hope">
<meta property="og:url" content="http://tiramisutes.github.io/page/8/index.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="hope bioinformatics blog">
<meta property="og:locale" content="en">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="hope">
<meta name="twitter:description" content="hope bioinformatics blog">



  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml"/>




  <link rel="canonical" href="http://tiramisutes.github.io/page/8/"/>



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>hope – A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</title>
  






  <script>
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>







  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br/>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br/>Archives</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-code">

    
    
    
      
    

    

    <a href="/code/" rel="section"><i class="menu-item-icon fa fa-fw fa-code"></i> <br/>Code</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br/>Categories</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br/>Tags</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-footprint">

    
    
    
      
    

    

    <a href="/footprint/" rel="section"><i class="menu-item-icon fa fa-fw fa-image"></i> <br/>Footprint</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-qrt-pcr">

    
    
    
      
    

    

    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8="><i class="menu-item-icon fa fa-fw fa-cloud"></i> <br/>qRT-PCR</span>

  </li>
        
        
        
          
          <li class="menu-item menu-item-project">

    
    
    
      
    

    

    <a href="/project/" rel="section"><i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br/>Project</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-publish">

    
    
    
      
    

    

    <a href="/publish/" rel="section"><i class="menu-item-icon fa fa-fw fa-book"></i> <br/>Publish</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-sci-figure">

    
    
    
      
    

    

    <a href="/SCIFigure/" rel="section"><i class="menu-item-icon fa fa-fw fa-angellist"></i> <br/>SCI Figure</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br/>About</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-cv">

    
    
    
      
    

    

    <a href="/resume/" rel="section"><i class="menu-item-icon fa fa-fw fa-address-card"></i> <br/>CV</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br/>Search</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/04/01/Summary-Statistics.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/04/01/Summary-Statistics.html" class="post-title-link" itemprop="url">Analyzing Data</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-04-01 14:17:34" itemprop="dateCreated datePublished" datetime="2016-04-01T14:17:34+08:00">2016-04-01</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/04/01/Summary-Statistics.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/04/01/Summary-Statistics.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/04/01/Summary-Statistics.html" class="leancloud_visitors" data-flag-title="Analyzing Data">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Summary_Statistics">Summary Statistics</h2><h3 id="mean(),max(),min(),range()，sd()">mean(),max(),min(),range()，sd()</h3><p>以上函数运算中，若存在NA，则返回结果为NA，设置na.rm=TRUE忽略NA值。<br>mean()中移除异常值：trim（<code>mean(x, trim = 0.1)</code>：先把x的最大的10%的数和最小的10%的数去掉，然后剩下的数算平均）<br>range():同时返回最大/最小值。<br>sd()：标准差。</p>
<h3 id="quantile(),fivenum()，IQR()">quantile(),fivenum()，IQR()</h3><p><code>quantile(dow30$Open, probs=c(0,0.25,0.5,0.75,1.0))</code><br>返回不同的百分位数值，probs指定百分位。<br>fivenum()：返回(minimum, 25th percentile, median, 75th percentile, and maximum)。<br>IQR()：返回25%与75%的差值<br>以上函数既可用于单独数组，也可用于apply, tapply对数据框的操作。</p>
<h3 id="summary()">summary()</h3><p>对于数值变量计算了五个分位点和均值，对于分类变量则计算了频数。</p>
<h2 id="Statistical_Tests">Statistical Tests</h2><p>基于<strong>正态分布</strong>的检验；自然群体大都为正态分布。</p>
<h3 id="Comparing_means">Comparing <blod>means</blod></h3><p>Specifically, suppose that you have a set of observations<br>x1, x2, …, xn with experimental mean μ and want to know if the experimental<br>mean is different from the null hypothesis mean μ0. Furthermore, assume that the<br>observations are normally distributed. To test the validity of the hypothesis, you can<br>use a t-test. In R, you would use the function <code>t.test</code>;</p>
<h3 id="Comparing_paired_data(means)">Comparing paired data(<blod>means</blod>)</h3><p>For example, you might have two observations per subject: one before an experiment and one after the experiment.<br>In this case, you would use a paired t-test. You can use the t.test function, specifying<br><code>paired=TRUE</code>, to perform this test.</p>
<h3 id="Comparing_variances_of_two_populations">Comparing <blod>variances</blod> of two populations</h3><p>To compare the variances of two samples from normal populations, R includes the<br><code>var.test</code> function which performs an <em>F-test</em>;</p>
<h3 id="Comparing_means_across_more_than_two_groups">Comparing means across more than two groups</h3><p>ANOVA单因素方差分析与R实现：<a href="http://tiramisutes.github.io/2015/10/08/ANOVA.html">http://tiramisutes.github.io/2015/10/08/ANOVA.html</a></p>
<h3 id="Correlation_tests"><strong>Correlation</strong> tests</h3><p>If you’d like to check whether there is a statistically significant<br>correlation between two vectors, you can use the <code>cor.test</code> function；</p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/22/SQLite.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/22/SQLite.html" class="post-title-link" itemprop="url">SQLite数据库简单操作</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-22 13:43:04" itemprop="dateCreated datePublished" datetime="2016-03-22T13:43:04+08:00">2016-03-22</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/22/SQLite.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/22/SQLite.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/22/SQLite.html" class="leancloud_visitors" data-flag-title="SQLite数据库简单操作">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="SQLite数据库整体查询命令">SQLite数据库整体查询命令</h2><h3 id="dot-commands">dot-commands</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$sqlite3</span> </span><br><span class="line">sqlite&gt; .<span class="built_in">help</span></span><br><span class="line">.backup ?DB? FILE      Backup DB (default <span class="string">"main"</span>) to FILE</span><br><span class="line">.bail ON|OFF           Stop after hitting an error.  Default OFF</span><br><span class="line">.databases             List names and files of attached databases</span><br><span class="line">.dump ?TABLE? ...      Dump the database <span class="keyword">in</span> an SQL text format</span><br><span class="line">                         If TABLE specified, only dump tables matching</span><br><span class="line">                         LIKE pattern TABLE.</span><br><span class="line">.<span class="built_in">echo</span> ON|OFF           Turn <span class="built_in">command</span> <span class="built_in">echo</span> on or off</span><br><span class="line">.<span class="built_in">exit</span>                  Exit this program</span><br><span class="line">.explain ON|OFF        Turn output mode suitable <span class="keyword">for</span> EXPLAIN on or off.</span><br><span class="line">.genfkey ?OPTIONS?     Options are:</span><br><span class="line">                         --no-drop: Do not drop old fkey triggers.</span><br><span class="line">                         --ignore-errors: Ignore tables with fkey errors</span><br><span class="line">                         --<span class="built_in">exec</span>: Execute generated SQL immediately</span><br><span class="line">                       See file tool/genfkey.README <span class="keyword">in</span> the <span class="built_in">source</span> </span><br><span class="line">                       distribution <span class="keyword">for</span> further information.</span><br><span class="line">.header(s) ON|OFF      Turn display of headers on or off</span><br><span class="line">.<span class="built_in">help</span>                  Show this message</span><br><span class="line">.import FILE TABLE     Import data from FILE into TABLE</span><br><span class="line">.indices ?TABLE?       Show names of all indices</span><br><span class="line">                         If TABLE specified, only show indices <span class="keyword">for</span> tables</span><br><span class="line">                         matching LIKE pattern TABLE.</span><br><span class="line">.load FILE ?ENTRY?     Load an extension library</span><br><span class="line">.mode MODE ?TABLE?     Set output mode <span class="built_in">where</span> MODE is one of:</span><br><span class="line">                         csv      Comma-separated values</span><br><span class="line">                         column   Left-aligned columns.  (See .width)</span><br><span class="line">                         html     HTML &lt;table&gt; code</span><br><span class="line">                         insert   SQL insert statements <span class="keyword">for</span> TABLE</span><br><span class="line">                         line     One value per line</span><br><span class="line">                         list     Values delimited by .separator string</span><br><span class="line">                         tabs     Tab-separated values</span><br><span class="line">                         tcl      TCL list elements</span><br><span class="line">.nullvalue STRING      Print STRING <span class="keyword">in</span> place of NULL values</span><br><span class="line">.output FILENAME       Send output to FILENAME</span><br><span class="line">.output stdout         Send output to the screen</span><br><span class="line">.prompt MAIN CONTINUE  Replace the standard prompts</span><br><span class="line">.quit                  Exit this program</span><br><span class="line">.<span class="built_in">read</span> FILENAME         Execute SQL <span class="keyword">in</span> FILENAME</span><br><span class="line">.restore ?DB? FILE     Restore content of DB (default <span class="string">"main"</span>) from FILE</span><br><span class="line">.schema ?TABLE?        Show the CREATE statements</span><br><span class="line">                         If TABLE specified, only show tables matching</span><br><span class="line">                         LIKE pattern TABLE.</span><br><span class="line">.separator STRING      Change separator used by output mode and .import</span><br><span class="line">.show                  Show the current values <span class="keyword">for</span> various settings</span><br><span class="line">.tables ?TABLE?        List names of tables</span><br><span class="line">                         If TABLE specified, only list tables matching</span><br><span class="line">                         LIKE pattern TABLE.</span><br><span class="line">.timeout MS            Try opening locked tables <span class="keyword">for</span> MS milliseconds</span><br><span class="line">.width NUM NUM ...     Set column widths <span class="keyword">for</span> <span class="string">"column"</span> mode</span><br><span class="line">.timer ON|OFF          Turn the CPU timer measurement on or off</span><br></pre></td></tr></table></figure>
<h2 id="SQLite数据类型">SQLite数据类型</h2><p><li>text</li></p>
<p><li>integer </li></p>
<p><li>real </li></p>
<p><li>NULL, used for missing data, or no value </li></p>
<p><li>BLOB, which stands for binary large object, and stores any type of object as bytes </li><br>注：SQLite并没有强制同列必须使用相同类型的数据，每个表的每一列都有优先类型（type affinity），但为了下游分析方便，最好同一列保持相同数据类型。<br>当某一列是混合数据类型时，排序原则为：NULL values, integer and real values (sorted numerically), text values, and finally blob values。</p>
<h2 id="数据库内容查询">数据库内容查询</h2><h3 id="万能的SELECT命令">万能的SELECT命令</h3><p>语法：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SELECT &lt;columns&gt; FROM &lt;tablename&gt;;</span><br></pre></td></tr></table></figure></p>
<p>基本形式：SELECT选择指令从一个table中抓取所有列的所有行（columns设定为<code>*</code>）。<br>选取特定列：不同列之间逗号分隔（SELECT trait, chrom, position, strongest_risk_snp, pvalue FROM gwascat LIMIT 5;）<br>SELECT语句除了在sqlite中交互查询外，还可在命令行中直接查询<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#交互</span></span><br><span class="line">sqlite&gt; SELECT * FROM gwascat;</span><br><span class="line"><span class="comment">#命令行</span></span><br><span class="line">sqlite3 gwascat.db <span class="string">"SELECT * FROM gwascat"</span> &gt; results.txt</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-commenting-o"></i>SQLite默认输出不规则，可做以下设置输出排列整齐易读输出：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT trait, chrom, position, strongest_risk_snp, pvalue</span><br><span class="line">   ...&gt; FROM gwascat LIMIT 5;</span><br><span class="line">trait|chrom|position|strongest_risk_snp|pvalue</span><br><span class="line">Asthma and hay fever|6|32658824|rs9273373|4.0e-14</span><br><span class="line">Asthma and hay fever|4|38798089|rs4833095|5.0e-12</span><br><span class="line">Asthma and hay fever|5|111131801|rs1438673|3.0e-11</span><br><span class="line">Asthma and hay fever|2|102350089|rs10197862|4.0e-11</span><br><span class="line">Asthma and hay fever|17|39966427|rs7212938|4.0e-10</span><br><span class="line">sqlite&gt; .header on</span><br><span class="line">sqlite&gt; .mode column</span><br><span class="line">sqlite&gt; SELECT trait, chrom, position, strongest_risk_snp, pvalue</span><br><span class="line">   ...&gt; FROM gwascat LIMIT 5;</span><br><span class="line">trait                 chrom       position    strongest_risk_snp  pvalue</span><br><span class="line">--------------------  ----------  ----------  ------------------  ----------</span><br><span class="line">Asthma and hay fever  6           32658824    rs9273373           4.0e-14</span><br><span class="line">Asthma and hay fever  4           38798089    rs4833095           5.0e-12</span><br><span class="line">Asthma and hay fever  5           111131801   rs1438673           3.0e-11</span><br><span class="line">Asthma and hay fever  2           102350089   rs10197862          4.0e-11</span><br><span class="line">Asthma and hay fever  17          39966427    rs7212938           4.0e-10</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-cog fa-spin"></i>SELECT可选参数：<br>LIMIT    输出查询行数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT * FROM gwascat LIMIT 2;</span><br></pre></td></tr></table></figure></p>
<p>ORDER BY     输出结果排序<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">SELECT author, trait, journal FROM &lt;tablename&gt; ORDER BY author DESC LIMIT 5; </span><br><span class="line"><span class="comment">#（按author降序排序），排序有助于异常值检测。</span></span><br><span class="line"><span class="comment">#若所指定排序列含由NULL值，可通过 IS NOT NULL 排除NULL值</span></span><br><span class="line">SELECT chrom, position, trait, strongest_risk_snp, pvalue FROM  &lt;tablename&gt; WHERE pvalue IS NOT NULL ORDER BY pvalue LIMIT 5;</span><br></pre></td></tr></table></figure></p>
<p>WHERE    数据筛选<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">SELECT chrom, position, trait, strongest_risk_snp, pvalue FROM &lt;tablename&gt; WHERE lower(strongest_risk_snp) = <span class="string">"rs429358"</span>;</span><br><span class="line"><span class="comment">#SQLite大小写敏感，所以匹配时最好用lower() 转换。</span></span><br></pre></td></tr></table></figure></p>
<p>多条件筛选：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT chrom, position, strongest_risk_snp, pvalue FROM gwascat</span><br><span class="line">   ...&gt; WHERE chrom IN (<span class="string">"1"</span>, <span class="string">"2"</span>, <span class="string">"3"</span>) AND pvalue &lt; 10e-11</span><br><span class="line">   ...&gt; ORDER BY pvalue LIMIT 5;</span><br><span class="line"><span class="comment">#或者</span></span><br><span class="line">sqlite&gt; SELECT chrom, position, strongest_risk_snp, pvalue</span><br><span class="line">   ...&gt; FROM gwascat WHERE chrom = <span class="string">"22"</span></span><br><span class="line">   ...&gt; AND position BETWEEN 24000000 AND 25000000</span><br><span class="line">   ...&gt; AND pvalue IS NOT NULL ORDER BY pvalue LIMIT 5;</span><br></pre></td></tr></table></figure></p>
<p>AS    对原始数据的修改：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT lower(trait) AS trait,</span><br><span class="line">   ...&gt; <span class="string">"chr"</span> || chrom || <span class="string">":"</span> || position AS region FROM gwascat LIMIT 5;</span><br><span class="line"><span class="comment">#||为连接运算符，用来连接两个字符串</span></span><br><span class="line"><span class="comment">#NULL的替换，ifnull()函数</span></span><br><span class="line">sqlite&gt; SELECT ifnull(chrom, <span class="string">"NA"</span>) AS chrom, ifnull(position, <span class="string">"NA"</span>) AS position,</span><br><span class="line">   ...&gt; strongest_risk_snp, ifnull(pvalue, <span class="string">"NA"</span>) AS pvalue FROM gwascat</span><br><span class="line">   ...&gt; WHERE strongest_risk_snp = <span class="string">"rs429358"</span>;</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-cutlery"></i>更多SQLite内置函数<br>Function    Description<br>ifnull(x, val)    If x is NULL, return with val, otherwise return x; shorthand for coalesce() with two arguments<br>min(a, b, c, …)    Return minimum in a, b, c, …<br>max(a, b, c, …)    Return maximum in a, b, c, …<br>abs(x)    Absolute value<br>coalesce(a, b, c, …)    Return first non-NULL value in a, b, c, … or NULL if all values are NULL<br>length(x)    Returns number of characters in x<br>lower(x)    Return x in lowercase<br>upper(x)    Return x in uppercase<br>replace(x, str, repl)    Return x with all occurrences of str replaced with repl<br>round(x, digits)    Round x to digits (default 0)<br>trim(x, chars), ltrim(x, chars), rtrim(x, chars)    Trim off chars (spaces if chars is not specified) from both sides, left side, and right side of x, respectively.<br>substr(x, start, length)    Extract a substring from x starting from character start and is length characters long</p>
<h3 id="集合函数（Aggregate）">集合函数（Aggregate）</h3><p>count(colname)函数：<br>返回总行数（无视NULL的存在）：sqlite&gt; SELECT count(*) FROM gwascat;<br>若colname是具体某列则返回出去NULL值的总行数.<br>其他相似函数：avg(x),max(x),min(x),sum(x),total(x)<br><i class="fa fa-modx"></i>计算列非重复值(unique)个数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT count(DISTINCT 列) AS unique_rs FROM gwascat;</span><br></pre></td></tr></table></figure></p>
<h3 id="行分组(GROUP_BY)">行分组(GROUP BY)</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; SELECT chrom, count(*) FROM gwascat GROUP BY chrom;</span><br><span class="line">chrom       count(*)</span><br><span class="line">----------  ----------</span><br><span class="line">            70</span><br><span class="line">1           1458</span><br><span class="line">10          930</span><br><span class="line">11          988</span><br><span class="line">12          858</span><br><span class="line">13          432</span><br><span class="line">[...]</span><br><span class="line"><span class="comment">#列重命名，降序，分组计算</span></span><br><span class="line">sqlite&gt; SELECT chrom, count(*) as nhits FROM gwascat GROUP BY chrom</span><br><span class="line">   ...&gt; ORDER BY nhits DESC;</span><br><span class="line">chrom       nhits</span><br><span class="line">----------  ----------</span><br><span class="line">6           1658</span><br><span class="line">1           1458</span><br><span class="line">2           1432</span><br><span class="line">3           1033</span><br><span class="line">11          988</span><br><span class="line">10          930</span><br><span class="line">[...]</span><br><span class="line"><span class="comment">#多列分组，不同列之间逗号分隔</span></span><br><span class="line">sqlite&gt; select strongest_risk_snp, strongest_risk_allele, count(*) AS count</span><br><span class="line">   ...&gt; FROM gwascat GROUP BY strongest_risk_snp, strongest_risk_allele</span><br><span class="line">   ...&gt; ORDER BY count DESC LIMIT 10;</span><br><span class="line">strongest_risk_snp  strongest_risk_allele  count</span><br><span class="line">------------------  ---------------------  ----------</span><br><span class="line">rs1260326           T                      22</span><br><span class="line">rs2186369           G                      22</span><br><span class="line">rs1800562           A                      20</span><br><span class="line">rs909674            C                      20</span><br><span class="line">rs11710456          G                      19</span><br></pre></td></tr></table></figure>
<h2 id="自己动手写数据库">自己动手写数据库</h2><p>we’ll use the basic SQL syntax to create tables and insert records into tables. Then load data into SQLite using Python’s sqlite3 module.</p>
<h3 id="创建tables">创建tables</h3><p>基本语法：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">CREATE TABLE tablename(</span><br><span class="line">  id <span class="built_in">integer</span> primary key,</span><br><span class="line">  column1 column1_type,</span><br><span class="line">  column2 column2_type,</span><br><span class="line">  ...</span><br><span class="line">);</span><br></pre></td></tr></table></figure></p>
<p>注意到所有SQLite数据库第一列总是id integer primary key，primary key是非重复整数来识别table中每一条记录。<br>创建table：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ sqlite3 practice.dbsqlite&gt; CREATE TABLE variants(</span><br><span class="line">   ...&gt;   id <span class="built_in">integer</span> primary key,</span><br><span class="line">   ...&gt;   chrom text,</span><br><span class="line">   ...&gt;   start <span class="built_in">integer</span>,</span><br><span class="line">   ...&gt;   end <span class="built_in">integer</span>,</span><br><span class="line">   ...&gt;   strand text,</span><br><span class="line">   ...&gt;   name text);</span><br></pre></td></tr></table></figure></p>
<h3 id="数据写入table">数据写入table</h3><p>基本语法：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">INSERT INTO tablename(column1, column2)</span><br><span class="line">VALUES (value1, value2);</span><br></pre></td></tr></table></figure></p>
<h3 id="建立索引">建立索引</h3><p>基本语法：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">sqlite&gt; CREATE INDEX &lt;columns-name_idx&gt; ON &lt;tablename&gt;(&lt;columns-name&gt;);</span><br><span class="line"><span class="comment">#察看索引</span></span><br><span class="line">sqlite&gt; .indices</span><br><span class="line">columns-name_idx</span><br><span class="line"><span class="comment">#删除索引</span></span><br><span class="line">sqlite&gt; DROP INDEX columns-name_idx;</span><br></pre></td></tr></table></figure></p>
<h3 id="修改/删除table">修改/删除table</h3><p>删除：DROP TABLE<br>修改：ALTER TABLE</p>
<h2 id="python中交互操作SQLite">python中交互操作SQLite</h2><h3 id="连接SQLite数据库并创建table">连接SQLite数据库并创建table</h3><p>create_table.py<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">import sqlite3</span><br><span class="line"></span><br><span class="line"><span class="comment"># the filename of this SQLite database</span></span><br><span class="line">db_filename = <span class="string">"variants.db"</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># initialize database connection</span></span><br><span class="line">conn = sqlite3.connect(db_filename) <span class="comment">#connect() 连接数据库</span></span><br><span class="line"></span><br><span class="line">c = conn.cursor() <span class="comment">#在python中用cursor()与SQLite数据库交互</span></span><br><span class="line"></span><br><span class="line">table_def = <span class="string">""</span><span class="string">"\ </span></span><br><span class="line"><span class="string">CREATE TABLE variants(</span></span><br><span class="line"><span class="string">  id integer primary key,</span></span><br><span class="line"><span class="string">  chrom test,</span></span><br><span class="line"><span class="string">  start integer,</span></span><br><span class="line"><span class="string">  end integer,</span></span><br><span class="line"><span class="string">  strand text,</span></span><br><span class="line"><span class="string">  rsid text);</span></span><br><span class="line"><span class="string">"</span><span class="string">""</span></span><br><span class="line"></span><br><span class="line">c.execute(table_def)  <span class="comment">#SQL语法，相当于确认</span></span><br><span class="line">conn.commit()  <span class="comment">#提交跟新内容到SQLite数据库</span></span><br><span class="line">conn.close()   <span class="comment">#关闭与数据库的连接</span></span><br></pre></td></tr></table></figure></p>
<h3 id="数据载入table">数据载入table</h3><p>load_variants.py<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##load_variants.py data.txt</span></span><br><span class="line">import sys</span><br><span class="line">import sqlite3</span><br><span class="line">from collections import OrderedDict</span><br><span class="line"></span><br><span class="line"><span class="comment"># the filename of this SQLite database</span></span><br><span class="line">db_filename = <span class="string">"variants.db"</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># initialize database connection</span></span><br><span class="line">conn = sqlite3.connect(db_filename)</span><br><span class="line">c = conn.cursor()</span><br><span class="line"></span><br><span class="line"><span class="comment">## Load Data</span></span><br><span class="line"><span class="comment"># columns (other than id, which is automatically incremented</span></span><br><span class="line">tbl_cols = OrderedDict([(<span class="string">"chrom"</span>, str), (<span class="string">"start"</span>, int), </span><br><span class="line">                        (<span class="string">"end"</span>, int), (<span class="string">"strand"</span>, str),</span><br><span class="line">                        (<span class="string">"rsid"</span>, str)])</span><br><span class="line"></span><br><span class="line">with open(sys.argv[1]) as input_file:</span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> input_file:</span><br><span class="line">        <span class="comment"># split a tab-delimited line</span></span><br><span class="line">        values = line.strip().split(<span class="string">"\t"</span>)</span><br><span class="line"></span><br><span class="line">        <span class="comment"># pair each value with its column name</span></span><br><span class="line">        cols_values = zip(tbl_cols.keys(), values)</span><br><span class="line">		<span class="comment"># use the column name to lookup an appropriate function to coerce each</span></span><br><span class="line">        <span class="comment"># value to the appropriate type</span></span><br><span class="line">        coerced_values = [tbl_cols[col](value) <span class="keyword">for</span> col, value <span class="keyword">in</span> cols_values]</span><br><span class="line"></span><br><span class="line">        <span class="comment"># create an empty list of placeholders</span></span><br><span class="line">        placeholders = [<span class="string">"?"</span>] * len(tbl_cols)</span><br><span class="line"></span><br><span class="line">        <span class="comment"># create the query by joining column names and placeholders quotation</span></span><br><span class="line">        <span class="comment"># marks into comma-separated strings</span></span><br><span class="line">        colnames = <span class="string">", "</span>.join(tbl_cols.keys())</span><br><span class="line">        placeholders = <span class="string">", "</span>.join(placeholders)</span><br><span class="line">        query = <span class="string">"INSERT INTO variants(%s) VALUES (%s);"</span>%(colnames, placeholders)</span><br><span class="line"></span><br><span class="line">        <span class="comment"># execute query</span></span><br><span class="line">        c.execute(query, coerced_values)</span><br><span class="line"></span><br><span class="line">conn.commit() <span class="comment"># commit these inserts</span></span><br><span class="line">conn.close()</span><br></pre></td></tr></table></figure></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/21/find-rm.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/21/find-rm.html" class="post-title-link" itemprop="url">Linux中如何正确删除：find-rm</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-21 12:18:46" itemprop="dateCreated datePublished" datetime="2016-03-21T12:18:46+08:00">2016-03-21</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/21/find-rm.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/21/find-rm.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/21/find-rm.html" class="leancloud_visitors" data-flag-title="Linux中如何正确删除：find-rm">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>假如当前目录下有诸多fastq文件想要删除，通常我们选择这样：</p>
<h2 id="直接删除">直接删除</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">rm *-temp.fastq</span><br></pre></td></tr></table></figure>
<p>但是如果一时疏忽输入<code>rm <em> -temp.fastq</em></code>（号和-号之间多了空格），那结果就惨了…..<br>而如果结合<a href="http://tiramisutes.github.io/2015/11/29/find/" target="_blank">Linux常用命令之find</a>命令那么一切就简单多了<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">find . -name <span class="string">"*-temp.fastq"</span> -<span class="built_in">exec</span> rm -i &#123;&#125; \;</span><br><span class="line"><span class="comment">#or</span></span><br><span class="line">find . -name <span class="string">"*-temp.fastq"</span> | xargs rm</span><br></pre></td></tr></table></figure></p>
<p>-exec 表示由find找到的匹配项会作为”-exec后面设定的命令”的参数（｛｝中输入值）<br>-i    交互删除<br>若想要删除文件夹，则<code>-delete</code> 代替<code>-exec -rm {}</code> </p>
<h2 id="打印删除rm命令并检查">打印删除rm命令并检查</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">find . -name <span class="string">"*-temp.fastq"</span> | xargs -n 1 <span class="built_in">echo</span> <span class="string">"rm -i"</span> &gt; delete-temp.sh</span><br><span class="line">cat delete-temp.sh</span><br><span class="line">rm -i ./zmaysA_R1-temp.fastq</span><br><span class="line">rm -i ./zmaysA_R2-temp.fastq</span><br><span class="line">rm -i ./zmaysC_R1-temp.fastq</span><br><span class="line">rm -i ./zmaysC_R2-temp.fastq</span><br><span class="line">bash delete-temp.sh</span><br><span class="line"><span class="comment">#or</span></span><br><span class="line">find . -name <span class="string">"*.fastq"</span> | xargs -n 1 -P 4 bash script.sh</span><br></pre></td></tr></table></figure>
<p>-n 1    表示find擦找到的参数每次只有一个输入到xargs中<br>-P    并行运算</p>
<h2 id="将删除文件放入临时文件夹（tmp）">将删除文件放入临时文件夹（tmp）</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">myrm</span></span>()&#123; D=/tmp/$(date +%Y%m%d%H%M%S); mkdir -p <span class="variable">$D</span>; mv <span class="string">"<span class="variable">$@</span>"</span> <span class="variable">$D</span> &amp;&amp; <span class="built_in">echo</span> <span class="string">"moved to <span class="variable">$D</span> ok"</span>; &#125;</span><br></pre></td></tr></table></figure>
          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/20/Samtools.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/20/Samtools.html" class="post-title-link" itemprop="url">被忽视的Samtools参数</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-20 22:54:42" itemprop="dateCreated datePublished" datetime="2016-03-20T22:54:42+08:00">2016-03-20</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/20/Samtools.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/20/Samtools.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/20/Samtools.html" class="leancloud_visitors" data-flag-title="被忽视的Samtools参数">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p><span class="exturl" data-url="aHR0cDovL3d3dy5wbG9iLm9yZy90YWcvc2FtdG9vbHM=" title="View all posts in samtools">Samtools<i class="fa fa-external-link"></i></span>是一个用于操作序列比对结果sam和bam文件的工具合集。</p>
<h2 id="sam文件格式">sam文件格式</h2><p>SAM格式由两部分组成：头部区和比对区，都以tab分列。<br><strong>头部区:</strong>以’@’开始，体现了比对的一些总体信息。比对的SAM格式版本，比对的参考序列，比对使用的软件等。<br><strong>比对区:</strong> 比对结果，每一个比对结果是一行，有11个主列和1个可选列。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">@HD VN:1.0 SO:unsorted  </span><br><span class="line">头部区第一行：VN是格式版本；SO表示比对排序的类型，有unkown(default),unsorted,</span><br><span class="line">queryname和coordinate几种。samtools软件在进行排序后不能自动更新bam文件的SO值。</span><br><span class="line">picard却可以。</span><br><span class="line">@SQ SN:A.auricula_all_contig_1 LN:9401</span><br><span class="line">参考序列名。这些参考序列决定了比对结果sort的顺序。SN是参考序列名；LN是参考序列</span><br><span class="line">长度;</span><br><span class="line">@RG ID:sample01</span><br><span class="line">Read Group. 1个sample的测序结果为1个Read Group；该sample可以有多个library</span><br><span class="line">的测序结果。</span><br><span class="line">@PG ID:bowtie2 PN:bowtie2 VN:2.0.0-beta7</span><br><span class="line">比对所使用的软件。</span><br><span class="line"></span><br><span class="line">比对区11个列和可选列的解释</span><br><span class="line">1  QNAME  比对的序列名,即单端或双端fa/fq 中的reads编号, ‘*’ indicates the information is unavailable.</span><br><span class="line">2  FLAG   Bwise FLAG(表明比对类型：pairing，strand，mate strand等)</span><br><span class="line">3  RNAME  比对上的参考序列名,An unmapped segment without coordinate has a ‘*’ at this field. </span><br><span class="line">4  POS    1-Based的比对上的最左边的定位,POS is <span class="built_in">set</span> as 0 <span class="keyword">for</span> an unmapped <span class="built_in">read</span> without coordinate. If POS is 0, no assumptions can be made about RNAME and CIGAR.</span><br><span class="line">5  MAPQ   比对质量，255表示没有map</span><br><span class="line">6  CIGAR  Extended CIGAR string (操作符：MIDNSHP) 比对结果信息：匹配碱基数，可变剪接等，*表示不可用。</span><br><span class="line">7  RNEXT   相匹配的另外一条序列，比对上的参考序列名,‘*’ when the information is unavailable, and <span class="built_in">set</span> as ‘=’ <span class="keyword">if</span> RNEXT is identical RNAME</span><br><span class="line">8  PNEXT   1-Based leftmost Mate POsition,0 when the information is unavailable.</span><br><span class="line">9  TLEN  插入片段长度,<span class="built_in">set</span> as 0 <span class="keyword">for</span> single-segment template or when the information is unavailable</span><br><span class="line">10 SEQ    和参考序列在同一个琏上的比对序列(若比对结果在负意链上，则序列是其反向重复序列), ‘*’ when the sequence is not stored</span><br><span class="line">11 QUAL   比对序列的质量(ASCII-33=Phred base quality),‘*’ when quality is not stored</span><br><span class="line">12 可选的行，以TAG：TYPE：VALUE的形式提供额外的信息</span><br></pre></td></tr></table></figure></p>
<h3 id="比对区解释">比对区解释</h3><p><strong>sam/bam比对区包含有此次比对的结果信息，</strong>其中主要信息解释如下：</p>
<p><li><strong>FLAG部分</strong></li><br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/flag.png" alt=""><br><strong>0x800</strong> 表明相应位置的比对属于嵌合体比对；<br><strong>0x4</strong> 没有map上的reads；<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/sam_output2.png" alt=""></p>
<p><li><strong>CIGAR部分</strong></li><br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/cigar.png" alt=""><br>对于mRNA到基因组的比对，N表示内含子。<br>More: <span class="exturl" data-url="aHR0cDovL3NhbXRvb2xzLmdpdGh1Yi5pby9odHMtc3BlY3MvU0FNdjEucGRm" title="http://samtools.github.io/hts-specs/SAMv1.pdf">http://samtools.github.io/hts-specs/SAMv1.pdf<i class="fa fa-external-link"></i></span></p>
<h2 id="sam文件的几个特例解释">sam文件的几个特例解释</h2><h3 id="Unmapped_reads_统计">Unmapped reads 统计</h3><p>Each alignment is one line of the SAM file, but not all lines are successful alignments. <strong>Unmapped reads在sam文件中的标记：<code>FLAG</code>列为4而且<code>RNAME</code>列为星号*；</strong><br><figure class="highlight nimrod"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#统计包含星号的比对行数</span></span><br><span class="line">cut -f3 smallRNA-<span class="built_in">seq</span>.sam | grep -c \*</span><br><span class="line"><span class="comment">#总的比对行数</span></span><br><span class="line">grep -c -v <span class="string">"^@"</span> smallRNA-<span class="built_in">seq</span>.sam</span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_different_read_IDs_are_in_the_file?">How many different read IDs are in the file?</h3><p>The query (read) ID is in field 1. Some reads may have multiple alignments, so the number of lines is not necessarily the number of reads.<br><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"@"</span> HR<span class="number">-1</span>B.fq.gz.sam | <span class="type">cut</span> -f1 | <span class="type">sort</span> | <span class="type">uniq</span> | <span class="type">wc</span> -l </span><br><span class="line">#同时统计此次比对的单端/双端 fa/fq文件发现两者结果不同，说明确实有些reads未能比对上去。</span><br><span class="line">zcat pepper/RNA-seq/sgs-clean-reads/HR<span class="number">-1</span>B.fq.gz | <span class="type">wc</span> -l</span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_different_read_sequences_are_in_the_file?">How many different read sequences are in the file?</h3><p>第一列的reads ID仅能表示测序过程中的不同reads，但他们的序列可能因为PCR扩增原因或文库偏好而完全相同，所以统计第10列的uniq序列数能够准确的表示reads总数。<br><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cut</span> -f10 smallRNAseq.sam | <span class="type">sort</span> | <span class="type">uniq</span> | <span class="type">wc</span> -l</span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_reads_are_uniquely_mapped?">How many reads are <strong>uniquely mapped</strong>?</h3><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cut</span> -f10 smallRNAseq.sam | <span class="type">sort</span> | <span class="type">uniq</span> -u | <span class="type">wc</span> -l</span><br></pre></td></tr></table></figure>
<p>对于 BWA比对结果也可用<code>grep -c XT:A:U smallRNA-seq.sam</code>来准确统计。</p>
<h3 id="How_many_reads_are_multi-hits?">How many reads are <strong>multi-hits</strong>?</h3><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cut</span> -f1 smallRNA-seq.sam | <span class="type">sort</span> | <span class="type">uniq</span> -d | <span class="type">wc</span> -l</span><br></pre></td></tr></table></figure>
<p>对于 BWA比对结果也可用<code>grep -c XT:A:R smallRNAseq.sam</code>来准确统计。</p>
<h3 id="How_many_alignments_are_reported_for_each_read?">How many alignments are reported for each read?</h3><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"^@"</span> smallRNA-seq.sam | <span class="type">cut</span> -f1 | <span class="type">sort</span> | <span class="type">uniq</span> -c | <span class="type">sort</span> -nr &gt; sortedreadcount.txt</span><br><span class="line">grep -v <span class="string">"^@"</span> smallRNA-seq.sam | <span class="type">cut</span> -f1 |<span class="type">sort</span> | <span class="type">uniq</span> -c | <span class="type">sort</span> -nr | <span class="type">cut</span> -c1<span class="number">-8</span> | <span class="type">sort</span> | <span class="type">uniq</span> -c</span><br></pre></td></tr></table></figure>
<h3 id="How_many_different_reference_sequences_are_represented_in_the_file?">How many different reference sequences are represented in the file?</h3><figure class="highlight coq"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"^@"</span> smallRNA-seq.sam | <span class="type">cut</span> –f3 | <span class="type">sort</span> | <span class="type">uniq</span> | <span class="type">wc</span> -l</span><br></pre></td></tr></table></figure>
<h2 id="view">view</h2><p>-c    计数<br>-f    返回指定区间/flags比对结果<br>-q    返回比对质量大于等于指定值的比对数目<br><strong>-F 4：</strong>统计map 上的 reads总数；<br><strong>-f 4：</strong>统计没有map 上的 reads总数；<br>To get the <strong>unmapped</strong> reads from a <span class="exturl" data-url="aHR0cDovL3NhbXRvb2xzLnNvdXJjZWZvcmdlLm5ldC9TQU0xLnBkZg==" title="http://samtools.sourceforge.net/SAM1.pdf">bam<i class="fa fa-external-link"></i></span> file use :<br><code>samtools view -f 4 file.bam &gt; unmapped.sam</code>, the output will be in <span class="exturl" data-url="aHR0cDovL3NhbXRvb2xzLnNvdXJjZWZvcmdlLm5ldC9TQU0xLnBkZg==" title="http://samtools.sourceforge.net/SAM1.pdf">sam<i class="fa fa-external-link"></i></span><br>to get the output in <span class="exturl" data-url="aHR0cDovL3NhbXRvb2xzLnNvdXJjZWZvcmdlLm5ldC9TQU0xLnBkZg==" title="http://samtools.sourceforge.net/SAM1.pdf">bam<i class="fa fa-external-link"></i></span> use : <code>samtools view -b -f 4 file.bam &gt; unmapped.bam</code><br>To get only the <strong>mapped</strong> reads use the parameter ‘F’, which works like <code>-v</code> of grep and skips the alignments for a specific flag.<br><code>samtools view -b -F 4 file.bam &gt; mapped.bam</code><br>samtools view -b -F 4 -f 8 file.bam &gt; onlyThisEndMapped.bam<br>samtools view -b -F 8 -f 4 file.bam &gt; onlyThatEndMapped.bam<br>samtools view -b -F12 file.bam &gt; bothEndsMapped.bam<br>samtools merge merged.bam onlyThisEndMapped.bam onlyThatEndMapped.bam bothEndsMapped.bam<br><strong>对于tophat比对结果：</strong><br><code><span style="background-color: rgb(249, 242, 244);">samtools view -b -f 2  accepted_hits.bam &gt; mappedPairs.bam</span></code><br>Better with:<br><code><span style="background-color: rgb(249, 242, 244);">samtools view -b -f 0x2 accepted_hits.bam &gt; mappedPairs.bam</span></code></p>
<h2 id="sort">sort</h2><p>-m    指定运算内存，支持K，M，G等缩写<br>-@    并行运算核数</p>
<h2 id="index">index</h2><p>必须对bam文件进行默认情况下的排序后，才能进行index。否则会报错。</p>
<p>建立索引后将产生后缀为.bai的文件，用于快速的随机处理。很多情况下需要有bai文件的存在，特别是显示序列比对情况下。比如samtool的tview命令就需要；gbrowse2显示reads的比对图形的时候也需要。</p>
<h2 id="faidx">faidx</h2><p>对fasta文件建立索引,生成的索引文件以.fai后缀结尾。该命令也能依据索引文件快速提取fasta文件中的某一条（子）序列。<br>See more： <a href="http://tiramisutes.github.io/2016/03/18/bedtools.html" target="_blank">bedtools 使用小结</a></p>
<h2 id="flagstat">flagstat</h2><p>给出BAM文件的比对结果<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flagstat example.bam</span><br><span class="line">11945742 + 0 <span class="keyword">in</span> total (QC-passed reads + QC-failed reads)</span><br><span class="line"><span class="comment">#总共的reads数</span></span><br><span class="line">0 + 0 duplicates</span><br><span class="line">7536364 + 0 mapped (63.09%:-nan%)</span><br><span class="line"><span class="comment">#总体上reads的匹配率</span></span><br><span class="line">11945742 + 0 paired <span class="keyword">in</span> sequencing</span><br><span class="line"><span class="comment">#有多少reads是属于paired reads</span></span><br><span class="line">5972871 + 0 read1</span><br><span class="line"><span class="comment">#reads1中的reads数</span></span><br><span class="line">5972871 + 0 read2</span><br><span class="line"><span class="comment">#reads2中的reads数</span></span><br><span class="line">6412042 + 0 properly paired (53.68%:-nan%)</span><br><span class="line"><span class="comment">#完美匹配的reads数：比对到同一条参考序列，并且两条reads之间的距离符合设置的阈值</span></span><br><span class="line">6899708 + 0 with itself and mate mapped</span><br><span class="line"><span class="comment">#paired reads中两条都比对到参考序列上的reads数</span></span><br><span class="line">636656 + 0 singletons (5.33%:-nan%)</span><br><span class="line"><span class="comment">#单独一条匹配到参考序列上的reads数，和上一个相加，则是总的匹配上的reads数。</span></span><br><span class="line">469868 + 0 with mate mapped to a different chr</span><br><span class="line"><span class="comment">#paired reads中两条分别比对到两条不同的参考序列的reads数</span></span><br><span class="line">243047 + 0 with mate mapped to a different chr (mapQ&gt;=5)</span><br></pre></td></tr></table></figure></p>
<h2 id="mpileup">mpileup</h2><p>samtools还有个非常重要的命令mpileup，以前为pileup。该命令用于生成bcf文件，再使用bcftools进行SNP和Indel的分析。bcftools是samtool中附带的软件，在samtools的安装文件夹中可以找到。<br>-f 来输入有索引文件的fasta参考序列；<br>-g 输出到bcf格式。</p>
<h2 id="depth">depth</h2><p>计算每一个位点或者区域的测序深度；<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools depth sorted<span class="selector-class">.bam</span> &gt; sorted<span class="selector-class">.bam</span><span class="selector-class">.txt</span></span><br></pre></td></tr></table></figure></p>
<p>一共得到3列以指标分隔符分隔的数据，第一列为染色体名称，第二列为位点，第三列为覆盖深度。</p>
<p></p><p></p><br>贡献来源<br><span class="exturl" data-url="aHR0cDovL3d3dy5wbG9iLm9yZy8yMDE0LzAxLzI2LzcxMTIuaHRtbA==" title="http://www.plob.org/2014/01/26/7112.html">http://www.plob.org/2014/01/26/7112.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL2Jsb2cuc2luYS5jb20uY24vcy9ibG9nXzY3MDQ0NTI0MDEwMWwzMGsuaHRtbA==" title="http://blog.sina.com.cn/s/blog_670445240101l30k.html">http://blog.sina.com.cn/s/blog_670445240101l30k.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly93d3cuYmlvc3RhcnMub3JnL3AvNTYyNDYv" title="https://www.biostars.org/p/56246/">https://www.biostars.org/p/56246/<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly93d3cuYmlvc3RhcnMub3JnL3AvMTEwMDM5Lw==" title="https://www.biostars.org/p/110039/">https://www.biostars.org/p/110039/<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly93d3cuYmlvc3RhcnMub3JnL3AvOTU5Mjkv" title="https://www.biostars.org/p/95929/">https://www.biostars.org/p/95929/<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly93d3cuYmlvc3RhcnMub3JnL3AvMTEwMTU3Lw==" title="https://www.biostars.org/p/110157/">https://www.biostars.org/p/110157/<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly9ncm91cHMuZ29vZ2xlLmNvbS9mb3J1bS8jIWZvcnVtL2JlZHRvb2xzLWRpc2N1c3M=" title="https://groups.google.com/forum/#!forum/bedtools-discuss">https://groups.google.com/forum/#!forum/bedtools-discuss<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly9jb2RlLmdvb2dsZS5jb20vcC9oeWRyYS1zdi8=" title="https://code.google.com/p/hydra-sv/">https://code.google.com/p/hydra-sv/<i class="fa fa-external-link"></i></span><p></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/18/bedtools.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/18/bedtools.html" class="post-title-link" itemprop="url">bedtools 使用小结</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-18 16:29:32" itemprop="dateCreated datePublished" datetime="2016-03-18T16:29:32+08:00">2016-03-18</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/18/bedtools.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/18/bedtools.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/18/bedtools.html" class="leancloud_visitors" data-flag-title="bedtools 使用小结">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="概述">概述</h2><p>BEDTools是可用于genomic features的比较，相关操作及进行注释的工具。而genomic features通常使用Browser Extensible Data (BED) 或者 General Feature Format (GFF)文件表示，用UCSC Genome Browser进行可视化比较。</p>
<h2 id="与BEDTools使用相关的基本概念">与BEDTools使用相关的基本概念</h2><p>已有的一些genome features信息一般由BED格式或者GFF格式进行存储。<br>genome features: 功能元素（gene）， 遗传多态性 (SNPs, INDELs, or structural variants), 已经由测序或者其他方法得到的注释信息，也可以是自定义的一些特征信息。<br>Overlapping/intersecting features: 两个genome features的区域至少有一个bp的共同片段。</p>
<h2 id="BED和GFF文件的一个差异">BED和GFF文件的一个差异</h2><p><strong>BED文件中起始坐标为0，结束坐标至少是1,； GFF中起始坐标是1而结束坐标至少是1。</strong></p>
<h2 id="相关格式">相关格式</h2><h3 id="BED_format">BED format</h3><p>BEDTools主要使用BED格式的前三列，BED可以最多有12列。BED格式的常用列描述如下：<br>chrom: 染色体信息， 如chr1, III, myCHrom, contig1112.23, 必须有<br>start: genome feature的起始位点，从0开始， 必须有<br>end: genome feature的终止位点，至少为1， 必须有<br>score: 可以是p值等等一些可以刻量化的数值信息<br>strands: 正反链信息</p>
<h3 id="GFF_format">GFF format</h3><p><strong>seqname</strong> - name of the chromosome or scaffold; chromosome names can be given with or without the ‘chr’ prefix. <b>Important note</b>: the seqname must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.<br><strong>source</strong> - name of the program that generated this feature, or the data source (database or project name)<br><strong>feature</strong> - feature type name, e.g. Gene, Variation, Similarity<br><strong>start</strong> - Start position of the feature, with sequence numbering starting at 1.end - End position of the feature, with sequence numbering starting at 1.<br><strong>score</strong> - A floating point value.strand - defined as + (forward) or - (reverse).<br><strong>frame</strong> - One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on..<br><strong>attribute</strong> - A semicolon-separated list of tag-value pairs, providing additional information about each feature.<br>See more from <i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5lbnNlbWJsLm9yZy9pbmZvL3dlYnNpdGUvdXBsb2FkL2dmZi5odG1s" title="http://www.ensembl.org/info/website/upload/gff.html">http://www.ensembl.org/info/website/upload/gff.html<i class="fa fa-external-link"></i></span></p>
<h3 id="genome_files">genome files</h3><p>BEDTools中的一些工具（genomeCoverageBed, complementBed, slopBed）需要物种的染色体大小的信息，genome file一般就是每行都是tab隔开，两列，一列为染色体的名字，第二列为这个染色体的大小。<strong>一般常用物种的genome file在BEDTools安装目录的/genome里面</strong>。<br>自定义基因组genome files文件生成方法见我的另一篇博文：<strong><a href="http://tiramisutes.github.io/2015/11/08/fa-length.html" target="_blank">批量求fasta格式序列长度</a>。</strong></p>
<h2 id="BEDTools使用总结">BEDTools使用总结</h2><h3 id="intersect/intersectBed：计算_Overlaps">intersect/intersectBed：计算 Overlaps</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect -a A.bed -b B.bed -wa -wb</span><br></pre></td></tr></table></figure>
<p>用来求两个BED或者BAM文件中的overlap，overlap可以进行自定义是整个genome features的overlap还是局部。<br>默认的结果描述如下图：<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/bed1.jpg" alt=""><br>加-wa参数可以报告出原始的在A文件中的feature， 如下图<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/bed2.jpg" alt=""><br>加-wb参数可以报告出原始的在B文件中的feature, 加-c参数可以报告出两个文件中的overlap的feature的数量。<br><i class="fa fa-bolt"></i>当用bedtools intersect 处理大文件时比较耗内存，有效的方法是对A和B文件按照染色体名字(chromosome)和位置(position)排序(<code>sort -k1,1 -k2,2n</code>),然后用<code>-sorted</code>参数重新intersect。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect -a A-sorted.bed -b B-sorted.bed --sorted</span><br></pre></td></tr></table></figure></p>
<p>其他参数：<br>-wo 返回overlap碱基数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$bedtools</span>  intersect -a A.bed -b B.bed -wo</span><br><span class="line">chr1    0       15      a       chr1    0       4       x       4</span><br><span class="line">chr1    0       15      a       chr1    9       15      z       6</span><br><span class="line">chr1    25      29      b       chr1    18      28      y       3</span><br><span class="line">chr1    18      18      c       chr1    18      28      y       1</span><br><span class="line">chr1    10      14      d       chr1    9       15      z       4</span><br><span class="line">chr1    20      23      e       chr1    18      28      y       3</span><br></pre></td></tr></table></figure></p>
<p>-v 返回非overlap区间<br>-s 相同链上的feature<br>-c 两个文件中的overlap的feature的数量</p>
<h3 id="complement：返回基因组非覆盖区">complement：返回基因组非覆盖区</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools complement -i &lt;BED/GFF/VCF&gt; -g &lt;genome files&gt;</span><br></pre></td></tr></table></figure>
<h3 id="Slop：增加特征区间大小">Slop：增加特征区间大小</h3><p>要求：单个输入bed文件（-i指定）和genome files<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">cat ranges-qry.bed </span><br><span class="line">chr1    0       15      a</span><br><span class="line">chr1    25      29      b</span><br><span class="line">chr1    18      18      c</span><br><span class="line">chr1    10      14      d</span><br><span class="line">chr1    20      23      e</span><br><span class="line">chr1    6       7       f</span><br><span class="line">bedtools slop -i ranges-qry.bed -g genome.txt -b 4</span><br><span class="line">chr1    0       19      a</span><br><span class="line">chr1    21      33      b</span><br><span class="line">chr1    14      22      c</span><br><span class="line">chr1    6       18      d</span><br><span class="line">chr1    16      27      e</span><br><span class="line">chr1    2       11      f</span><br><span class="line"><span class="comment">#-b 4	:两端同时缩短4个碱基</span></span><br></pre></td></tr></table></figure></p>
<p>-l 3 -r 5:增加左3右5</p>
<h3 id="flank：提取特定区域(启动子区)">flank：提取特定区域(启动子区)</h3><p>要求：基因组GTF文件（-i指定）和genome files<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">bedtools flank -i mm_GRCm38.75_protein_coding_genes.gtf \</span><br><span class="line">                 -g Mus_musculus.GRCm38_genome.txt \</span><br><span class="line">                 -l 3000 -r 0 &gt; mm_GRCm38_3kb_promoters.gtf</span><br><span class="line">cut -f1,4,5,7 mm_GRCm38_3kb_promoters.gtf | head -n 3</span><br><span class="line">1       3671499 3674498 -</span><br><span class="line">1       4360315 4363314 -</span><br><span class="line">1       4496414 4499413 -</span><br></pre></td></tr></table></figure></p>
<h3 id="getfasta：提取序列">getfasta：提取序列</h3><p>要求：基因组fasta文件（-fi指定）和提取区间GTF文件(-bed指定)<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bedtools getfasta -<span class="keyword">fi</span> Mus_musculus.GRCm38.75.dna_rm.toplevel_chr1.fa \</span><br><span class="line">   -bed mm_GRCm38_3kb_promoters.gtf -fo mm_GRCm38_3kb_promoters.fasta</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-bell"></i>-tab    Report extract sequences in a tab-delimited format instead of in FASTA format.</p>
<p></p><p></p><br><i class="fa fa-commenting-o"></i>提取序列之<code>samtools</code>（速度较快<i class="fa fa-plane"></i>）<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#首先建立fai索引文件（第一列为染色体名字，第二列为序列碱基数）</span></span><br><span class="line">samtools faidx Mus_musculus.GRCm38.75.dna.chromosome.8.fa</span><br><span class="line"><span class="comment">#序列提取，多提取区间空格隔开</span></span><br><span class="line">samtools faidx Mus_musculus.GRCm38.75.dna.chromosome.8.fa \</span><br><span class="line">     8:123407082-123410744 8:123518835-123536649</span><br><span class="line">&gt;8:123407082-123410744</span><br><span class="line">GAGAAAAGCTCCCTTCTTCTCCAGAGTCCCGTCTACCCTGGCTTGGCGAGGGAAAGGAAC</span><br><span class="line">CAGACATATATCAGAGGCAAGTAACCAAGAAGTCTGGAGGTGTTGAGTTTAGGCATGTCT</span><br><span class="line">[...]</span><br><span class="line">&gt;8:123518835-123536649</span><br><span class="line">TCTCGCGAGGATTTGAGAACCAGCACGGGATCTAGTCGGAGTTGCCAGGAGACCGCGCAG</span><br><span class="line">CCTCCTCTGACCAGCGCCCATCCCGGATTAGTGGAAGTGCTGGACTGCTGGCACCATGGT</span><br><span class="line">[...]</span><br></pre></td></tr></table></figure><p></p>
<h3 id="nuc:_计算GC含量即各碱基数">nuc: 计算GC含量即各碱基数</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools nuc -<span class="keyword">fi</span> hg19.fa -bed CDS.bed</span><br></pre></td></tr></table></figure>
<p>输出结果解释：在原bed文件每行结尾增加以下几列<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">Output format: </span><br><span class="line">The following information will be reported after each BED entry:</span><br><span class="line">    1) %AT content</span><br><span class="line">    2) %GC content</span><br><span class="line">    3) Number of As observed</span><br><span class="line">    4) Number of Cs observed</span><br><span class="line">    5) Number of Gs observed</span><br><span class="line">    6) Number of Ts observed</span><br><span class="line">    7) Number of Ns observed</span><br><span class="line">    8) Number of other bases observed</span><br><span class="line">    9) The length of the explored sequence/interval.</span><br><span class="line">    10) The seq. extracted from the FASTA file. (opt., <span class="keyword">if</span> -seq is used)</span><br><span class="line">    11) The number of <span class="built_in">times</span> a user<span class="string">'s pattern was observed.</span></span><br><span class="line"><span class="string">        (opt., if -pattern is used.)</span></span><br></pre></td></tr></table></figure></p>
<h3 id="genomecov：染色体和全基因组覆盖度计算">genomecov：染色体和全基因组覆盖度计算</h3><p>要求：单个输入bed文件（-i指定）和genome files；如果输入为bam(-ibam指定)文件，则不需要genome files。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">cat ranges-cov-sorted.bed</span><br><span class="line">chr1    4       9</span><br><span class="line">chr1    1       6</span><br><span class="line">chr1    8       19</span><br><span class="line">chr1    25      30</span><br><span class="line">chr2    0       20</span><br><span class="line">$ cat cov.txt</span><br><span class="line">chr1    30</span><br><span class="line">chr2    20</span><br><span class="line">bedtools genomecov -i ranges-cov-sorted.bed -g cov.txt</span><br><span class="line">chr1    0       7       30      0.233333 1</span><br><span class="line">chr1    1       20      30      0.666667</span><br><span class="line">chr1    2       3       30      0.1</span><br><span class="line">chr2    1       20      20      1 2</span><br><span class="line">genome  0       7       50      0.14 3</span><br><span class="line">genome  1       40      50      0.8</span><br><span class="line">genome  2       3       50      0.06</span><br><span class="line"><span class="comment">#name 覆盖次数 覆盖碱基数 总碱基数   覆盖度</span></span><br><span class="line"><span class="comment">#同时计算单染色体和全基因组覆盖度</span></span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-cogs"></i><li>ranges-cov.bed文件需提前排序sort -k1,1 ranges-cov.bed &gt; ranges-cov-sorted.bed</li></p>
<p><li>-bg参数可得到每个碱基的覆盖度。</li></p>
<h3 id="coverage：计算染色体给定区间覆盖度">coverage：计算染色体给定区间覆盖度</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">$ cat A.bed</span><br><span class="line">chr1  0   100</span><br><span class="line">chr1  100 200</span><br><span class="line">chr2  0   100</span><br><span class="line"></span><br><span class="line">$ cat B.bed</span><br><span class="line">chr1  10  20</span><br><span class="line">chr1  20  30</span><br><span class="line">chr1  30  40</span><br><span class="line">chr1  100 200</span><br><span class="line"></span><br><span class="line">$ bedtools coverage -a A.bed -b B.bed</span><br><span class="line">chr1  0   100  3  30  100 0.3000000</span><br><span class="line">chr1  100 200  1  100 100 1.0000000</span><br><span class="line">chr2  0   100  0  0   100 0.0000000</span><br></pre></td></tr></table></figure>
<p></p><p></p><br>贡献来源<br><span class="exturl" data-url="aHR0cDovL3d3dy5wbG9iLm9yZy8yMDEyLzA5LzI2LzM3NDguaHRtbA==" title="http://www.plob.org/2012/09/26/3748.html">http://www.plob.org/2012/09/26/3748.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL2JlZHRvb2xzLnJlYWR0aGVkb2NzLm9yZy9lbi9sYXRlc3QvY29udGVudC9iZWR0b29scy1zdWl0ZS5odG1s" title="http://bedtools.readthedocs.org/en/latest/content/bedtools-suite.html">http://bedtools.readthedocs.org/en/latest/content/bedtools-suite.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly9jb2RlLmdvb2dsZS5jb20vYXJjaGl2ZS9wL2JlZHRvb2xzL3dpa2lzL1VzYWdlLndpa2k=" title="https://code.google.com/archive/p/bedtools/wikis/Usage.wiki">https://code.google.com/archive/p/bedtools/wikis/Usage.wiki<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly9jb2RlLmdvb2dsZS5jb20vYXJjaGl2ZS9wL2JlZHRvb2xzL3dpa2lzL1VzYWdlQWR2YW5jZWQud2lraQ==" title="https://code.google.com/archive/p/bedtools/wikis/UsageAdvanced.wiki">https://code.google.com/archive/p/bedtools/wikis/UsageAdvanced.wiki<i class="fa fa-external-link"></i></span><p></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/15/dplyr.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/15/dplyr.html" class="post-title-link" itemprop="url">R数据整形术之 dplyr</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-15 21:10:33" itemprop="dateCreated datePublished" datetime="2016-03-15T21:10:33+08:00">2016-03-15</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/15/dplyr.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/15/dplyr.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/15/dplyr.html" class="leancloud_visitors" data-flag-title="R数据整形术之 dplyr">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="数据集类型">数据集类型</h2><p>将过长过大的数据集转换为显示更友好的 tbl_df 类型:<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">hflights_df &lt;- tbl_df(hflights)</span><br></pre></td></tr></table></figure></p>
<p>可以 hflights_df 感受一下不再被刷屏的感觉.</p>
<h2 id="基本操作">基本操作</h2><p><i class="fa fa-book"></i>常用的数据操作行为归纳为以下五种:</p>
<h3 id="筛选:_filter()">筛选: filter()</h3><p>按给定的逻辑判断筛选出符合要求的子数据集, 类似于 base::subset() 函数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">filter(hflights_df, Month == 1, DayofMonth == 1)</span><br><span class="line"><span class="comment">#对同一对象的任意个条件组合</span></span><br><span class="line">filter(hflights_df, Month == 1 | Month == 2)</span><br></pre></td></tr></table></figure></p>
<h3 id="排列:_arrange()">排列: arrange()</h3><p>按给定的列名依次对行进行排序<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">arrange(hflights_df, DayofMonth, Month, Year)</span><br><span class="line"><span class="comment">#对列名加 desc() 进行倒序:</span></span><br><span class="line">arrange(hflights_df, desc(ArrDelay))</span><br><span class="line"><span class="comment">#这个函数和 plyr::arrange() 是一样的, 类似于 order()</span></span><br></pre></td></tr></table></figure></p>
<h3 id="选择:_select()">选择: select()</h3><p>用列名作参数来选择子数据集:<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">select(hflights_df, Year, Month, DayOfWeek)</span><br><span class="line"><span class="comment">#还可以用 : 来连接列名, 没错, 就是把列名当作数字一样使用</span></span><br><span class="line">select(hflights_df, Year:DayOfWeek)</span><br><span class="line">用 - 来排除列名:</span><br><span class="line">select(hflights_df, -DrayOfWeek)</span><br></pre></td></tr></table></figure></p>
<h3 id="变形:_mutate()">变形: mutate()</h3><p>对已有列进行数据运算并添加为新列:<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mutate(hflights_df,   gain = ArrDelay - DepDelay,   speed = Distance / AirTime * 60)</span><br></pre></td></tr></table></figure></p>
<h3 id="汇总:_summarise()">汇总: summarise()</h3><p>对数据框调用其它函数进行汇总操作, 返回一维的结果:<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">summarise(hflights_df, delay = mean(DepDelay, na.rm = TRUE))</span><br></pre></td></tr></table></figure></p>
<h2 id="分组动作_group_by()">分组动作 group_by()</h2><p>以上5个动词函数已经很方便了, 但是当它们跟分组操作这个概念结合起来时, 那才叫真正的强大! 当对数据集通过 group_by() 添加了分组信息后,mutate(), arrange() 和 summarise() 函数会自动对这些 tbl 类数据执行分组操作 (R语言泛型函数的优势).<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">mtfs_df %&gt;%</span><br><span class="line">     group_by(chr) %&gt;%</span><br><span class="line">     summarize(max_recom = max(recom), mean_recom = mean(recom), num=n())</span><br><span class="line"></span><br><span class="line">Source: <span class="built_in">local</span> data frame [23 x 4]</span><br><span class="line">     chr max_recom mean_recom  num</span><br><span class="line">1   chr1   41.5648   2.217759 2095</span><br><span class="line">2  chr10   42.4129   2.162635 1029</span><br><span class="line">3  chr11   36.1703   2.774918  560</span><br><span class="line">[...]</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-book"></i>summarise中用到的函数<br><i class="fa fa-fort-awesome"></i><br>n(): 计算个数 n_distinct(): 计算 x 中唯一值的个数. (原文为 count_distinct(x))<br>first(x), last(x) 和 nth(x, n): 返回对应秩的值, 类似于自带函数 x[1], x[length(x)], 和 x[n]</p>
<h2 id="连接符_%&gt;%">连接符 %&gt;%</h2><p>使用时把数据名作为开头, 然后依次对此数据进行多步操作.</p>
<h2 id="join功能">join功能</h2><p>两数据集取交，并集等。</p>
<ul>
<li>inner_join(x,y)  交集</li>
<li>semi_join(x,y) </li>
<li>left_join(x,y)</li>
<li>anti_join(x,y)</li>
<li>inner_join(y,x)</li>
<li>semi_join(y,x)</li>
<li>left_join(y,x)</li>
<li>anti_join(y,x)</li>
<li>full_join(x,y)  并集<br>【<span class="exturl" data-url="aHR0cDovL3N0YXQ1NDUuY29tL2JpdDAwMV9kcGx5ci1jaGVhdHNoZWV0Lmh0bWw=" title="http://stat545.com/bit001_dplyr-cheatsheet.html">Cheatsheet for dplyr join functions<i class="fa fa-external-link"></i></span>】<h2 id="深入学习">深入学习</h2>dplyr 包自带的60页<span class="exturl" data-url="aHR0cDovL2NyYW4ucnN0dWRpby5jb20vd2ViL3BhY2thZ2VzL2RwbHlyL2RwbHlyLnBkZg==" title="http://cran.rstudio.com/web/packages/dplyr/dplyr.pdf">详细文档<i class="fa fa-external-link"></i></span>.<br>其余几个vignettes (<span class="exturl" data-url="aHR0cDovL2NyYW4ucnN0dWRpby5jb20vd2ViL3BhY2thZ2VzL2RwbHlyL3ZpZ25ldHRlcy8=" title="http://cran.rstudio.com/web/packages/dplyr/vignettes/">网页<i class="fa fa-external-link"></i></span>) 或 vignette(package = “dplyr”),包含了数据库相关, 混合编程, 运算性能比较, 以及新的 window-functions 等内容.<br>简单看了下<span class="exturl" data-url="aHR0cDovL2NyYW4ucnN0dWRpby5jb20vd2ViL3BhY2thZ2VzL2RwbHlyL3ZpZ25ldHRlcy93aW5kb3ctZnVuY3Rpb25zLmh0bWw=" title="http://cran.rstudio.com/web/packages/dplyr/vignettes/window-functions.html">vignette(“window-functions”, package = “dplyr”)<i class="fa fa-external-link"></i></span>, 提供了一系列函数, 扩展了原来只能返回一个数值的聚焦类函数(如sum(), mean())至返回等长度的值, 变成 cumsum()和 cummean(), 以及 n(), lead() 和 lag()等便捷功能.<br>plyr 包的相关文档: <span class="exturl" data-url="aHR0cDovL3BseXIuaGFkLmNvLm56Lw==" title="http://plyr.had.co.nz/">主页<i class="fa fa-external-link"></i></span><br>还有<span class="exturl" data-url="aHR0cDovL2NyYW4ucnN0dWRpby5jb20vd2ViL3BhY2thZ2VzL2RhdGEudGFibGUv" title="http://cran.rstudio.com/web/packages/data.table/">data.table<i class="fa fa-external-link"></i></span>包也是很强大的哦, 空下来可以学一学.</li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/09/tidyr.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/09/tidyr.html" class="post-title-link" itemprop="url">R数据整形术之 tidyr</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-09 21:07:46" itemprop="dateCreated datePublished" datetime="2016-03-09T21:07:46+08:00">2016-03-09</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/09/tidyr.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/09/tidyr.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/09/tidyr.html" class="leancloud_visitors" data-flag-title="R数据整形术之 tidyr">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p><blockquote></blockquote></p>
<p></p><p>Happy families are all alike; every unhappy family is unhappy in its own way — Leo Tolstoy</p><br><br>R数据整形包之一tidyr最近迎来更新(<span class="exturl" data-url="aHR0cDovL2Jsb2cucnN0dWRpby5vcmcvMjAxNi8wMi8wMi90aWR5ci0wLTQtMC8=" title="http://blog.rstudio.org/2016/02/02/tidyr-0-4-0/">tidyr 0.4.0<i class="fa fa-external-link"></i></span>)，所以有必要对其<span class="exturl" data-url="aHR0cHM6Ly9jcmFuLnItcHJvamVjdC5vcmcvd2ViL3BhY2thZ2VzL3RpZHlyL3ZpZ25ldHRlcy90aWR5LWRhdGEuaHRtbA==" title="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html">Tidy data<i class="fa fa-external-link"></i></span>进行学习。<br>以下为个人简单总结：<p></p>
<pre class="r"><code>#gather--mutate--separate--select--arrange
setwd(&quot;F:/Rwork/tidyr&quot;)
library(tidyr)</code></pre>
<pre><code>## Warning: package 'tidyr' was built under R version 3.2.3</code></pre>
<pre class="r"><code>library(dplyr)</code></pre>
<pre><code>## Warning: package 'dplyr' was built under R version 3.2.2</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'
## 
## The following objects are masked from 'package:stats':
## 
##     filter, lag
## 
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>preg &lt;-read.csv(&quot;preg.csv&quot;,stringsAsFactors = FALSE)
preg</code></pre>
<pre><code>##           name treatmenta treatmentb
## 1   John Smith         NA         18
## 2     Jane Doe          4          1
## 3 Mary Johnson          6          7</code></pre>
<pre class="r"><code>preg6 &lt;-tbl_df(read.csv(&quot;preg.csv&quot;,stringsAsFactors = FALSE))
preg6</code></pre>
<pre><code>## Source: local data frame [3 x 3]
## 
##           name treatmenta treatmentb
##          (chr)      (int)      (int)
## 1   John Smith         NA         18
## 2     Jane Doe          4          1
## 3 Mary Johnson          6          7</code></pre>
<pre class="r"><code>preg2&lt;-preg %&gt;% 
  gather(treatment,n,treatmenta:treatmentb) %&gt;%  
#The first argument, is the name of the key column, which is the name of the variable defined by the values of the column headings. 
#The second argument is the name of the value column.
#The third argument defines the columns to gather, here, every column except religion.
 #gather(treatment,n,-name,na.rm = TRUE)    #the same above # na.rm to drop any missing values from the gather columns
  mutate(treatment=gsub(&quot;treatment&quot;,&quot;&quot;,treatment)) %&gt;%
  arrange(name,treatment)   #arrange=sort
preg2</code></pre>
<pre><code>##           name treatment  n
## 1     Jane Doe         a  4
## 2     Jane Doe         b  1
## 3   John Smith         a NA
## 4   John Smith         b 18
## 5 Mary Johnson         a  6
## 6 Mary Johnson         b  7</code></pre>
<pre class="r"><code>#each deplay one by one 
preg3&lt;-preg %&gt;% 
  gather(treatment,n,treatmenta:treatmentb)
preg3</code></pre>
<pre><code>##           name  treatment  n
## 1   John Smith treatmenta NA
## 2     Jane Doe treatmenta  4
## 3 Mary Johnson treatmenta  6
## 4   John Smith treatmentb 18
## 5     Jane Doe treatmentb  1
## 6 Mary Johnson treatmentb  7</code></pre>
<pre class="r"><code>preg33&lt;-preg3 %&gt;% separate(treatment, c(&quot;Treatments&quot;, &quot;group&quot;),9 )  #separate
preg33</code></pre>
<pre><code>##           name Treatments group  n
## 1   John Smith  treatment     a NA
## 2     Jane Doe  treatment     a  4
## 3 Mary Johnson  treatment     a  6
## 4   John Smith  treatment     b 18
## 5     Jane Doe  treatment     b  1
## 6 Mary Johnson  treatment     b  7</code></pre>
<pre class="r"><code>preg333&lt;-preg33 %&gt;% select(name,group,n)  # select
preg333</code></pre>
<pre><code>##           name group  n
## 1   John Smith     a NA
## 2     Jane Doe     a  4
## 3 Mary Johnson     a  6
## 4   John Smith     b 18
## 5     Jane Doe     b  1
## 6 Mary Johnson     b  7</code></pre>
<pre class="r"><code>preg3333&lt;-preg333 %&gt;% spread(group,n)  #spread: one column become two column
preg3333</code></pre>
<pre><code>##           name  a  b
## 1     Jane Doe  4  1
## 2   John Smith NA 18
## 3 Mary Johnson  6  7</code></pre>
<pre class="r"><code>preg4&lt;-preg3 %&gt;% mutate(treatment=gsub(&quot;treatment&quot;,&quot;&quot;,treatment))
preg4</code></pre>
<pre><code>##           name treatment  n
## 1   John Smith         a NA
## 2     Jane Doe         a  4
## 3 Mary Johnson         a  6
## 4   John Smith         b 18
## 5     Jane Doe         b  1
## 6 Mary Johnson         b  7</code></pre>
<pre class="r"><code>preg5&lt;-preg4 %&gt;% arrange(name,treatment)
preg5</code></pre>
<pre><code>##           name treatment  n
## 1     Jane Doe         a  4
## 2     Jane Doe         b  1
## 3   John Smith         a NA
## 4   John Smith         b 18
## 5 Mary Johnson         a  6
## 6 Mary Johnson         b  7</code></pre>
<pre class="r"><code>#reads all files from the same locaed pathway  into a single data frame.
library(plyr)</code></pre>
<pre><code>## -------------------------------------------------------------------------
## You have loaded plyr after dplyr - this is likely to cause problems.
## If you need functions from both plyr and dplyr, please load plyr first, then dplyr:
## library(plyr); library(dplyr)
## -------------------------------------------------------------------------
## 
## Attaching package: 'plyr'
## 
## The following objects are masked from 'package:dplyr':
## 
##     arrange, count, desc, failwith, id, mutate, rename, summarise,
##     summarize</code></pre>
<pre class="r"><code>paths &lt;- dir(&quot;F:/Rwork/tidyr&quot;, pattern = &quot;\\.csv$&quot;, full.names = TRUE)
names(paths) &lt;- basename(paths)
all&lt;-ldply(paths, read.csv, stringsAsFactors = FALSE)
all</code></pre>
<pre><code>##               .id         name treatmenta treatmentb
## 1 preg - Copy.csv   John Smith         NA         18
## 2 preg - Copy.csv     Jane Doe          4          1
## 3 preg - Copy.csv Mary Johnson          6          7
## 4        preg.csv   John Smith         NA         18
## 5        preg.csv     Jane Doe          4          1
## 6        preg.csv Mary Johnson          6          7</code></pre>
<pre class="r"><code>#get some data from name column incloud John Smith and Jane Doe  located in preg3 data frame 
subset(preg3, name %in% c(&quot;John Smith&quot;, &quot;Jane Doe&quot;))</code></pre>
<pre><code>##         name  treatment  n
## 1 John Smith treatmenta NA
## 2   Jane Doe treatmenta  4
## 4 John Smith treatmentb 18
## 5   Jane Doe treatmentb  1</code></pre>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/09/Economist-Graph.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/09/Economist-Graph.html" class="post-title-link" itemprop="url">Economist_Graph：复杂图形修改</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-09 17:23:10" itemprop="dateCreated datePublished" datetime="2016-03-09T17:23:10+08:00">2016-03-09</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/09/Economist-Graph.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/09/Economist-Graph.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/09/Economist-Graph.html" class="leancloud_visitors" data-flag-title="Economist_Graph：复杂图形修改">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p></p><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/20111210_WOC210.gif" width="600" height="300"></p><br>原图：<span class="exturl" data-url="aHR0cDovL3d3dy5lY29ub21pc3QuY29tL25vZGUvMjE1NDExNzg=" title="http://www.economist.com/node/21541178">http://www.economist.com/node/21541178<i class="fa fa-external-link"></i></span><p></p>
<h2 id="Basic_plot">Basic plot</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">dat &lt;- read.csv(<span class="string">"EconomistData.csv"</span>)</span><br><span class="line">pc1 &lt;- ggplot(dat, aes(x = CPI, y = HDI, color = Region))+</span><br><span class="line">       geom_point()</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter1.png" alt=""></p>
<h2 id="Trend_line">Trend line</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pc2 &lt;- pc1 +</span><br><span class="line">   geom_smooth(aes(group = 1),</span><br><span class="line">               method = <span class="string">"lm"</span>,</span><br><span class="line">               formula = y ~ <span class="built_in">log</span>(x),</span><br><span class="line">               se = FALSE,</span><br><span class="line">               color = <span class="string">"red"</span>) +</span><br><span class="line">   geom_line()</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter2.png" alt=""></p>
<h2 id="Open_points">Open points</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pc3 &lt;- pc2 +</span><br><span class="line">  geom_point(shape = 1, size = 4)</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter3.png" alt=""></p>
<h2 id="Labelling_points">Labelling points</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">pointsToLabel &lt;- c(<span class="string">"Russia"</span>, <span class="string">"Venezuela"</span>, <span class="string">"Iraq"</span>, <span class="string">"Myanmar"</span>, <span class="string">"Sudan"</span>,</span><br><span class="line">                   <span class="string">"Afghanistan"</span>, <span class="string">"Congo"</span>, <span class="string">"Greece"</span>, <span class="string">"Argentina"</span>, <span class="string">"Brazil"</span>,</span><br><span class="line">                   <span class="string">"India"</span>, <span class="string">"Italy"</span>, <span class="string">"China"</span>, <span class="string">"South Africa"</span>, <span class="string">"Spane"</span>,</span><br><span class="line">                   <span class="string">"Botswana"</span>, <span class="string">"Cape Verde"</span>, <span class="string">"Bhutan"</span>, <span class="string">"Rwanda"</span>, <span class="string">"France"</span>,</span><br><span class="line">                   <span class="string">"United States"</span>, <span class="string">"Germany"</span>, <span class="string">"Britain"</span>, <span class="string">"Barbados"</span>, <span class="string">"Norway"</span>, <span class="string">"Japan"</span>,</span><br><span class="line">                   <span class="string">"New Zealand"</span>, <span class="string">"Singapore"</span>)</span><br><span class="line">library(<span class="string">"ggrepel"</span>)</span><br><span class="line">pc4 &lt;- pc3 +  geom_text_repel(aes(label = Country),</span><br><span class="line">            color = <span class="string">"gray20"</span>,</span><br><span class="line">            data = subset(dat, Country %<span class="keyword">in</span>% pointsToLabel),</span><br><span class="line">            force = 10)</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter4.png" alt=""><br>选择性的标注想要的点</p>
<h2 id="Change_the_region_labels_and_order">Change the region labels and order</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">dat<span class="variable">$Region</span> &lt;- factor(dat<span class="variable">$Region</span>,</span><br><span class="line">                     levels = c(<span class="string">"EU W. Europe"</span>,</span><br><span class="line">                                <span class="string">"Americas"</span>,</span><br><span class="line">                                <span class="string">"Asia Pacific"</span>,</span><br><span class="line">                                <span class="string">"East EU Cemt Asia"</span>,</span><br><span class="line">                                <span class="string">"MENA"</span>,</span><br><span class="line">                                <span class="string">"SSA"</span>),</span><br><span class="line">                     labels = c(<span class="string">"OECD"</span>,</span><br><span class="line">                                <span class="string">"Americas"</span>,</span><br><span class="line">                                <span class="string">"Asia &amp;\nOceania"</span>,</span><br><span class="line">                                <span class="string">"Central &amp;\nEastern Europe"</span>,</span><br><span class="line">                                <span class="string">"Middle East &amp;\nnorth Africa"</span>,</span><br><span class="line">                                <span class="string">"Sub-Saharan\nAfrica"</span>))</span><br><span class="line">pc4<span class="variable">$data</span> &lt;- dat</span><br><span class="line">pc4</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter6.png" alt=""><br>修改图例值和顺序</p>
<h2 id="Add_title_and_format_axes">Add title and format axes</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">library(grid)</span><br><span class="line">pc5 &lt;- pc4 +</span><br><span class="line">  scale_x_continuous(name = <span class="string">"Corruption Perceptions Index, 2011 (10=least corrupt)"</span>,</span><br><span class="line">                     limits = c(.9, 10.5),</span><br><span class="line">                     breaks = 1:10) +</span><br><span class="line">  scale_y_continuous(name = <span class="string">"Human Development Index, 2011 (1=Best)"</span>,</span><br><span class="line">                     limits = c(0.2, 1.0),</span><br><span class="line">                     breaks = seq(0.2, 1.0, by = 0.1)) +</span><br><span class="line">  scale_color_manual(name = <span class="string">""</span>,</span><br><span class="line">                     values = c(<span class="string">"#24576D"</span>,</span><br><span class="line">                                <span class="string">"#099DD7"</span>,</span><br><span class="line">                                <span class="string">"#28AADC"</span>,</span><br><span class="line">                                <span class="string">"#248E84"</span>,</span><br><span class="line">                                <span class="string">"#F2583F"</span>,</span><br><span class="line">                                <span class="string">"#96503F"</span>)) +</span><br><span class="line">  ggtitle(<span class="string">"Corruption and Human development"</span>))</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter7.png" alt=""><br>利用scale来修改x，y轴，颜色和标出title</p>
<h2 id="Theme_tweaks">Theme tweaks</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">library(grid) <span class="comment"># for the 'unit' function</span></span><br><span class="line">pc6 &lt;- pc5 +</span><br><span class="line">  theme_minimal() + <span class="comment"># start with a minimal theme and add what we need</span></span><br><span class="line">  theme(text = element_text(color = <span class="string">"gray20"</span>),</span><br><span class="line">        legend.position = c(<span class="string">"top"</span>), <span class="comment"># position the legend in the upper left </span></span><br><span class="line">        legend.direction = <span class="string">"horizontal"</span>,</span><br><span class="line">        legend.justification = 0.1, <span class="comment"># anchor point for legend.position.</span></span><br><span class="line">        legend.text = element_text(size = 11, color = <span class="string">"gray10"</span>),</span><br><span class="line">        axis.text = element_text(face = <span class="string">"italic"</span>),</span><br><span class="line">        axis.title.x = element_text(vjust = -1), <span class="comment"># move title away from axis</span></span><br><span class="line">        axis.title.y = element_text(vjust = 2), <span class="comment"># move away for axis</span></span><br><span class="line">        axis.ticks.y = element_blank(), <span class="comment"># element_blank() is how we remove elements</span></span><br><span class="line">        axis.line = element_line(color = <span class="string">"gray40"</span>, size = 0.5),</span><br><span class="line">        axis.line.y = element_blank(),</span><br><span class="line">        panel.grid.major = element_line(color = <span class="string">"gray50"</span>, size = 0.5),</span><br><span class="line">        panel.grid.major.x = element_blank()</span><br><span class="line">        ))</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter8.png" alt=""><br>微调主题</p>
<h2 id="Add_model_R^2_and_source_note">Add model R^2 and source note</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">mR2 &lt;- summary(lm(HDI ~ <span class="built_in">log</span>(CPI), data = dat))<span class="variable">$r</span>.squared</span><br><span class="line">library(grid)</span><br><span class="line">png(file = <span class="string">"images/econScatter10.png"</span>, width = 800, height = 600)</span><br><span class="line">pc6 </span><br><span class="line">grid.text(<span class="string">"Sources: Transparency International; UN Human Development Report"</span>,</span><br><span class="line">         x = .02, y = .03,</span><br><span class="line">         just = <span class="string">"left"</span>,</span><br><span class="line">         draw = TRUE)</span><br><span class="line">grid.segments(x0 = 0.81, x1 = 0.825,</span><br><span class="line">              y0 = 0.90, y1 = 0.90,</span><br><span class="line">              gp = gpar(col = <span class="string">"red"</span>),</span><br><span class="line">              draw = TRUE)</span><br><span class="line">grid.text(paste0(<span class="string">"R² = "</span>,</span><br><span class="line">                 as.integer(mR2*100),</span><br><span class="line">                 <span class="string">"%"</span>),</span><br><span class="line">          x = 0.835, y = 0.90,</span><br><span class="line">          gp = gpar(col = <span class="string">"gray20"</span>),</span><br><span class="line">          draw = TRUE,</span><br><span class="line">          just = <span class="string">"left"</span>)</span><br><span class="line"></span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/econScatter10.png" alt=""><br>Contribution from ：<br><span class="exturl" data-url="aHR0cDovL3R1dG9yaWFscy5pcS5oYXJ2YXJkLmVkdS9SL1JncmFwaGljcy9SZ3JhcGhpY3MuaHRtbA==" title="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html">http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/09/ggplot2-learning.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/03/09/ggplot2-learning.html" class="post-title-link" itemprop="url">myself ggplot2 learning tip (再学ggplot2： 遗漏细节)</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-03-09 13:44:37" itemprop="dateCreated datePublished" datetime="2016-03-09T13:44:37+08:00">2016-03-09</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/03/09/ggplot2-learning.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/03/09/ggplot2-learning.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/03/09/ggplot2-learning.html" class="leancloud_visitors" data-flag-title="myself ggplot2 learning tip (再学ggplot2： 遗漏细节)">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p></p><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/20111210_WOC210.gif" width="500" height="250"></p><br>原图：<span class="exturl" data-url="aHR0cDovL3d3dy5lY29ub21pc3QuY29tL25vZGUvMjE1NDExNzg=" title="http://www.economist.com/node/21541178">http://www.economist.com/node/21541178<i class="fa fa-external-link"></i></span><p></p>
<h2 id="ggplot2绘图语法结构：">ggplot2绘图语法结构：</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">ggplot(data = &lt;default data <span class="built_in">set</span>&gt;, </span><br><span class="line">       aes(x = &lt;default x axis variable&gt;,</span><br><span class="line">           y = &lt;default y axis variable&gt;,</span><br><span class="line">           ... &lt;other default aesthetic mappings&gt;),</span><br><span class="line">       ... &lt;other plot defaults&gt;) +</span><br><span class="line"></span><br><span class="line">       geom_&lt;geom <span class="built_in">type</span>&gt;(aes(size = &lt;size variable <span class="keyword">for</span> this geom&gt;, </span><br><span class="line">                      ... &lt;other aesthetic mappings&gt;),</span><br><span class="line">                  data = &lt;data <span class="keyword">for</span> this point geom&gt;,</span><br><span class="line">                  <span class="built_in">stat</span> = &lt;statistic string or <span class="keyword">function</span>&gt;,</span><br><span class="line">                  position = &lt;position string or <span class="keyword">function</span>&gt;,</span><br><span class="line">                  color = &lt;<span class="string">"fixed color specification"</span>&gt;,</span><br><span class="line">                  &lt;other arguments, possibly passed to the _stat_ <span class="keyword">function</span>) +</span><br><span class="line"></span><br><span class="line">  scale_&lt;aesthetic&gt;_&lt;<span class="built_in">type</span>&gt;(name = &lt;<span class="string">"scale label"</span>&gt;,</span><br><span class="line">                     breaks = &lt;<span class="built_in">where</span> to put tick marks&gt;,</span><br><span class="line">                     labels = &lt;labels <span class="keyword">for</span> tick marks&gt;,</span><br><span class="line">                     ... &lt;other options <span class="keyword">for</span> the scale&gt;) +</span><br><span class="line"></span><br><span class="line">  theme(plot.background = element_rect(fill = <span class="string">"gray"</span>),</span><br><span class="line">        ... &lt;other theme elements&gt;)</span><br></pre></td></tr></table></figure>
<h2 id="Geometric_Objects(geom_XX)_and_Aesthetics_(aes())">Geometric Objects(geom_XX) and Aesthetics (aes())</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">geom_&lt;tab&gt;查看所有Geometric Objects（几何学对象）</span><br></pre></td></tr></table></figure>
<h2 id="Statistical_Transformations（统计变换）">Statistical Transformations（统计变换）</h2><p>统计变换 （stat_） 比如求均值，求方差等，当我们需要展示出某个变量的某种统计特征的时候，需要用到统计变换。<br>每一个<code>geom_XX</code>都有一个默认的统计量，可通过<code>args(geom_XX)</code>查看（<code>args(stat_bin)</code>）例如<code>geom_bar</code>的默认统计量是<code>stat_count</code>,表示进行计数。什么意思呢？举例如下：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; library(ggplot2)</span><br><span class="line">&gt; df &lt;- data.frame(trt = c(<span class="string">"a"</span>,<span class="string">"a"</span>, <span class="string">"b"</span>, <span class="string">"c"</span>), outcome = c(2.3,5, 1.9, 3.2))</span><br><span class="line">&gt; df</span><br><span class="line">  trt outcome</span><br><span class="line">1   a     2.3</span><br><span class="line">2   a     5.0</span><br><span class="line">3   b     1.9</span><br><span class="line">4   c     3.2</span><br><span class="line">&gt; <span class="comment">##stat()</span></span><br><span class="line">&gt; ggplot(df, aes(trt, outcome)) +</span><br><span class="line">  geom_bar()</span><br><span class="line">Error: stat_count() must not be used with a y aesthetic.</span><br><span class="line"><span class="comment">#成功报错，因为第一句aes()中我们指定了y轴为outcome，即一个x轴的trt值对应一个y轴的outcome值，而geom_bar()</span></span><br><span class="line"><span class="comment">#中stat_count要进行计数，即计数trt中相同值出现的次数.最终y轴指定的outcome与需要表示的count值冲突，报错。</span></span><br><span class="line"><span class="comment">##解决：</span></span><br><span class="line">&gt; ggplot(df, aes(trt)) +</span><br><span class="line">  geom_bar()</span><br></pre></td></tr></table></figure></p>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/count.png" alt=""><br>如果想要表示相同trt值相加后的对应值，修改stat=indentity即可。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; ggplot(df, aes(trt, outcome)) +</span><br><span class="line">+   geom_bar(<span class="built_in">stat</span> = <span class="string">"identity"</span>)</span><br></pre></td></tr></table></figure></p>
<p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/indentity.png" alt=""><br>ggplot2中包含的统计变换有如下多种：<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/stat.png" alt=""><br>更多参见：<span class="exturl" data-url="aHR0cHM6Ly93d3cuemhpaHUuY29tL3F1ZXN0aW9uLzI0Nzc5MDE3" title="https://www.zhihu.com/question/24779017">https://www.zhihu.com/question/24779017<i class="fa fa-external-link"></i></span></p>
<h2 id="Scales">Scales</h2><h3 id="控制aes()映射">控制<code>aes()</code>映射</h3><p>Aesthetic (<code>aes()</code>) 映射仅仅是告诉一个变量应该映射到一个aesthetic，但并没有说明如何映射？例如，当我们用<code>aes(shape=x)</code>去映射一个变量到shape时，并没有说明用什么shape；同样的，<code>aes(color=z)</code>并无说明用什么颜色；通常在我们未定义这些时ggplot2会自动用默认值，而我们可以通过<em>scale</em>来修改这些值。<br>在ggplot2中scales包括：</p>
<li>position</li><br><li>color and fill</li><br><li>size</li><br><li>shape</li><br><li>line type</li><br><li>x,y</li><br>### 修改方式<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">scale_&lt;aesthetic&gt;_&lt;<span class="built_in">type</span>&gt;(name,limits,breaks,labels)</span><br><span class="line">通过键入scale_&lt;tab&gt;查看全部可修改函数。</span><br></pre></td></tr></table></figure><br><br>特殊scale函数有额外参数，例如对颜色的修改<code>scale_color_continuous</code>函数有<code>low</code>和<code>high</code>参数。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">scale_x_discrete(name=<span class="string">"State Abbreviation"</span>) +</span><br><span class="line">scale_color_continuous(name=<span class="string">""</span>,</span><br><span class="line">                         breaks = c(19751, 19941, 20131),</span><br><span class="line">                         labels = c(1971, 1994, 2013),</span><br><span class="line">                         low = <span class="string">"blue"</span>, high = <span class="string">"red"</span>)</span><br></pre></td></tr></table></figure><br><br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/modifyScales3.png" alt=""><br>## Faceting<br>### 分面: 一页多图<br><code>facet_wrap()</code>：对数据分类只能应用一个标准，例<code>facet_wrap(~State, ncol = 10))</code>,按State分组后每行设置10个小图依次画出全部。<br><code>facet_grid()</code>：多个标准对数据进行分组绘图,<code>facet_grid(color~cut，margins=TRUE)</code>，波浪号前为小图分行标准，后面为分列标准，margins指用于分面的包含每个变量元素所有数据的数据组，相当于每个小图一个title。<br>## Themes<br>### 主题<br>更多ggplot2主题演示如下：<br><span class="exturl" data-url="aHR0cDovL2RvY3MuZ2dwbG90Mi5vcmcvZGV2L3ZpZ25ldHRlcy90aGVtZXMuaHRtbA==" title="http://docs.ggplot2.org/dev/vignettes/themes.html">http://docs.ggplot2.org/dev/vignettes/themes.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cHM6Ly9naXRodWIuY29tL2pybm9sZC9nZ3RoZW1lcw==" title="https://github.com/jrnold/ggthemes">https://github.com/jrnold/ggthemes<i class="fa fa-external-link"></i></span><br>### 修改主题默认值<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">theme_minimal() +</span><br><span class="line">  theme(text = element_text(color = <span class="string">"turquoise"</span>))</span><br></pre></td></tr></table></figure><br><br>### 自定义主题<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">theme_new &lt;- theme_bw() +</span><br><span class="line">  theme(plot.background = element_rect(size = 1, color = <span class="string">"blue"</span>, fill = <span class="string">"black"</span>),</span><br><span class="line">        text=element_text(size = 12, family = <span class="string">"Serif"</span>, color = <span class="string">"ivory"</span>),</span><br><span class="line">        axis.text.y = element_text(colour = <span class="string">"purple"</span>),</span><br><span class="line">        axis.text.x = element_text(colour = <span class="string">"red"</span>),</span><br><span class="line">        panel.background = element_rect(fill = <span class="string">"pink"</span>),</span><br><span class="line">        strip.background = element_rect(fill = muted(<span class="string">"orange"</span>)))</span><br><span class="line"></span><br><span class="line">p5 + theme_new</span><br></pre></td></tr></table></figure><br><br>## 关于晕晕的aes()<br><li><strong>任何与数据向量顺序相关，需要逐个指定的参数都必须写在aes里</strong></li><br><li>什么？还是搞不清该放aes里面还是外面？那就记着想统一整个图层时就放到aes外，想分成不同组调整，并且已经有一个与x、y长度一致的分组变量了，那就放到aes里</li><br>##其他总结<br><li>加注释，所有注释的实现都是通过annotate函数实现的，geom_text()是兼职的。</li><br><li>theme函数最妙的地方是将对于数据相关的美学调整和与数据无关的美学调整分离，将数据处理与数据美学分开，数据美学与数据无关的调整分开。</li>



          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/01/29/ncRNA.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2016/01/29/ncRNA.html" class="post-title-link" itemprop="url">Noncoding RNAs (ncRNAs)</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2016-01-29 10:29:16" itemprop="dateCreated datePublished" datetime="2016-01-29T10:29:16+08:00">2016-01-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/molecular-biology/" itemprop="url" rel="index"><span itemprop="name">molecular biology</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2016/01/29/ncRNA.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2016/01/29/ncRNA.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2016/01/29/ncRNA.html" class="leancloud_visitors" data-flag-title="Noncoding RNAs (ncRNAs)">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Noncoding_RNAs_(ncRNAs)">Noncoding RNAs (ncRNAs)</h2><p><li>most ncRNAs operate as RNA-protein complexes, including ribosomes, snRNPs, snoRNPs, telomerase, microRNAs, and long ncRNAs.</li></p>
<p><li></li></p>
<h2 id="Non-coding_RNA_database">Non-coding RNA database</h2><p><span class="exturl" data-url="aHR0cDovL3Jlc2VhcmNoLmltYi51cS5lZHUuYXUvcm5hZGIv" title="http://research.imb.uq.edu.au/rnadb/">http://research.imb.uq.edu.au/rnadb/<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/7/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/7/">7</a><span class="page-number current">8</span><a class="page-number" href="/page/9/">9</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/9/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      

      <div class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <a href="/" class="site-author-image" rel="start" style="border:none">
              <img class="site-author-image" itemprop="image"
                src="/images/author.jpg"
                alt="tiramisutes"/>
              </a>
            
              <p class="site-author-name" itemprop="name">tiramisutes</p>
              <div class="site-description motion-element" itemprop="description">hope bioinformatics blog</div>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">148</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  
                    
                      <a href="/categories/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">17</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                      <a href="/tags/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">112</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly9naXRodWIuY29tL3RpcmFtaXN1dGVz" title="GitHub &rarr; https://github.com/tiramisutes"><i class="fa fa-fw fa-github"></i>GitHub</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="bWFpbHRvOnRpcmFtaXN1dGVze2F0fTEyNi5jb20=" title="E-Mail &rarr; mailto:tiramisutes{at}126.com"><i class="fa fa-fw fa-envelope"></i>E-Mail</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cDovL3dlaWJvLmNvbS90aXJhbWlzdXRlc1Rlcw==" title="Weibo &rarr; http://weibo.com/tiramisutesTes"><i class="fa fa-fw fa-globe"></i>Weibo</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly90d2l0dGVyLmNvbS9ob3BldG9neQ==" title="Twitter &rarr; https://twitter.com/hopetogy"><i class="fa fa-fw fa-twitter"></i>Twitter</span>
                </span>
              
            </div>
          

          
             <div class="cc-license motion-element" itemprop="license">
              
              
                
              
              
              
              <span class="exturl cc-opacity" data-url="aHR0cHM6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LW5jLXNhLzQuMC8="><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"/></span>
             </div>
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-inline">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8=" title="https://ihope.shinyapps.io/qRT-PCR-Pipeline/">qRT-PCR-Pipeline</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly90aXJhbWlzdXRlcy5naXRib29rcy5pby9hd2svY29udGVudC8=" title="https://tiramisutes.gitbooks.io/awk/content/">GitBook《awk学习总结》</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2NvdHRvbi5oemF1LmVkdS5jbi8=" title="http://cotton.hzau.edu.cn/">棉花遗传改良团队</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW8taW5mby10cmFpbmVlLmNvbS8=" title="http://www.bio-info-trainee.com/">生信菜鸟团-深圳大学城</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW90cmFpbmVlLmNvbS9mb3J1bS5waHA=" title="http://www.biotrainee.com/forum.php">生信技能树</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5pd2hnYW8uY29tLw==" title="http://www.iwhgao.com/">rabbit gao's blog</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9odWFucy5naXRodWIuaW8v" title="https://huans.github.io/">Huans(代谢+GWAS)</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5zZXF1ZW5jaW5nLnVpby5uby8=" title="http://www.sequencing.uio.no/">NSC-sequenceing</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2Jsb2cuam9iYm9sZS5jb20v" title="http://blog.jobbole.com/">伯乐在线</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9saW51eC5jbi8=" title="https://linux.cn/">Linux开源中文社区</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5yLWJsb2dnZXJzLmNvbS8=" title="http://www.r-bloggers.com/">R-bloggers</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9jcmFuLnItcHJvamVjdC5vcmcvb3RoZXItZG9jcy5odG1s" title="https://cran.r-project.org/other-docs.html">R-resource</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3R1dG9yaWFscy5pcS5oYXJ2YXJkLmVkdS9SL1JncmFwaGljcy9SZ3JhcGhpY3MuaHRtbA==" title="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html">ggplot2</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2xpbGliZWkubmV0Lw==" title="http://lilibei.net/">pele' blog</span>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </div>

      

      

    </div>
  </aside>
  


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2014 – <span itemprop="copyrightYear">2023</span>
  <span class="with-love" id="animate">
    <i class="fa fa-heart"></i>
  </span>
  <a class="theme-link" target="_blank" href="http://tiramisutes.github.io/about/"><span class="author" itemprop="copyrightHolder">tiramisutes</span></a>

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Symbols count total"></span>
  

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Reading time total"></span>
  
</div>


  <div class="powered-by">Powered: <span class="exturl theme-link" data-url="aHR0cHM6Ly9oZXhvLmlv">Hexo</span></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">▽ – <span class="exturl theme-link" data-url="aHR0cHM6Ly9naXRodWIuY29tL25leHQtdGhlbWUvaGV4by10aGVtZS1uZXh0">NexT.Muse</span></div>




        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  

  

  
    <span class="post-meta-item-icon">
      <i class="fa fa-line-chart"></i>
    </span>
    <span class="site-pv" title="Total Views">
    <a style="line-height: 1.5;" href="https://tongji.baidu.com/web/welcome/login" target="_blank">The total visits
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>&nbsp;times</a>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>












  



  
    
    
      
    
  
  <script color='0,0,255' opacity='0.5' zIndex='-1' count='20' src="//cdn.jsdelivr.net/gh/theme-next/theme-next-canvas-nest@1/canvas-nest-nomobile.min.js"></script>













  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.js"></script>


  


  <script src="/js/src/utils.js?v=7.0.1"></script>

  <script src="/js/src/motion.js?v=7.0.1"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.1"></script>



  

  


  <script src="/js/src/next-boot.js?v=7.0.1"></script>


  

  
  <script src="/js/src/exturl.js?v=7.0.1"></script>


  
  
  
    
  

  
    
      
    
  

  
  
  
    <script src="//cdn.jsdelivr.net/npm/quicklink@1/dist/quicklink.umd.js"></script>
    <script>
      
        window.addEventListener('load', () => {
      
        quicklink({
          timeout: 3000,
          priority: true,
          ignores: [uri => uri.includes('#'),uri => uri == 'http://tiramisutes.github.io/page/8/',]
        });
      
        });
      
    </script>
  


  
  

  

<script src="//cdn.jsdelivr.net/npm/leancloud-storage@3/dist/av-min.js"></script>



<script src="//unpkg.com/valine/dist/Valine.min.js"></script>

<script>
  var GUEST = ['nick', 'mail', 'link'];
  var guest = 'nick,mail,link';
  guest = guest.split(',').filter(function(item) {
    return GUEST.indexOf(item) > -1;
  });
  new Valine({
    el: '#comments',
    verify: false,
    notify: true,
    appId: 'c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz',
    appKey: 'iVJ6ST0e1O5xaiIhVwOyhbf3',
    placeholder: 'Just go go',
    avatar: 'mm',
    meta: guest,
    pageSize: '10' || 10,
    visitor: true
  });
</script>





  
  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  

  
  
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  

  


  
<style>
.pdfobject-container {
  position: relative;
  overflow: auto;
  width: 100%;
}
</style>
<script>
if ($('body').find('div.pdf').length) {
  $.ajax({
    type: 'GET',
    url: '//cdn.jsdelivr.net/npm/pdfobject@2.1.1/pdfobject.min.js',
    dataType: 'script',
    cache: true,
    success: function() {
      $('body').find('div.pdf').each(function(i, o) {
        PDFObject.embed($(o).attr('target'), $(o), {
          pdfOpenParams: {
            navpanes: 0,
            toolbar: 0,
            statusbar: 0,
            pagemode: 'thumbs',
            view: 'FitH'
          },
          PDFJS_URL: '/lib/pdf/web/viewer.html',
          height: $(o).attr('height') || '500px'
        });
      });
    },
  });
}
</script>


  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      bp.src = (curProtocol === 'https') ? 'https://zz.bdstatic.com/linksubmit/push.js' : 'http://push.zhanzhang.baidu.com/push.js';
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  
  
  
    
  
  <script src="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.js"></script>
  <script>
    
      pbOptions = {};
      
        pbOptions.iconStyle = "box";
      
        pbOptions.boxForm = "horizontal";
      
        pbOptions.position = "bottomCenter";
      
        pbOptions.networks = "Weibo,Wechat,Douban,QQZone,Twitter,Evernote,Facebook";
      
      new needShareButton('#needsharebutton-postbottom', pbOptions);
    
    
  </script>


  

  

  

  

</body>
</html>
