<!DOCTYPE html>












  


<html class="theme-next muse use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge"/>
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222"/>
<script type="text/javascript" src="http://tajs.qq.com/stats?sId=66189719" charset="UTF-8"></script>


  
  
    
  
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.css"/>





  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/Han/3.3.0/han.min.css"/>



  <meta http-equiv="Cache-Control" content="no-transform"/>
  <meta http-equiv="Cache-Control" content="no-siteapp"/>



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg"/>



  <meta name="msvalidate.01" content="74A687E04D4EF9914B7AA20824207409"/>









  
  
    
  
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.css"/>




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link rel="stylesheet" href="https://fonts.loli.net/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext"/>
  






<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2"/>

<link rel="stylesheet" href="/css/main.css?v=7.0.1"/>


  <link rel="apple-touch-icon" sizes="180x180" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon.png?v=7.0.1">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/small_favicon.png?v=7.0.1">


  <link rel="mask-icon" href="/images/favicon.svg?v=7.0.1" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '7.0.1',
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":true,"dimmer":false},
    back2top: true,
    back2top_sidebar: false,
    fancybox: true,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="hope bioinformatics blog">
<meta property="og:type" content="website">
<meta property="og:title" content="hope">
<meta property="og:url" content="http://tiramisutes.github.io/page/10/index.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="hope bioinformatics blog">
<meta property="og:locale" content="en">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="hope">
<meta name="twitter:description" content="hope bioinformatics blog">



  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml"/>




  <link rel="canonical" href="http://tiramisutes.github.io/page/10/"/>



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>hope – A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</title>
  






  <script>
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>







  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br/>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br/>Archives</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-code">

    
    
    
      
    

    

    <a href="/code/" rel="section"><i class="menu-item-icon fa fa-fw fa-code"></i> <br/>Code</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br/>Categories</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br/>Tags</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-footprint">

    
    
    
      
    

    

    <a href="/footprint/" rel="section"><i class="menu-item-icon fa fa-fw fa-image"></i> <br/>Footprint</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-qrt-pcr">

    
    
    
      
    

    

    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8="><i class="menu-item-icon fa fa-fw fa-cloud"></i> <br/>qRT-PCR</span>

  </li>
        
        
        
          
          <li class="menu-item menu-item-project">

    
    
    
      
    

    

    <a href="/project/" rel="section"><i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br/>Project</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-publish">

    
    
    
      
    

    

    <a href="/publish/" rel="section"><i class="menu-item-icon fa fa-fw fa-book"></i> <br/>Publish</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-sci-figure">

    
    
    
      
    

    

    <a href="/SCIFigure/" rel="section"><i class="menu-item-icon fa fa-fw fa-angellist"></i> <br/>SCI Figure</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br/>About</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-cv">

    
    
    
      
    

    

    <a href="/resume/" rel="section"><i class="menu-item-icon fa fa-fw fa-address-card"></i> <br/>CV</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br/>Search</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/12/01/miRNA.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/12/01/miRNA.html" class="post-title-link" itemprop="url">miRNA</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-12-01 14:06:06" itemprop="dateCreated datePublished" datetime="2015-12-01T14:06:06+08:00">2015-12-01</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/molecular-biology/" itemprop="url" rel="index"><span itemprop="name">molecular biology</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/12/01/miRNA.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/12/01/miRNA.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/12/01/miRNA.html" class="leancloud_visitors" data-flag-title="miRNA">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="简介">简介</h2><p>MicroRNAs (miRNAs) are short in sequence and are generated by enzymatic excision from precursor transcripts called primary miRNAs (pri-miRs), which until now had been assumed not to encode any proteins.<br><a href="/nature/journal/v520/n7545/full/nature14346.html">Lauressergues et al</a> find that plant pri-miRNAs contain short open reading frame sequences that encode regulatory peptides. This is the first report of a functional peptide being encoded by a pri-miR and provides a fresh perspective on the significance of pri-miR regions beyond those that directly give rise to miRNAs.<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/miRNA.png" alt=""><br>图片注释：MicroRNAs and their associated peptides. The precursors of plant microRNAs (miRNAs) are<br>pri-miR sequences, which are transcribed from DNA by the enzyme RNA polymerase II (Pol II). They are<br>then modified by capping and addition of a poly(A) tail. The miRNA duplexes are subsequently excised by<br>the enzyme Dicer-like1 and transported to the cytoplasm; other parts of the pri-miR are degraded. After<br>further processing, the resulting miRNA sequence guides repression of gene expression as part of the<br>RISC complex. Lauressergues et al.report that some pri-miRs contain short open reading frame (ORF)<br>sequences that can produce peptides (miPEP). The miPEPs enhance expression of the pri-miR, leading<br>to more miRNA and so more effective cleavage of the target gene’s messenger RNA. Such ORFs may avoid<br>degradation as part of pri-miRs that might exit the nucleus without being processed by Dicer-like1.</p>
<h2 id="关键信息总结">关键信息总结</h2><p>1）    Both plant and animal pri-miRs are transcribed from DNA in the nucleus by the enzyme RNA polymerase II.<br>2）    The structured (fold-back) region of the transcript surrounding the miRNA sequence is recognized and processed by one of two enzymes-Drosha or Dicer-like1(In plants, Dicer-like1 cuts out the miRNA in a duplex form).<br>3)    Transporter proteins export the excised sequences to the cytoplasm, where they are further processed before becoming competent to guide the RNA-induced silencing complex (RISC) in repressing target genes through either cleavage or translational repression of their mRNAs.<br>4)    The miPEPs had the same tissue distribution as their associated mature miRNAs and enhanced the expression and effectiveness of these miRNAs. Moreover, the miPEPs promoted the transcription of their corresponding pri-miR, rather than enhancing miRNA stability.</p>
<h2 id="miRNA合成途径">miRNA合成途径</h2><p>miRNA的合成大体经过：初级miRNA(pri-miRNA, primary miRNA)→前体miRNA(pre-miRNA, precursor-miRNA)→成熟miRNA(miRNA, mature miRNA)。<br>其具体合成途径存在两种形式：经典合成途径和Mirtron合成途径；<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/nihms311773f1.jpg" alt=""></p>
<h3 id="经典合成途径">经典合成途径</h3><p>来源：基因间隔区或编码基因内含子中；<br>细胞核内编码miRNA 的基因通过RNA聚合酶II 或RNA聚合酶III 转录生成初级miRNA(pri-miRNA)，pri-miRNA与来自蛋白质编码基因mRNA相似，有5’ 端帽式结构和3’ 端多聚腺苷酸尾结构，长度可达数千个碱基。接着pri-miRNA在一种RNaseIII(Drosha 酶)和它的伴侣分子(DiGeorge syndrome critical region gene 8,DGCR8)组成的复合物作用下，剪切为70~80 个核苷酸长度、具有茎环结构的miRNA 前体(pre-miRNA)。miRNA 前体在Ran-GTP 依赖的核质/ 细胞质转运蛋白——输出蛋白5(exportin 5)的作用下，从核内运输到胞质中。随后miRNA前体在另一种RNaseIII(Dicer酶)的作用下被剪切成21~25 个核苷酸长度，而且其5’端磷酸化和3’端有2 nt 的悬垂序列，类似于siRNA 的不完全配对的双链RNA，它们是由成熟miRNA与miRNA* 组成的二聚体，miRNA*是pre-miRNA中的一段RNA，其位置恰好与成熟的miRNA 相对。最后在RNA解旋酶作用下生成成熟miRNA 和miRNA*，成熟miRNA结合到RNA诱导的基因沉默复合物(RNA-induced silencing complex, RISC)中发挥作用，miRNA* 则被降解。<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/ruby-et-al-nature-mirtron.jpg" alt=""></p>
<h3 id="Mirtron途径">Mirtron途径</h3><p>2007年，Ruby等在研究黑腹果蝇和秀丽隐杆线虫的小RNA 序列时首次发现mirtron，是一类定位于mRNA编码基因内含子内的miRNA，与<strong>经典miRNA合成相比其合成过程不需要经过Drosha酶切割</strong>，其形式功能与miRNA相同，具有以下特点：<br>(1)来自长度为56 nt 的内含子序列；<br>(2)能够形成miRNA:miRNA*复合物；<br>(3)内含子序列及其二级结构非常保守；<br>(4)内含子存在于基因组中的外显子中间并且有”GU-AG”的典型内含子特点；<br>(5)最后形成的成熟miRNA总来自于3’ 端侧的序列；<br>(6)合成需要Ldbr(lariat-debranching enzyme,套索分支酶)(参与内含子剪接) 的作用；<br>(7)与经典的miRNA一样，依赖Ran-GTP/ 输出蛋白5 的转运机制转运出核，在Dicer 酶的作用下成为成熟的miRNA。<br><code>同时，mirtron在哺乳动物中和在无脊椎动物中有明显的不同，如：</code><br>(1)哺乳动物中的mirtron途径合成的功能miRNA大多数来自mirtron的5’茎序列；<br>(2)不管哺乳动物中的mirtron3’ 茎序列能否成为成熟miRNA，它的起始核苷酸总是尿嘧啶(U)或胞嘧啶(C);<br>(3)哺乳动物中mirtron悬垂序列(overhang, 发夹结构末端突出序列)大多数为单核苷酸(G)，只有少数是双核苷酸(AG)；<br>(4)哺乳动物中mirtron比经典的miRNA前体和无脊椎动物mirtron的GC 含量要高，自由能要低。</p>
<h2 id="阻断miRNA对mRNA的调控作用">阻断miRNA对mRNA的调控作用</h2><h3 id="miRNA海绵(miRNA_sponges)">miRNA海绵(miRNA sponges)</h3><p>miRNA海绵是一种miRNA靶基因的竞争性抑制剂，是将若干个miRNA的反义序列串联在一起，连接到合适的载体中表达，其转录物”吸附”相应的miRNA，与miRNA靶基因形成竞争，导致靶基因去抑制化。<br><span class="exturl" data-url="aHR0cDovL2Qud2FuZmFuZ2RhdGEuY29tLmNuL1BlcmlvZGljYWwvZ3d5eC1menN3MjAxNDA0MDEw" title="http://d.wanfangdata.com.cn/Periodical/gwyx-fzsw201404010">miRNA海绵表达载体的构建策略<i class="fa fa-external-link"></i></span></p>
<h3 id="circRNA">circRNA</h3><p>环状RNA（circRNA）对miRNA的吸附作用。<i class="fa fa-link" aria-hidden="true"></i><a href="http://tiramisutes.github.io/2016/04/12/Circular-RNAs/">http://tiramisutes.github.io/2016/04/12/Circular-RNAs/</a></p>
<h2 id="miRNA数据库">miRNA数据库</h2><p>miocroRNA databases:<span class="exturl" data-url="aHR0cDovL21pY3Jvcm5hZGF0YWJhc2VzLmNvbS9EYXRhYmFzZXMuaHRtbA==" title="http://micrornadatabases.com/Databases.html">http://micrornadatabases.com/Databases.html<i class="fa fa-external-link"></i></span></p>
<h2 id="miRNA相关工具">miRNA相关工具</h2><h3 id="mirtron_prediction">mirtron prediction</h3><p><span class="exturl" data-url="aHR0cDovL2NiaW8ubXNrY2Mub3JnL2xlc2xpZWxhYi9taXJ0cm9ucy8=" title="http://cbio.mskcc.org/leslielab/mirtrons/">Mirtron’SVM’prediction<i class="fa fa-external-link"></i></span></p>
<h3 id="MicroRNA_Target_Prediction">MicroRNA Target Prediction</h3><p><span class="exturl" data-url="aHR0cDovL3d3dy5taWNyb3JuYS5vcmcvbWljcm9ybmEvaG9tZS5kbw==" title="http://www.microrna.org/microrna/home.do">miRanda<i class="fa fa-external-link"></i></span> &#8212; miRNA target prediction for human, drosophila and zebrafish genomes<br><span class="exturl" data-url="aHR0cDovL3d3dy5taXJiYXNlLm9yZy8=" title="http://www.mirbase.org/">miRBase<i class="fa fa-external-link"></i></span> &#8212; a comprehensive repository for miRNAs and their predicted targets<br><span class="exturl" data-url="aHR0cDovL21pcmRiLm9yZy9taVJEQi8=" title="http://mirdb.org/miRDB/">miRDB<i class="fa fa-external-link"></i></span> &#8212; an online database for miRNA target prediction and functional annotations in animals<br><span class="exturl" data-url="aHR0cDovL21pcm5hbWFwLm1iYy5uY3R1LmVkdS50dy8=" title="http://mirnamap.mbc.nctu.edu.tw/">miRNAMap<i class="fa fa-external-link"></i></span> &#8212; a genomic maps of microRNA genes and their target genes in mammalian genomes<br><span class="exturl" data-url="aHR0cDovL3dhdHNvbi5jb21wYmlvLml1cHVpLmVkdTo4MDgwL21pcjJkaXNlYXNlLw==" title="http://watson.compbio.iupui.edu:8080/mir2disease/">miR2Disease<i class="fa fa-external-link"></i></span> &#8212; a database providing comprehensive resource of miRNA deregulation in various human diseases<br><span class="exturl" data-url="aHR0cDovL2RpYW5hLmNzbGFiLmVjZS5udHVhLmdyL3RhcmJhc2Uv" title="http://diana.cslab.ece.ntua.gr/tarbase/">TarBase<i class="fa fa-external-link"></i></span> &#8212; a comprehensive database of experimentally supported animal microRNA targets<br><span class="exturl" data-url="aHR0cDovL3BpY3Rhci5tZGMtYmVybGluLmRlLw==" title="http://pictar.mdc-berlin.de/">PicTar<i class="fa fa-external-link"></i></span> &#8212; microRNA targets for vertebrates, fly and nematodes<br><span class="exturl" data-url="aHR0cDovL3d3dy50YXJnZXRzY2FuLm9yZy8=" title="http://www.targetscan.org/">TargetScan<i class="fa fa-external-link"></i></span> &#8212; a search for the presence of conserved sites that match the seed of each miRNA<br><span class="exturl" data-url="aHR0cDovL3d3dy5ydXNzZWxsLmVtYmwuZGUvbWlSTkFzLw==" title="http://www.russell.embl.de/miRNAs/">Target Gene Prediction at EMBL<i class="fa fa-external-link"></i></span> &#8212; miRNA-Target predictions for Drosophila miRNAs</p>
<h3 id="Databases_for_microRNA_Expression">Databases for microRNA Expression</h3><p><span class="exturl" data-url="aHR0cDovL3d3dy5taWNyb3JuYS5vcmcvbWljcm9ybmEvZ2V0RXhwckZvcm0uZG8=" title="http://www.microrna.org/microrna/getExprForm.do">microRNA.org<i class="fa fa-external-link"></i></span> &#8212; predicted microRNA targets &amp; target downregulation scores. Experimentally observed expression patterns<br><span class="exturl" data-url="aHR0cDovLzIwMi4zOC4xMjYuMTUxL2htZGQvbWlybmEvbWQv" title="http://202.38.126.151/hmdd/mirna/md/">HMDD<i class="fa fa-external-link"></i></span> &#8212; Human MicroRNA Disease Database (HMDD) is a database that contains the experimentally supported miRNA-disease<br>association data, which are manually curated from publications. The dysfunction evidence or miRNAs and literature PubMed ID are also given<br><span class="exturl" data-url="aHR0cDovLzIwMi4zOC4xMjYuMTUxL2htZGQvbWlybmEvdGYv" title="http://202.38.126.151/hmdd/mirna/tf/">TransmiR<i class="fa fa-external-link"></i></span> &#8212; a web query-driven database integrating the experimentally supported transcription factor and miRNA regulator relations</p>
<h3 id="RNA_Secondary_Structure_Prediction">RNA Secondary Structure Prediction</h3><p><span class="exturl" data-url="aHR0cDovL2RpYW5hLnBjYmkudXBlbm4uZWR1L2NnaS1iaW4vbWljcm9fdC5jZ2kv" title="http://diana.pcbi.upenn.edu/cgi-bin/micro_t.cgi/">DIANA MicroTest<i class="fa fa-external-link"></i></span> &#8212; a prediction of miRNA-mRNA interaction<br><span class="exturl" data-url="aHR0cDovL21mb2xkLmJpb2luZm8ucnBpLmVkdS8=" title="http://mfold.bioinfo.rpi.edu/">mfold<i class="fa fa-external-link"></i></span> &#8212; tools for predicting the secondary structure of RNA and DNA, mainly by using thermodynamic methods<br><span class="exturl" data-url="aHR0cDovL21pcm5hLmltYmIuZm9ydGguZ3IvbWljcm9pbnNwZWN0b3Iv" title="http://mirna.imbb.forth.gr/microinspector/">microInspector<i class="fa fa-external-link"></i></span> &#8212; a web tool for detection of miRNA binding sites in an RNA sequence<br><span class="exturl" data-url="aHR0cDovL3d3dy5taXJuYWluZm8uY29tL0VuZXJneUNhbGMuYXNweA==" title="http://www.mirnainfo.com/EnergyCalc.aspx">miRNA Bioinfor<i class="fa fa-external-link"></i></span> &#8212; miRNA End Energy calculator which takes miRNA duplex to calculate free energy for 5 base pairs at<br>one end plus a dangling nucleotide<br><span class="exturl" data-url="aHR0cDovL21pcnJpbS5uY3JuYS5vcmcv" title="http://mirrim.ncrna.org/">miRRim<i class="fa fa-external-link"></i></span> &#8212; a method for detecting miRNA foldbacks based on hidden Markov model (HMM)<br><span class="exturl" data-url="aHR0cDovL214c2Nhcm5hLm5jcm5hLm9yZy8=" title="http://mxscarna.ncrna.org/">MXSCARNA<i class="fa fa-external-link"></i></span> &#8212; a multiple alignment tool for RNA sequences using progressive alignment based on pairwise structural alignment algorithm of SCARNA. Good for large scale analyses.<br><span class="exturl" data-url="aHR0cDovL2JpYmlzZXJ2LnRlY2hmYWsudW5pLWJpZWxlZmVsZC5kZS9ybmFoeWJyaWQv" title="http://bibiserv.techfak.uni-bielefeld.de/rnahybrid/">RNAhybrid<i class="fa fa-external-link"></i></span>&#8212; a tool for finding the minimum free energy hybridisation of a long and a short RNA</p>
<h3 id="MicroRNA_Homologous_Prediction">MicroRNA Homologous Prediction</h3><p><span class="exturl" data-url="aHR0cDovL2dyb3Vwcy5jc2FpbC5taXQuZWR1L3BhZy9taXJuYW1pbmVyLw==" title="http://groups.csail.mit.edu/pag/mirnaminer/">miRNAminer<i class="fa fa-external-link"></i></span> &#8212; a web-based tool used for homologous miRNA gene search in several species<br><span class="exturl" data-url="aHR0cDovL3Blb3BsZS5jc2FpbC5taXQuZWR1L2FraWV6dW4vbWlSdmlld2VyLw==" title="http://people.csail.mit.edu/akiezun/miRviewer/">miRviewer<i class="fa fa-external-link"></i></span> &#8212; a global view of homologous miRNA genes in many species<br><span class="exturl" data-url="aHR0cDovL3d3dy5uY2JpLm5sbS5uaWguZ292L3BtYy9hcnRpY2xlcy9QTUMyNjc2MjU3L3BkZi8xNDcxLTIxMDUtMTAtMTA1LnBkZg==" title="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2676257/pdf/1471-2105-10-105.pdf">RISCbinder<i class="fa fa-external-link"></i></span>&#8212; prediction of guide strand of microRNAs<br><span class="exturl" data-url="aHR0cDovL3RhZ2MudW5pdi1tcnMuZnIvbWlyZXZhbC8=" title="http://tagc.univ-mrs.fr/mireval/">Mireval<i class="fa fa-external-link"></i></span> &#8212; Sequence evaluation of microRNA properties</p>
<h3 id="MicroRNA_Deep_Sequencing">MicroRNA Deep Sequencing</h3><p><span class="exturl" data-url="aHR0cDovL3dlYi5iaW9pbmZvcm1hdGljcy5jaWNiaW9ndW5lLmVzL21pY3JvUk5BL21pUmFuYWx5c2VyLnBocA==" title="http://web.bioinformatics.cicbiogune.es/microRNA/miRanalyser.php">miRanalyzer<i class="fa fa-external-link"></i></span>&#8212; A microRNA detection and analysis tool for next-generation sequencing experiments<br><span class="exturl" data-url="aHR0cDovL2liaXMudGF1LmFjLmlsL21pUk5Ba2V5Lw==" title="http://ibis.tau.ac.il/miRNAkey/">miRNAkey<i class="fa fa-external-link"></i></span>&#8212; A software pipeline for the analysis of microRNA Deep Sequencing data<br><span class="exturl" data-url="aHR0cDovL3d3dy5tZGMtYmVybGluLmRlL2VuL3Jlc2VhcmNoL3Jlc2VhcmNoX3RlYW1zL3N5c3RlbXNfYmlvbG9neV9vZl9nZW5lX3JlZ3VsYXRvcnlfZWxlbWVudHMvcHJvamVjdHMvbWlSRGVlcC9pbmRleC5odG1s" title="http://www.mdc-berlin.de/en/research/research_teams/systems_biology_of_gene_regulatory_elements/projects/miRDeep/index.html">miRDeep<i class="fa fa-external-link"></i></span>&#8212; Discovering known and novel miRNAs from deep sequencing data</p>
<h2 id="miRNA百科">miRNA百科</h2><p><i class="fa fa-link" aria-hidden="true"></i><span class="exturl" data-url="aHR0cDovL3d3dy5nZW5lLXF1YW50aWZpY2F0aW9uLmRlL21pY3JvLXJuYS5odG1s" title="http://www.gene-quantification.de/micro-rna.html">microRNA.gene-quantification.info<i class="fa fa-external-link"></i></span></p>
<p>参考文献：<br><i class="fa fa-link" aria-hidden="true"></i><span class="exturl" data-url="aHR0cDovL3d3dy5uYXR1cmUuY29tL25hdHVyZS9qb3VybmFsL3Y1MjAvbjc1NDUvZnVsbC9uYXR1cmUxNDM3OC5odG1s" title="http://www.nature.com/nature/journal/v520/n7545/full/nature14378.html">Plant biology: Coding in non-coding RNAs<i class="fa fa-external-link"></i></span><br><i class="fa fa-link" aria-hidden="true"></i><span class="exturl" data-url="aHR0cDovL3d3dy5uY2JpLm5sbS5uaWguZ292L3BtYy9hcnRpY2xlcy9QTUMzMTg1MTg5Lw==" title="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3185189/">Mirtrons: microRNA biogenesis via splicing<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/12/01/chip-seq.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/12/01/chip-seq.html" class="post-title-link" itemprop="url">ChIP-sequencing介绍</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-12-01 13:21:42" itemprop="dateCreated datePublished" datetime="2015-12-01T13:21:42+08:00">2015-12-01</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/molecular-biology/" itemprop="url" rel="index"><span itemprop="name">molecular biology</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/12/01/chip-seq.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/12/01/chip-seq.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/12/01/chip-seq.html" class="leancloud_visitors" data-flag-title="ChIP-sequencing介绍">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="简介(From_Wikipedia,_the_free_encyclopedia)">简介(From Wikipedia, the free encyclopedia)</h2><p>ChIP-sequencing, also known as ChIP-seq, is a method used to analyze protein interactions with DNA. ChIP-seq combines chromatin immunoprecipitation (ChIP) with massively parallel DNA sequencing to identify the binding sites of DNA-associated proteins. It can be used to map global binding sites precisely for any protein of interest. Previously, ChIP-on-chip was the most common technique utilized to study these protein–DNA relations.<br>ChIP-seq is used primarily to determine how transcription factors and other chromatin-associated proteins influence phenotype-affecting mechanisms. Determining how proteins interact with DNA to regulate gene expression is essential for fully understanding many biological processes and disease states. </p>
<h2 id="流程">流程</h2><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/chip-seq.png" alt=""></p>
<h3 id="ChIP">ChIP</h3><p>The ChIP process enriches specific crosslinked DNA-protein complexes  using an <a href="/wiki/Antibody" title="Antibody">antibody</a> against the protein of interest. For a good description of the ChIP wet lab protocol see <a href="/wiki/ChIP-on-chip" title="ChIP-on-chip">ChIP-on-chip</a>. <a href="/wiki/Oligonucleotide" title="Oligonucleotide">Oligonucleotide</a> adaptors are then added to the small stretches of DNA that were bound to the protein of interest to enable massively parallel sequencing.</p>
<h3 id="Sequencing">Sequencing</h3><p>After size selection, all the resulting ChIP-DNA fragments are sequenced simultaneously using a genome sequencer. A single sequencing run can scan for genome-wide associations with high resolution, as opposed to large sets of tilingarrays required for lower resolution ChIP-chip.</p>
<h2 id="中文解释">中文解释</h2><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/chip2.png" alt=""><br>ChIP 测序的基本原分为 ChIP 和测序两个步骤。ChIP 是英文 Chromatin immunoprecipitation 的缩写，即染色质免疫沉淀，其步骤包括：细胞内蛋白质和 DNA 的交联、DNA 分子分离及片段化、免疫沉淀和解除交联。测序就是对解除交联后的 DNA 片段进行测序，制备文库时也包括连接测序接头和片段筛选等步骤。由于免疫沉淀的 DNA 片段在蛋白质结合区域周围富集，因此识别蛋白质结合区域转化为检测测序标签富集的区域。在信息处理中，这是一个信号检测的问题。在转录因子 ChIP 测序数据分析中，由于测序标签还会在转录因子结合位点周围形成分布，通过对分布的统计建模分析，可从数据中精确定位结合位点。 </p>
<h2 id="Computational_analysis">Computational analysis</h2><p>The read count data generated by ChIP-seq is massive. It motivates the development of computational analysis methods. To predict DNA-binding sites from ChIP-seq read count data, <a href="/wiki/Peak_calling" title="Peak calling">peak calling</a> methods have been developed. The most popular method is MACS which empirically models the shift size of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.<sup id="cite_ref-10" class="reference"><a href="#cite_note-10"><span>[</span>10<span>]</span></a></sup></p>
<p></p><p>Another relevant computational problem is Differential peak calling, which identifies significant differences in two ChIP-seq signals from distinct biological conditions. Differential peak callers segment two ChIP-seq signals and identify differential peaks using <a href="/wiki/Hidden_Markov_Model" title="Hidden Markov Model" class="mw-redirect">Hidden Markov Models</a>. Examples for two-stage differential peak callers are ChIPDiff<sup id="cite_ref-11" class="reference"><a href="#cite_note-11"><span>[</span>11<span>]</span></a></sup> and ODIN.<br>具体数据分析详细见：<i class="fa fa-link"></i><span class="exturl" data-url="aHR0cDovL3d3dy5wbG9iLm9yZy8yMDEyLzA5LzI5LzM3NjAuaHRtbA==" title="http://www.plob.org/2012/09/29/3760.html">http://www.plob.org/2012/09/29/3760.html<i class="fa fa-external-link"></i></span></p>
<p>Contribution from ：<br><span class="exturl" data-url="aHR0cHM6Ly9lbi53aWtpcGVkaWEub3JnL3dpa2kvQ2hJUC1zZXF1ZW5jaW5n" title="https://en.wikipedia.org/wiki/ChIP-sequencing">https://en.wikipedia.org/wiki/ChIP-sequencing<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL3d3dy5pbGx1bWluYS5jb20vZG9jdW1lbnRzL3Byb2R1Y3RzL2RhdGFzaGVldHMvZGF0YXNoZWV0X2NoaXBfc2VxdWVuY2UucGRm" title="http://www.illumina.com/documents/products/datasheets/datasheet_chip_sequence.pdf">http://www.illumina.com/documents/products/datasheets/datasheet_chip_sequence.pdf<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL2Jpb2luZm8uYXUudHNpbmdodWEuZWR1LmNuL21lbWJlci94d2FuZy9maWxlcy9UaGVzaXNfWGlXYW5nX09MLnBkZg==" title="http://bioinfo.au.tsinghua.edu.cn/member/xwang/files/Thesis_XiWang_OL.pdf">http://bioinfo.au.tsinghua.edu.cn/member/xwang/files/Thesis_XiWang_OL.pdf<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/12/01/wegt-http.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/12/01/wegt-http.html" class="post-title-link" itemprop="url">Linux 下批量下载 http 中链接内容</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-12-01 08:31:16" itemprop="dateCreated datePublished" datetime="2015-12-01T08:31:16+08:00">2015-12-01</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/12/01/wegt-http.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/12/01/wegt-http.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/12/01/wegt-http.html" class="leancloud_visitors" data-flag-title="Linux 下批量下载 http 中链接内容">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/kobas.png" alt=""><br>KOBAS这个软件所需数据库有3540个物种数据，如何实现批量下载…<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/kobas2.png" alt=""><br>察看html网页源文件可发现：所有这些下载链接都整齐排列于 <code><a href="../download/sqlite3/ko.db.gz">KO</a><br></code>这样一个html语法结构中，所以可以通过正则表达式匹配找出所有的下载链接；</p>
<h2 id="方法1:python抓取网页并下载"><i class="fa fa-key"></i>方法1:python抓取网页并下载</h2><p>python中实现爬虫，解析网页html文件和下载文件的简单容易上手的优秀模块分别为requests，BeautifulSoup和re，通过三者的组合可实现爬虫网页并批量下载。<br>导入所需模块<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#!~/bin/Python-2.7.10/</span></span><br><span class="line">from BeautifulSoup import BeautifulSoup</span><br><span class="line">import requests</span><br><span class="line">import re</span><br><span class="line"><span class="comment">#—--name:url-batch-download.py--</span></span><br><span class="line"><span class="comment">#—--修改:hope--</span></span><br><span class="line"><span class="comment"># point to output directory</span></span><br><span class="line">outpath = <span class="string">'/backend3/'</span></span><br><span class="line">url = <span class="string">'http://kobas.cbi.pku.edu.cn/site/download_db.jsp'</span></span><br><span class="line">mbyte=1024*1024</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> <span class="string">'Reading: '</span>, url</span><br><span class="line">html = requests.get(url).text</span><br><span class="line">soup = BeautifulSoup(html)</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> <span class="string">'Processing: '</span>, url</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> soup.findAll(<span class="string">'a'</span>, href=True):</span><br><span class="line">    gzurl = name[<span class="string">'href'</span>]</span><br><span class="line">    strinfo = re.compile(<span class="string">'\../'</span>)</span><br><span class="line">    xzp= strinfo.sub(<span class="string">'http://kobas.cbi.pku.edu.cn/'</span>,gzurl)</span><br><span class="line">    <span class="built_in">print</span> xzp</span><br><span class="line">    <span class="keyword">if</span>( xzp.endswith(<span class="string">'.gz'</span>) ):</span><br><span class="line">        outfname = outpath + xzp.split(<span class="string">'/'</span>)[-1]</span><br><span class="line">        r = requests.get(xzp, stream=True)</span><br><span class="line">        <span class="keyword">if</span>( r.status_code == requests.codes.ok ) :</span><br><span class="line">            fsize = int(r.headers[<span class="string">'content-length'</span>])</span><br><span class="line">            <span class="built_in">print</span> <span class="string">'Downloading %s (%sMb)'</span> % ( outfname, fsize/mbyte )</span><br><span class="line">            with open(outfname, <span class="string">'wb'</span>) as fd:</span><br><span class="line">                <span class="keyword">for</span> chunk <span class="keyword">in</span> r.iter_content(chunk_size=1024): <span class="comment"># chuck size can be larger</span></span><br><span class="line">                    <span class="keyword">if</span> chunk: <span class="comment"># ignore keep-alive requests</span></span><br><span class="line">                        fd.write(chunk)</span><br><span class="line">                fd.close()</span><br></pre></td></tr></table></figure></p>
<p>修稿py文件为可执行文件<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chmod +x url-batch-download.py</span><br></pre></td></tr></table></figure></p>
<p>开始爬虫并批量下载<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python2.7 url-batch-download.py</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-key"></i>方法2：wget批量下载<br>1） wget下载整个html页面后批量下载<br>wget都很熟悉，但是通常只是用来下载单个文件，如果需要批量下载呢？<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget http://kobas.cbi.pku.edu.cn/site/download_db.jsp</span><br></pre></td></tr></table></figure></p>
<p>对于这个链接下载完后是jsp后缀文件，可以重命名为index.html。<br>2） 开始批量下载<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wget -i index.html -F -B http://kobas.cbi.pku.edu.cn/site/download_db.jsp</span><br></pre></td></tr></table></figure></p>
<p>参数解释：<br>-i 表示从文件导入链接，默认是直接按行读取URL<br>-F 表示将文件以HTML的格式解析，其实就是解析<a><br>-B 因为发现解析出来的链接用的都是相对路径，而为了下载这个文件，必须在相对路径前添加上Base URL，-B就是用来添加Base URL。</a></p>
<p>这样就可以批量下载当前目录中的文件了。</p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/30/DNA-methylation.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/30/DNA-methylation.html" class="post-title-link" itemprop="url">DNA methylation</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-30 23:49:08" itemprop="dateCreated datePublished" datetime="2015-11-30T23:49:08+08:00">2015-11-30</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/molecular-biology/" itemprop="url" rel="index"><span itemprop="name">molecular biology</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/30/DNA-methylation.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/30/DNA-methylation.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/30/DNA-methylation.html" class="leancloud_visitors" data-flag-title="DNA methylation">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="简介">简介</h2><p>DNA methylation is a process by which methyl groups are added to DNA. Methylation modifies the function of the DNA. When located in a gene promoter, DNA methylation typically acts to repress gene transcription. DNA methylation is essential for normal development and is associated with a number of key processes including genomic imprinting, X-chromosome inactivation, repression of repetitive elements, and carcinogenesis.<br>DNA methylation at the 5 position of cytosine has the specific effect of reducing gene expression and has been found in every vertebrate examined. In adult somatic cells (cells in the body, not used for reproduction), DNA methylation typically occurs in a CpG dinucleotide context; non-CpG methylation is prevalent in embryonic stem cells,[5][6][7] and has also been indicated in neural development.</p>
<h2 id="原理">原理</h2><p>在甲基转移酶的催化下，DNA的CG两个核苷酸的胞嘧啶被选择性地添加甲基，形成5－甲基胞嘧啶，这常见于基因的5’-CG-3’序列。大多数脊椎动物基因组DNA都有少量的甲基化胞嘧啶，主要集中在基因5’端的非编码区，并成簇存在。甲基化位点可随DNA的复制而遗传，因为DNA复制后，甲基化酶可将新合成的未甲基化的位点进行甲基化。DNA的甲基化可引起基因的失活，DNA甲基化导致某些区域DNA构象变化，从而影响了蛋白质与DNA的相互作用，甲基化达到一定程度时会发生从常规的B-DNA向Z-DNA的过渡，由于Z-DNA结构收缩，螺旋加深，使许多蛋白质因子赖以结合的原件缩入大沟而不利于转录的起始，导致基因失活。<br>DNA甲基化主要形成5－甲基胞嘧啶（5-mC）和少量的N6-甲基腺嘌呤（N6-mA）及7－甲基鸟嘌呤（7-mG）。</p>
<h2 id="结构基因编辑">结构基因编辑</h2><p>含有很多CpG 结构，2CpG 和2GPC 中两个胞嘧啶的5 位碳原子通常被甲基化，且两个甲基集团在DNA 双链大沟中呈特定三维结构。基因组中60%～ 90% 的CpG 都被甲基化，未甲基化的CpG 成簇地组成CpG 岛，位于结构基因启动子的核心序列和转录起始点。有实验证明超甲基化阻遏转录的进行。DNA 甲基化可引起基因组中相应区域染色质结构变化，使DNA 失去核酶&ouml;限制性内切酶的切割位点，以及DNA 酶的敏感位点，使染色质高度螺旋化，凝缩成团，失去转录活性。5 位C 甲基化的胞嘧啶脱氨基生成胸腺嘧啶，由此可能导致基因置换突变，发生碱基错配： T2G，如果在细胞分裂过程中不被纠正，就会诱发遗传病或癌症，而且，生物体甲基化的方式是稳定的，可遗传的。</p>
<h2 id="The_Role_of_Methylation_in_Gene_Expression">The Role of Methylation in Gene Expression</h2><p>1) Not all genes are active at all times. DNA methylation is one of several epigenetic mechanisms that cells use to control gene expression.<br>2) 5-azacytidine Experiments Provide Early Clues to the Role of Methylation in Gene Expression.</p>
<h2 id="候选基因CpG岛区域和候选基因甲基化与表观基因组学">候选基因CpG岛区域和候选基因甲基化与表观基因组学</h2><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/jiajihua.png" alt=""><br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">Figure 1 | Altered DNA-methylation patterns <span class="keyword">in</span> tumorigenesis. The hypermethylation of CpG islands of tumoursuppressor</span><br><span class="line">genes is a common alteration <span class="keyword">in</span> cancer cells, and leads to the transcriptional inactivation of these genes</span><br><span class="line">and the loss of their normal cellular <span class="built_in">functions</span>. This contributes to many of the hallmarks of cancer cells. At the same</span><br><span class="line">time, the genome of the cancer cell undergoes global hypomethylation at repetitive sequences, and tissue-specific</span><br><span class="line">and imprinted genes can also show loss of DNA methylation. In some cases, this hypomethylation is known to</span><br><span class="line">contribute to cancer cell phenotypes, causing changes such as loss of imprinting, and might also contribute to the</span><br><span class="line">genomic instability that characterizes tumours. E, exon.</span><br></pre></td></tr></table></figure></p>
<p>如图1左所示，在正常细胞中，位于抑癌基因启动子区域的CpG岛处于低水平或未甲基化状态，此时抑癌基因处于正常的开放状态，抑癌基因不断表达抑制肿瘤的发生。而在肿瘤细胞中，该区域的CpG岛被高度甲基化，染色质构象发生改变，抑癌基因的表达被关闭，从而导致细胞进入细胞周期，凋亡丧失，DNA修复缺陷，血管生成以及细胞粘附功能缺失等，最终导致肿瘤发生。同样，如图1右所示，对于在正常细胞中处于高度甲基化的一些基因和重复序列，如果其甲基化水平降低，这些基因将表达和重复序列将激活，从而导致基因印记丢失，细胞过度增长，不合适的细胞特异性表达，基因组脆性增加，以及内寄生序列(endoparasitic sequence)的激活，最终也导致肿瘤发生。</p>
<h2 id="DNA甲基化检测方法">DNA甲基化检测方法</h2><p>据检测样本不同，可以分为DNA和mRNA。现有方法，绝大部分都是取样于细胞的DNA，根据研究水平，又将这些方法归为3大类，即：基因组甲基化水平(Methylation Content)的分析，候选基因甲基化分析，和基因组层次的DNA甲基化模式(Methylation pattern)与甲基化谱(Methylation Profiling)分析。</p>
<h3 id="基因组甲基化水平(Methylation_Content)">基因组甲基化水平(Methylation Content)</h3><h4 id="高效液相色谱(High-performance_Liquid_Chromatography,_HPLC)">高效液相色谱(High-performance Liquid Chromatography, HPLC)</h4><p>根据DNA或蛋白分子量和构象的不同而使其加以分离。由于在动态相和静态相下分子的光吸收度并不相同而加以定量。随着系统的压强的增加，其分辨率增高。故而能够定量测定基因组整体水平DNA甲基化水平。<br>过程是将DNA样品先经盐酸或氢氟酸水解成碱基，水解产物通过色谱柱，结果与标准品比较，用紫外光测定吸收峰值及其量，计算5 mC/(5mC+5C)的积分面积就得到基因组整体的甲基化水平。这是一种检测DNA甲基化水平的标准方法。</p>
<h4 id="高效毛细管电泳法(High-performance_Capillary_Electrophoresis,_HPCE)">高效毛细管电泳法(High-performance Capillary Electrophoresis, HPCE)</h4><p>利用窄孔熔融石英毛细管来从复合物中分离不同化学组分的技术。其基础是在强电场下不同分子的由于其所带电荷，大小，结构以及疏水性等不同而相互分开。用HPCE方法处理DNA水解产物来确定5mC水平，简便，经济且敏感性高。<br><i class="fa fa-comments-o"></i>以上各种方法虽然能够明确检测出目的序列中所有CpG位点的甲基化状况，但并不能对甲基化位点进行定位。</p>
<h3 id="候选基因(Candidate_Gene)">候选基因(Candidate Gene)</h3><h4 id="甲基化敏感性限制性内切酶-PCR/Southern法(_methylation-sensitive_restriction_Endonuclease_-PCR/Southern,_MSRE-PCR/Southern)">甲基化敏感性限制性内切酶-PCR/Southern法( methylation-sensitive restriction Endonuclease -PCR/Southern, MSRE-PCR/Southern)</h4><p>这种方法利用甲基化敏感性限制性内切酶对甲基化区的不切割的特性，将DNA消化为不同大小的片段后，进行Southern或PCR扩增分离产物，明确甲基化状态再进行分析。常使用的甲基化敏感的限制性内切酶有HpaⅡ-MspⅠ(CCGG)和SmaⅠ-Xmal(CCCGGG)等。</p>
<h4 id="重亚硫酸盐测序法(Bisulphite_Sequencing)">重亚硫酸盐测序法(Bisulphite  Sequencing)</h4><p>该方法首先用重亚硫酸盐使DNA中未发生甲基化的胞嘧啶脱氨基转变成尿嘧啶,而甲基化的胞嘧啶保持不变，行PCR扩增所需片段,则尿嘧啶全部转化成胸腺嘧啶，最后,对PCR产物进行测序并且与未经处理的序列比较,判断是否CpG位点发生甲基化。此方法是精确度很高，能明确目的片段中每一个CpG位点的甲基化状态，但需要大量的克隆测序,过程较为繁琐、昂贵。</p>
<h4 id="甲基化特异性的PCR(methylation-specific_PCR,_MS-PCR)">甲基化特异性的PCR(methylation-specific PCR, MS-PCR)</h4><p>该方法同样DNA先用重亚硫酸盐处理，随后行引物特异性的PCR。其设计两对引物，分别与重亚硫酸盐处理后的序列互补配对，即一对结合处理后的甲基化DNA链，另一对结合处理后的非甲基化DNA链。检测MS-PCR扩增产物，如果用针对处理后甲基化DNA链的引物能扩增出片段，则说明该被检测的位点存在甲基化；反之亦然。</p>
<h4 id="甲基化荧光法(MethyLight)">甲基化荧光法(MethyLight)</h4><p>结合重亚硫酸盐处理待测DNA片段，设计一个能与待测位点区互补的探针，探针的5’端连接报告荧光，3’端连接淬灭荧光，随后行实时定量PCR。如果探针能够与DNA杂交，则在PCR用引物延伸时，TaqDNA聚合酶5′到3′端的外切酶活性会将探针序列上5′端的报告荧光切下，淬灭荧光不再能对报告荧光进行抑制，这样报告荧光发光，测定每个循环报告荧光的强度即可得到该位点的甲基化情况及水平。本方法高效，迅速，具备可重复、所需样本量少、不需要电泳分离的特点。</p>
<h4 id="焦磷酸测序(Pyrosequencing)">焦磷酸测序(Pyrosequencing)</h4><p>该方法，由4种酶催化同一反应体系中的酶级联化学发光反应，在每一轮测序反应中，只加入一种dNTP，若该dNTP与模板配对，聚合酶就能将其加入到引物链中并释放出等摩尔数的焦磷酸(PPi)。PPi可最终转化为可见光信号，并由PyrogramTM转化为一个峰值，其高度与核苷酸数目成正比。当用于甲基化检测时，经重亚硫酸盐处理的序列可以看作是C-T型的SNP改变。其操作简单，结果准确可靠，可以行大规模分析。</p>
<h4 id="结合重亚硫酸盐的限制性内切酶法(combined_bisulfiterestriction_analysis,_COBRA)">结合重亚硫酸盐的限制性内切酶法(combined bisulfiterestriction analysis, COBRA)</h4><p>这种方法对标本DNA行重亚硫酸盐处理及PCR扩增，处理后原甲基化的胞嘧啶被保留，而非甲基化的胞嘧啶变为胸腺嘧啶。随后用限制性内切酶对转化后PCR产物切割的特性以识别原标本DNA的甲基化状况。该方法相对简单，不需预先知道CpG位点及样本序列。</p>
<h3 id="基因组范围的DNA甲基化模式(Methylation_pattern)与甲基化谱(Methylation_Profiling)">基因组范围的DNA甲基化模式(Methylation　pattern)与甲基化谱(Methylation Profiling)</h3><h4 id="限制性标记基因组扫描(Restriction_Landmark_Genomic_Scanning,_RLGS)">限制性标记基因组扫描(Restriction Landmark Genomic Scanning, RLGS)</h4><p>RLGS是最早适用于基因组范围DNA甲基化分析的方法之一。该方法先用甲基化敏感的稀频限制性内切酶NotⅠ消化基因组DNA，甲基化位点保留，标记末端、切割、行一维电泳，随后再用更高频的甲基化不敏感的内切酶切割，行二维电泳，这样甲基化的部分被切割开并在电泳时显带，得到RLGS图谱与正常对照得出缺失条带即为甲基化的可能部位。</p>
<h4 id="甲基化间区位点扩增(amplification_of_inter-methylated_sites,_AIMS)">甲基化间区位点扩增(amplification of inter-methylated sites, AIMS)</h4><p>AIMS是基于任意引物PCR(Arbitrary Primed PCR)的一种方法，由于任意引物PCR使用寡核苷酸连接子(linker) 进行连接，不需要依赖任何序列的先验信息。在该方法中，用来进行扩增的模板序列首先通过甲基化敏感的限制性内切酶进行消化而富集，其特异性由该酶酶切片断一端的特定序列结合连接子来保证。随后，由内切酶进行第二次消化，再次连接，提纯进行PCR扩增，最后电泳，提取目的序列进行测序。</p>
<h4 id="甲基化CpG岛扩增(Methylated_CpG-island_amplification,_MCA)">甲基化CpG岛扩增(Methylated  CpG-island amplification, MCA)</h4><p>MCA也是基于任意引物PCR的方法，该方法使用两种对甲基化具有不同敏感度的限制性内切酶(如SmaI和XmaI)先后进行消化，然后对甲基化敏感的限制性酶切片断进行接头(Adaptor)连接，行PCR，那些富含CpG的序列就会被选择性的扩增。该方法对甲基化分析和克隆甲基化差异性基因都非常有帮助。</p>
<h4 id="差异甲基化杂交(Differential_Methylation_Hybridization,DMH)">差异甲基化杂交(Differential Methylation Hybridization,DMH)</h4><p>DMH属于一种芯片技术，在该技术中，包括扩增子(Amplicon)生成和CGI文库筛选两个重要组成部分。在扩增子生成中，首先用MseI来酶切DNA样本，然后接上连接子，并去除重复序列，这时的样本一分为二，其一直接进行PCR扩增，生成仅由MseI处理过的扩增子，而另一半则用甲基化敏感的酶BstUI进行消化,然后进行PCR扩增，生成由MseI/BstUI共同处理过的扩增子。CGI文库通过筛查出重复序列，然后进行PCR，选出含有BstUI位点的克隆，最后这些克隆一式两份点到芯片上，制备成CGI芯片。然后把两种不同的扩增子分别杂交到相应的CGI克隆点上，最后通过差异性对比检测出那些未甲基化位点。</p>
<h4 id="由连接子介导PCR出的HpaII小片断富集分析(HpaII_tiny_fragement_Enrichment_by_Ligation-mediated_PCR,_HELP)">由连接子介导PCR出的HpaII小片断富集分析(HpaII tiny fragement Enrichment by Ligation-mediated PCR, HELP)</h4><p>该方法用HpaII与其对甲基化敏感同裂酶MspI对同一基因组序列进行消化，产生不同的代表性序列，然后对此序列进行连接子介导的PCR，瑞和进行电泳等比较分析或将此DNA样本共杂交到基因组芯片上进行分析。这种方法已经揭示了大量的组织特异性，差异甲基化区域，并用于正常细胞和癌症细胞的基因组比较分析。</p>
<h4 id="甲基化DNA免疫沉淀法(Methylated_DNA_immunoprecipitation,_MeDIP)">甲基化DNA免疫沉淀法(Methylated DNA immunoprecipitation, MeDIP)</h4><p>这是一种高效富集甲基化DNA的方法。在该方法中，可与5mC特异性结合的抗体加入到变性的基因组DNA片段中，从而使甲基化的基因组片断免疫沉淀，形成富集。通过与已有DNA微芯片技术相结合，从而进行大规模DNA甲基化分析。该方法简便，特异性高，适合DNA甲基化组学(DNA Methylome)的分析。</p>
<h2 id="候选基因甲基化分析网址">候选基因甲基化分析网址</h2><p><span class="exturl" data-url="aHR0cDovL2thdGFoZGluLm1zc20uZWR1L2tpc21ldGgvcmV2cGFnZS5wbA==" title="http://katahdin.mssm.edu/kismeth/revpage.pl">http://katahdin.mssm.edu/kismeth/revpage.pl<i class="fa fa-external-link"></i></span><br>Contribution from ：<span class="exturl" data-url="aHR0cDovL3d3dy5uYXR1cmUuY29tL3NjaXRhYmxlL3RvcGljcGFnZS90aGUtcm9sZS1vZi1tZXRoeWxhdGlvbi1pbi1nZW5lLWV4cHJlc3Npb24tMTA3MA==" title="http://www.nature.com/scitable/topicpage/the-role-of-methylation-in-gene-expression-1070">http://www.nature.com/scitable/topicpage/the-role-of-methylation-in-gene-expression-1070<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL3d3dy5uYXR1cmUuY29tL25yZy9qb3VybmFsL3Y4L240L2Z1bGwvbnJnMjAwNS5odG1s" title="http://www.nature.com/nrg/journal/v8/n4/full/nrg2005.html">http://www.nature.com/nrg/journal/v8/n4/full/nrg2005.html<i class="fa fa-external-link"></i></span><br><span class="exturl" data-url="aHR0cDovL3d3dy5saWZlb21pY3MuY29tLz9wPTE4NDU4" title="http://www.lifeomics.com/?p=18458">http://www.lifeomics.com/?p=18458<i class="fa fa-external-link"></i></span></p>
<h2 id="甲基化引物设计">甲基化引物设计</h2><p>Methyl Primer Express® Software v1.0<br><span class="exturl" data-url="aHR0cHM6Ly93d3cudGhlcm1vZmlzaGVyLmNvbS9vcmRlci9jYXRhbG9nL3Byb2R1Y3QvNDM3NjA0MQ==" title="https://www.thermofisher.com/order/catalog/product/4376041">https://www.thermofisher.com/order/catalog/product/4376041<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/30/degradome.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/30/degradome.html" class="post-title-link" itemprop="url">Degradome-降解组</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-30 23:36:38" itemprop="dateCreated datePublished" datetime="2015-11-30T23:36:38+08:00">2015-11-30</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Bioinformatics/" itemprop="url" rel="index"><span itemprop="name">Bioinformatics</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/30/degradome.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/30/degradome.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/30/degradome.html" class="leancloud_visitors" data-flag-title="Degradome-降解组">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="Degradome_sequencing">Degradome sequencing</h2><p>Degradome sequencing (Degradome-Seq),also referred to as parallel analysis of RNA ends (PARE),is a modified version of 5’-Rapid Amplification of cDNA Ends (RACE) using high-throughput, deep sequencing method using as Illumina’s SBS technology. Degradome sequencing provides a comprehensive means of analyzing patterns of RNA degradation.</p>
<p>Degradome sequencing has been used to identify microRNA (miRNA) cleavage sites,because miRNAs can cause endonucleolytic cleavage of mRNA by extensive and often perfect complementarity to mRNAs.Degradome sequencing revealed many known and novel plant miRNA (siRNA) targets. Recently, degradome sequencing also has been applied to identify animal (human and mouse) miRNA-derived cleavages.</p>
<h2 id="原理">原理</h2><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/degradome.jpg" alt=""><br>在植物体内绝大多数的miRNA是利用剪切作用调控靶基因的表达，且剪切常发生在miRNA与mRNA互补区域的第十位核苷酸上。靶基因经剪切产生二个片段，5’ 剪切片段和3’ 剪切片段。其中3’ 剪切片段，包含有自由的5’ 单磷酸和3’ polyA尾巴，可被RNA连接酶连接，连接产物可用于下游高通量测序；而含有5’ 帽子结构的完整基因，含有帽子结构的5’ 剪切片段或是其他缺少5’ 单磷酸基团的RNA是无法被RNA酶连接的，因而无法进入下游的测序实验；对测序数据进行深入地比对分析，可以直观地发现在mRNA序列的某个位点会出现一个波峰，而该处正是候选的miRNA剪切位点。</p>
<h2 id="測序數據的分析方法">測序數據的分析方法</h2><p>測序數據使用賓州大學 Addo-Quaye 等建立CleaveLand 分析方法進行比對分析，可以直觀地發現在 mRNA 序列的某個位點會出現一個波峰，而該處正是候選的 miRNA 剪切位元點。</p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/FISH.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/FISH.html" class="post-title-link" itemprop="url">Fluorescence in situ Hybridization(荧光原位杂交)</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 22:51:46" itemprop="dateCreated datePublished" datetime="2015-11-29T22:51:46+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/molecular-biology/" itemprop="url" rel="index"><span itemprop="name">molecular biology</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/FISH.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/FISH.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/FISH.html" class="leancloud_visitors" data-flag-title="Fluorescence in situ Hybridization(荧光原位杂交)">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/Multiplex_ViewRNA_FISH_Assay_in_Jurkat_and_HeLa_cells.jpg" alt=""></p>
<h2 id="简介">简介</h2><p>荧光原位杂交（fluorescence in situ hybridization，FISH）是在20世纪80年代末在放射性原位杂交技术的基础上发展起来的一种非放射性分子细胞遗传技术，以荧光标记取代同位素标记而形成的一种新的原位杂交方法。探针首先与某种介导分子（reporter molecule）结合，杂交后再通过免疫细胞化学过程连接上荧光染料。FISH的基本原理是将DNA（或RNA）探针用特殊的核苷酸分子标记，然后将探针直接杂交到染色体或DNA纤维切片上，再用与荧光素分子偶联的单克隆抗体与探针分子特异性结合来检测DNA序列在染色体或DNA纤维切片上的定性、定位、相对定量分析。FISH具有安全、快速、灵敏度高、探针能长期保存、能同时显示多种颜色等优点，不但能显示中期分裂相，还能显示于间期核。同时在荧光原位杂交基础上又发展了多彩色荧光原位杂交技术和染色质纤维荧光原位杂交技术.。</p>
<h2 id="Preparation_and_hybridization_process_–_DNA">Preparation and hybridization process – DNA</h2><p><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/FISH.jpg" alt=""><br>First, a probe is constructed. The probe must be large enough to hybridize specifically with its target but not so large as to impede the hybridization process. The probe is tagged directly with fluorophores, with targets for antibodies or with biotin. Tagging can be done in various ways, such as nick translation, or PCR using tagged nucleotides.</p>
<p>Then, an interphase or metaphase chromosome preparation is produced. The chromosomes are firmly attached to a substrate, usually glass. Repetitive DNA sequences must be blocked by adding short fragments of DNA to the sample. The probe is then applied to the chromosome DNA and incubated for approximately 12 hours while hybridizing. Several wash steps remove all unhybridized or partially hybridized probes. The results are then visualized and quantified using a microscope that is capable of exciting the dye and recording images.</p>
<p>If the fluorescent signal is weak, amplification of the signal may be necessary in order to exceed the detection threshold of the microscope. Fluorescent signal strength depends on many factors such as probe labeling efficiency, the type of probe, and the type of dye. Fluorescently tagged antibodies or streptavidin are bound to the dye molecule. These secondary components are selected so that they have a strong signal.</p>
<h2 id="Difference_between_Southern_Blot_and_Fluorescence_in_situ_hybridization_(FISH)_?">Difference between Southern Blot and Fluorescence in situ hybridization (FISH) ?</h2><p>FISH is performed on intact chromosomes through interphase or metaphase.it doesnt need DNA extraction and is performed on microscopic slides. its a type of kariotyping.its use is to determine the chromosomal abberations(numerical and structural) of a patient.<br>But,in Southern blotting you should first extract DNA of the specimen then cleave it with a restriction enzyme and then run on gel electrophoresis. the next step is to blot the DNA on a membrane and hybridaze it with a labeled probe. the use of suthern blot is in detecting mutations and restriction length fragment polymorphysms.<br><i class="fa fa-volume-up"></i><span class="exturl" data-url="aHR0cHM6Ly9hbnN3ZXJzLnlhaG9vLmNvbS9xdWVzdGlvbi9pbmRleD9xaWQ9MjAxMjA2MTEwOTU1NDBBQTFZWGhs" title="https://answers.yahoo.com/question/index?qid=20120611095540AA1YXhl">https://answers.yahoo.com/question/index?qid=20120611095540AA1YXhl<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/plyr.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/plyr.html" class="post-title-link" itemprop="url">R数据整形术之plyr</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 16:24:33" itemprop="dateCreated datePublished" datetime="2015-11-29T16:24:33+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/plyr.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/plyr.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/plyr.html" class="leancloud_visitors" data-flag-title="R数据整形术之plyr">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <p>plyr包可以进行类似于数据透视表的操作，将数据分割成更小的数据，对分割后的数据进行些操作，最后把操作的结果汇总。<br>本文主要介绍以下内容：<br>Split-Aapply-Combine 原理介绍<br>baby_names的名字排名<br>求分段拟合的系数<br>部分其他函数介绍<br>在正式开始之前，请确保电脑上已经安装plyr，如果没有，通过install.packages()函数安装。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">&gt; require(plyr)  <span class="comment">#载入plyr包</span></span><br><span class="line">&gt; <span class="comment">##假设有美国新生婴儿的取名汇总，每一年，会统计男孩和女孩的取名情况，形成如下的一张表。baby_names数据集包含1880 ~ 2008年间的数据， 包含统计的年份(year)，新生婴儿的性别、名字、以及改名字的比例。</span></span><br><span class="line">&gt; baby_names&lt;-read.csv(<span class="string">"baby-names.csv"</span>)</span><br><span class="line">&gt; head(baby_names)</span><br><span class="line">  year    name  percent sex</span><br><span class="line">1 1880    John 0.081541 boy</span><br><span class="line">2 1880 William 0.080511 boy</span><br><span class="line">3 1880   James 0.050057 boy</span><br><span class="line">4 1880 Charles 0.045167 boy</span><br><span class="line">5 1880  George 0.043292 boy</span><br><span class="line">6 1880   Frank 0.027380 boy</span><br></pre></td></tr></table></figure></p>
<h2 id="以提问并解决问题的形式对plyr做介绍。">以提问并解决问题的形式对plyr做介绍。</h2><p>想知道数据集中，每年都有多少记录？<br>数据集中，男孩和女孩名的各自排名？<br>男孩名和女孩名各自排名前100在当年中的比例？<br>数据集中，每年都有多少记录<br>先假设我们有某一年的数据，我们会如何统计其中的记录数呢？由于数据集中，每条记录一行，只需要统计对应的行数就可以得到对应的记录数。<br>写个函数试试<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt; record_count &lt;- <span class="keyword">function</span>(df) &#123;</span><br><span class="line">+     <span class="built_in">return</span>(data.frame(count = nrow(df)))</span><br><span class="line">+ &#125;</span><br></pre></td></tr></table></figure></p>
<p>返回值类型是data.frame类型，是为即将介绍的ddply()函数做铺垫。先来看看2008年，数据集中有多少记录。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; baby_names_2008 &lt;- subset(baby_names, year == 2008)</span><br><span class="line">&gt; record_count(baby_names_2008)</span><br><span class="line">  count</span><br><span class="line">1  2000</span><br></pre></td></tr></table></figure></p>
<p>结果显示2000条，貌似我们已经得到答案。下面想想，该如何得到1880 ~ 2008这129年间，每年的记录数呢？<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">&gt; baby_names_1880_2008&lt;-ddply(baby_names,     <span class="comment"># 数据集</span></span><br><span class="line">+                  .(year),        <span class="comment"># 分类的标准</span></span><br><span class="line">+                  record_count    <span class="comment"># 函数</span></span><br><span class="line">+                             )</span><br><span class="line">&gt; head(baby_names_1880_2008)</span><br><span class="line">  year count</span><br><span class="line">1 1880  2000</span><br><span class="line">2 1881  2000</span><br><span class="line">3 1882  2000</span><br><span class="line">4 1883  2000</span><br><span class="line">5 1884  2000</span><br><span class="line">6 1885  2000</span><br><span class="line">&gt; dim(baby_names_1880_2008)</span><br><span class="line">[1] 129   2</span><br></pre></td></tr></table></figure></p>
<p>ddply解释：<br>定义了一个负责计数的函数record_count()<br>调用ddply()，这里出现刚刚定义的函数<br>ddply()函数是plyr包中用于对data.frame结构的数据做处理的函数，其结果也是data.frame。ddply的参数列表如下：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">ddply(.data, .variables, .fun = NULL, ..., .progress = <span class="string">"none"</span>,</span><br><span class="line">  .inform = FALSE, .drop = TRUE, .parallel = FALSE, .paropts = NULL)</span><br></pre></td></tr></table></figure></p>
<p>各部分解释如下<br>第一个参数是要操作的原始数据集，比如baby<em>name<br>第二个参数是按照某个（也可以几个）变量，对数据集分割，比如按照year对数据集分割，可以写成.(year)的形式<br>第三个参数是具体执行操作的函数，对分割后的每一个子数据集，调用该函数<br>第四个参数可选，表示第三个参数对应函数所需的额外参数<br>其他参数，可以暂时不用考虑。ddply()函数会自动的将分割后的每一小部分的计算结果汇总，以data.frame的格式保存。分割后的数据，是fun的第一个参数。<br>在上面的描述中，提到的分割、<em>_操作、汇总</em></em>，在plyr包中是一种处理方式(“frame”)，即”Split - Apply - Combine”。在plyr包中有很多这种处理方式的函数，在介绍这些函数之前，我们再来看看ddply()的一些更深入的用法。</p>
<h2 id="各年，男孩名与女孩名的各自排名">各年，男孩名与女孩名的各自排名</h2><p>以2008年的数据为例，男孩名”Jacob”的比例最高，排名应当是第一，”Michael”紧跟其后，排名应当第二，依此类推。对于女孩名，”Emma”排名第一，”Isabella”排名第二，”Emily”排名第三等等。我们希望得到这样的结果。</p>
<p>对于2008年的数据，可以通过简单的rank即可得到，不过要对男孩和女孩分别排序。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">baby_names_2008_boy &lt;- subset(baby_names_2008, sex == <span class="string">"boy"</span>) <span class="comment"># 获取男孩名</span></span><br><span class="line">baby_names_2008_boy<span class="variable">$rank</span> &lt;- rank(- baby_names_2008_boy<span class="variable">$percent</span>) <span class="comment"># 排序</span></span><br><span class="line">head(baby_names_2008_boy) <span class="comment"># 查看</span></span><br></pre></td></tr></table></figure></p>
<p>如何利用ddply()对原始数据集做相应的操作呢？这里需要介绍R语言中的一个函数transform()，该函数对原始数据集做一些操作，并把结果存储在原始数据中，更详细的用法，参见帮助文档?transform。</p>
<p>第一个版本的处理方式是这样的<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">ddply(baby_names, </span><br><span class="line">      .(year, sex), </span><br><span class="line">      transform, </span><br><span class="line">      rank = rank(-percent, ties.method = <span class="string">"first"</span>)</span><br><span class="line">)</span><br></pre></td></tr></table></figure></p>
<p>第二个参数有点变化，除了year，还有sex，这表示对baby_name数据集，对year和sex分类（类似于SQL中的group by year, sex）。<br>第四个参数是transform的额外参数，如果查看transform的帮助文档，其函数调用方式如下：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">transform(_data, ...)</span><br></pre></td></tr></table></figure></p>
<p>第一参数为操作的数据，在ddply()中为按年份和性别分割后的子数据集；后面的…参数是tag = value的形式，这种tag:value将追加在数据中。<br>由于rank默认对数据进行升序排序，若要实现逆序排序，常规的做法是将数据的符号取反，这也就是上面的rank函数中出现-percent的原因。在plyr中，有一个类似的函数，实现取反的操作，是desc。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- 1:10</span><br><span class="line">desc(x)</span><br><span class="line"><span class="comment"># -1  -2  -3  -4  -5  -6  -7  -8  -9 -10</span></span><br></pre></td></tr></table></figure></p>
<p>所以，上面对percent取反的操作，可以写得更优雅些，就有了第二个版本的函数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">baby_names &lt;- ddply(baby_names, </span><br><span class="line">                    .(year, sex), </span><br><span class="line">                    transform, </span><br><span class="line">                    rank = rank(desc(percent), ties.method = <span class="string">"first"</span>)</span><br><span class="line">)</span><br></pre></td></tr></table></figure></p>
<p>注意这里把结果赋给了baby_name，因为后面还会用到排名的信息，就把结果保存下来。</p>
<h2 id="排名前100的男孩名与女孩名在当年中的比例">排名前100的男孩名与女孩名在当年中的比例</h2><p>跟前一问类似，处理方法是：</p>
<p>把每年排名前100的数据筛选出来<br>把男孩和女孩对应的percent相加<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">baby_names_top100 &lt;- subset(baby_names, rank &lt;= 100)  <span class="comment"># 将前100排名的数据筛选出来</span></span><br><span class="line">baby_names_top100_trend &lt;- ddply(baby_names_top100, </span><br><span class="line">                                 .(year, sex), <span class="comment"># 按年和性别分割</span></span><br><span class="line">                                 summarize, <span class="comment"># 汇总数据</span></span><br><span class="line">                                 trend = sum(percent)) <span class="comment"># 汇总方式（求和）</span></span><br></pre></td></tr></table></figure></p>
<p>这里出现一个新的操作函数summarize()，该函数是对数据做汇总，与transform不一样的是，该函数并不追加结果到原始数据，而是产生新的数据集。比如想知道，2008年的男孩名中，排名最高和最低的名字的百分比之差，可以通过如下方式求得：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">summarize(baby_names_2008_boy, trend = max(percent) - min(percent))</span><br><span class="line"><span class="comment"># 0.010266</span></span><br></pre></td></tr></table></figure></p>
<p>回到刚才的问题，从1880 ~ 2008年间，男孩名与女孩名的前100所占比例（可以衡量名字大众化的程度）到底是什么样的呢？画个图就知道了。<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/ddply.jpeg" alt=""><br>Contribution from ：<span class="exturl" data-url="aHR0cDovL3d3dy5qaWFuc2h1LmNvbS9wL2JmZGRmZTI5YWEzOQ==" title="http://www.jianshu.com/p/bfddfe29aa39">http://www.jianshu.com/p/bfddfe29aa39<i class="fa fa-external-link"></i></span></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/reshape2.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/reshape2.html" class="post-title-link" itemprop="url">R数据整形术之reshape2</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 15:23:03" itemprop="dateCreated datePublished" datetime="2015-11-29T15:23:03+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/reshape2.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/reshape2.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/reshape2.html" class="leancloud_visitors" data-flag-title="R数据整形术之reshape2">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="数据类型">数据类型</h2><p>宽数据<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#   ozone   wind  temp</span></span><br><span class="line"><span class="comment"># 1 23.62 11.623 65.55</span></span><br><span class="line"><span class="comment"># 2 29.44 10.267 79.10</span></span><br><span class="line"><span class="comment"># 3 59.12  8.942 83.90</span></span><br><span class="line"><span class="comment"># 4 59.96  8.794 83.97</span></span><br></pre></td></tr></table></figure></p>
<p>长数据<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#    variable  value</span></span><br><span class="line"><span class="comment"># 1     ozone 23.615</span></span><br><span class="line"><span class="comment"># 2     ozone 29.444</span></span><br><span class="line"><span class="comment"># 3     ozone 59.115</span></span><br><span class="line"><span class="comment"># 4     ozone 59.962</span></span><br><span class="line"><span class="comment"># 5      wind 11.623</span></span><br><span class="line"><span class="comment"># 6      wind 10.267</span></span><br><span class="line"><span class="comment"># 7      wind  8.942</span></span><br><span class="line"><span class="comment"># 8      wind  8.794</span></span><br><span class="line"><span class="comment"># 9      temp 65.548</span></span><br><span class="line"><span class="comment"># 10     temp 79.100</span></span><br><span class="line"><span class="comment"># 11     temp 83.903</span></span><br><span class="line"><span class="comment"># 12     temp 83.968</span></span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-desktop"></i>长数据有一列数据是变量的类型，有一列是变量的值。长数据不一定只有两列。ggplot2需要长类型的数据，plyr也需要长类型的数据，大多数的模型(比如lm(), glm()以及gam())也需要长数据。</p>
<h2 id="reshape2_包">reshape2 包</h2><p>reshape2 用得比较多的是melt和cast两个函数。</p>
<p>melt函数对宽数据进行处理，得到长数据；<br>cast函数对长数据进行处理，得到宽数据；</p>
<h2 id="melt函数">melt函数</h2><p>melt(参数)<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">melt(data,id.vars,measure.vars,</span><br><span class="line">    variable.name = <span class="string">"variable"</span>, ..., na.rm = FALSE,</span><br><span class="line">    value.name = <span class="string">"value"</span>)</span><br></pre></td></tr></table></figure></p>
<p>其中id.vars可以指定一系列变量，然后measure.vars就可以留空了，这样生成的新数据会保留id.vars的所有列，然后增加两个新列：variable和value，一个存储变量的名称一个存储变量值。<br>此处用R内置的airquality数据集<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(airquality)</span><br><span class="line">  Ozone Solar.R Wind Temp Month Day</span><br><span class="line">1    41     190  7.4   67     5   1</span><br><span class="line">2    36     118  8.0   72     5   2</span><br><span class="line">3    12     149 12.6   74     5   3</span><br><span class="line">4    18     313 11.5   62     5   4</span><br><span class="line">5    NA      NA 14.3   56     5   5</span><br><span class="line">6    28      NA 14.9   66     5   6</span><br></pre></td></tr></table></figure></p>
<p>首先将列名改成小写，然后查看相应的数据<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">&gt; names(airquality)&lt;- tolower(names(airquality))</span><br><span class="line">&gt; head(airquality)</span><br><span class="line">  ozone solar.r wind temp month day</span><br><span class="line">1    41     190  7.4   67     5   1</span><br><span class="line">2    36     118  8.0   72     5   2</span><br><span class="line">3    12     149 12.6   74     5   3</span><br><span class="line">4    18     313 11.5   62     5   4</span><br><span class="line">5    NA      NA 14.3   56     5   5</span><br><span class="line">6    28      NA 14.9   66     5   6</span><br><span class="line"><span class="comment">##直接用metl函数处理上述的数据</span></span><br><span class="line">&gt; library(<span class="string">"reshape2"</span>)</span><br><span class="line">&gt; aql &lt;- melt(airquality)</span><br><span class="line">No id variables; using all as measure variables</span><br><span class="line">&gt; head(aql)</span><br><span class="line">  variable value</span><br><span class="line">1    ozone    41</span><br><span class="line">2    ozone    36</span><br><span class="line">3    ozone    12</span><br><span class="line">4    ozone    18</span><br><span class="line">5    ozone    NA</span><br><span class="line">6    ozone    28</span><br><span class="line">&gt; tail(aql)</span><br><span class="line">    variable value</span><br><span class="line">913      day    25</span><br><span class="line">914      day    26</span><br><span class="line">915      day    27</span><br><span class="line">916      day    28</span><br><span class="line">917      day    29</span><br><span class="line">918      day    30</span><br></pre></td></tr></table></figure></p>
<p>默认情况下，melt认为所有数值列的变量均有值。很多情况下，这都是我们想要的情况。在这里，我们想知道每个月(month)以及每天(day)的ozone, solar.r, wind以及temp的值。因此，我们需要告诉melt，month和day是”ID variables”。ID variables就是那些能够区分不同行数据的变量，个人感觉类似于数据库中的主键。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">&gt; aql &lt;- melt(airquality, id.vars = c(<span class="string">"month"</span>, <span class="string">"day"</span>))</span><br><span class="line">&gt; head(aql)</span><br><span class="line">  month day variable value</span><br><span class="line">1     5   1    ozone    41</span><br><span class="line">2     5   2    ozone    36</span><br><span class="line">3     5   3    ozone    12</span><br><span class="line">4     5   4    ozone    18</span><br><span class="line">5     5   5    ozone    NA</span><br><span class="line">6     5   6    ozone    28</span><br></pre></td></tr></table></figure></p>
<p>如果我们想修改长数据中的列名，该如何操作呢?<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">&gt; aql &lt;- melt(airquality, id.vars = c(<span class="string">"month"</span>, <span class="string">"day"</span>),</span><br><span class="line">+             variable.name = <span class="string">"climate_variable"</span>, </span><br><span class="line">+             value.name = <span class="string">"climate_value"</span>)</span><br><span class="line">&gt; head(aql)</span><br><span class="line">  month day climate_variable climate_value</span><br><span class="line">1     5   1            ozone            41</span><br><span class="line">2     5   2            ozone            36</span><br><span class="line">3     5   3            ozone            12</span><br><span class="line">4     5   4            ozone            18</span><br><span class="line">5     5   5            ozone            NA</span><br><span class="line">6     5   6            ozone            28</span><br></pre></td></tr></table></figure></p>
<h2 id="cast函数">cast函数</h2><p>从宽格式数据变换到长格式的数据比较直观，然后反过来则需要一些二外的功夫。</p>
<p>在reshape2中有好几个cast版本的函数。若你经常使用data.frame，就需要使用dcast函数。acast函数返回向量、矩阵或者数组。</p>
<p>dcast借助于公式来描述数据的形状，左边参数表示”ID variables”，而右边的参数表示measured variables。可能需要几次尝试，才能找到合适的公式。</p>
<p>这里，我们需要告知dcast，month和day是ID variables，variable则表示measured variables。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">&gt; aql &lt;- melt(airquality, id.vars = c(<span class="string">"month"</span>, <span class="string">"day"</span>))</span><br><span class="line">&gt; aqw &lt;- dcast(aql, month + day ~ variable)</span><br><span class="line">&gt; head(aqw)</span><br><span class="line">  month day ozone solar.r wind temp</span><br><span class="line">1     5   1    41     190  7.4   67</span><br><span class="line">2     5   2    36     118  8.0   72</span><br><span class="line">3     5   3    12     149 12.6   74</span><br><span class="line">4     5   4    18     313 11.5   62</span><br><span class="line">5     5   5    NA      NA 14.3   56</span><br><span class="line">6     5   6    28      NA 14.9   66</span><br><span class="line">&gt; head(airquality) <span class="comment"># 与原始数据比较</span></span><br><span class="line">  ozone solar.r wind temp month day</span><br><span class="line">1    41     190  7.4   67     5   1</span><br><span class="line">2    36     118  8.0   72     5   2</span><br><span class="line">3    12     149 12.6   74     5   3</span><br><span class="line">4    18     313 11.5   62     5   4</span><br><span class="line">5    NA      NA 14.3   56     5   5</span><br><span class="line">6    28      NA 14.9   66     5   6</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-pencil-square"></i>If it isn’t clear to you what just happened there, then have a look at this illustration:<br><img src="https://raw.githubusercontent.com/wiki/tiramisutes/blog_image/dcast-illustration.png" alt=""><br><i class="fa fa-bars"></i>蓝色阴影块是能够表示每一行数据的ID variables；红色阴影块包含了将待生成数据的列名；而灰色的数据表示用于填充相关区域的数据。<br><i class="fa fa-user-times"></i>令人产生疑惑的情况往往是，一个数据单元有一个以上的数据。比如，我们的ID variables不包含day，<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; dcast(aql, month ~ variable)</span><br><span class="line">Aggregation <span class="keyword">function</span> missing: defaulting to length</span><br><span class="line">  month ozone solar.r wind temp</span><br><span class="line">1     5    31      31   31   31</span><br><span class="line">2     6    30      30   30   30</span><br><span class="line">3     7    31      31   31   31</span><br><span class="line">4     8    31      31   31   31</span><br><span class="line">5     9    30      30   30   30</span><br></pre></td></tr></table></figure></p>
<p>再次查看dcast的输出数据，可以看到每个单元是month与climate组合的个数。所得到数据是month对应的day的记录数。当每个单元有多个数据是，需要告诉dcast如何聚合(aggregate)这些数据，比如取均值(mean)，计算中位数(median)，或者简单的求和(sum)。比如，在这里，我们简单的计算下均值，同时通过na.rm = TRUE删除NA值。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; dcast(aql, month ~ variable, fun.aggregate = mean, na.rm = TRUE)</span><br><span class="line">  month    ozone  solar.r      wind     temp</span><br><span class="line">1     5 23.61538 181.2963 11.622581 65.54839</span><br><span class="line">2     6 29.44444 190.1667 10.266667 79.10000</span><br><span class="line">3     7 59.11538 216.4839  8.941935 83.90323</span><br><span class="line">4     8 59.96154 171.8571  8.793548 83.96774</span><br><span class="line">5     9 31.44828 167.4333 10.180000 76.90000</span><br></pre></td></tr></table></figure></p>
<h2 id="Additional_help">Additional help</h2><p>Read the package help:<br><code>help(package = &quot;reshape2&quot;)</code></p>

<p>See the <code>reshape2</code> website:<br><span class="exturl" data-url="aHR0cDovL2hhZC5jby5uei9yZXNoYXBlLw==" title="http://had.co.nz/reshape/">http://had.co.nz/reshape/<i class="fa fa-external-link"></i></span></p>

<p>And read the paper on <code>reshape</code>:<br>Wickham, H. (2007). Reshaping data with the <code>reshape</code> package.<br>21(12):1–20.<br><span class="exturl" data-url="aHR0cDovL3d3dy5qc3RhdHNvZnQub3JnL3YyMS9pMTI=" title="http://www.jstatsoft.org/v21/i12">http://www.jstatsoft.org/v21/i12<i class="fa fa-external-link"></i></span><br><br>(But note that the paper is written for the <code>reshape</code> package not the<br><code>reshape2</code> package.)</p>


          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/find.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/find.html" class="post-title-link" itemprop="url">Linux常用命令之find</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 15:14:58" itemprop="dateCreated datePublished" datetime="2015-11-29T15:14:58+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/find.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/find.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/find.html" class="leancloud_visitors" data-flag-title="Linux常用命令之find">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="想查看当前文件夹及子文件夹里有没有文件名为”abc”的文件">想查看当前文件夹及子文件夹里有没有文件名为”abc”的文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -name abc</span></span><br></pre></td></tr></table></figure>
<p><code>.</code> ：表示当前目录</p>
<p>-name：表示要根据名称查找</p>
<h2 id="想查看当前文件夹及子文件夹里有没有”xyz”目录">想查看当前文件夹及子文件夹里有没有”xyz”目录</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -type d -name xyz</span></span><br></pre></td></tr></table></figure>
<p>-type：表示设定类型，d表示文件夹类型，可以替换为f(普通文件)、l(链接文件)</p>
<h2 id="想找出当前文件夹及子文件夹里所有后缀是”-txt”的文件">想找出当前文件夹及子文件夹里所有后缀是”.txt”的文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -name "*.txt"</span></span><br></pre></td></tr></table></figure>
<h2 id="想查找当前目录及其子文件夹中”roc”用户自己的文件有哪些">想查找当前目录及其子文件夹中”roc”用户自己的文件有哪些</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -user roc</span></span><br></pre></td></tr></table></figure>
<p>-user：用于设定所属用户的名称，此处可替换为-group，即所属用户组的名称</p>
<h2 id="想查找当前文件夹及子文件夹里权限设定为755的所有文件">想查找当前文件夹及子文件夹里权限设定为755的所有文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -perm 755</span></span><br></pre></td></tr></table></figure>
<p>-perm：用于设定权限</p>
<h2 id="想查找当前文件夹及子文件夹里的同时含有b字符和3字符的文件：用到正则表达式技术">想查找当前文件夹及子文件夹里的同时含有b字符和3字符的文件：用到正则表达式技术</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -regex '.*b.*3'</span></span><br></pre></td></tr></table></figure>
<p>-regex：表示使用正则表达式进行匹配。请注意，此命令会和”全路径”进行匹配，也就是说前面要加.*，因为输出结果中会有”./“符号。</p>
<h2 id="如果想全部输出用find命令查找出的”*-abc”文件的内容">如果想全部输出用find命令查找出的”*.abc”文件的内容</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -type f -name "*.abc" -exec cat &#123;&#125; \;</span></span><br></pre></td></tr></table></figure>
<p>-exec 表示由find找到的匹配项会作为”-exec后面设定的命令”的参数</p>
<p>可以使用-ok代替-exec，这样对每个匹配项进行操作，都会要求用户确认（y为是，n为否）</p>
<p>命令最后的{} \; 别忘了写，其中{}代表用find查找到的结果中的每一个查找项。</p>
<h2 id="查找当前目录下在5分钟内被访问过的文件">查找当前目录下在5分钟内被访问过的文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -amin -5</span></span><br></pre></td></tr></table></figure>
<p>访问过用amin，修改过用mmin，文件状态改变过用cmin</p>
<p>精确到分钟的用amin,mmin,cmin，精确到天的用atime,mtime,ctime</p>
<p>在5分钟之内的用-5，在5分钟以上的用＋5</p>
<h2 id="想查找当前目录及子目录下文件大小大于10M的所有文件">想查找当前目录及子目录下文件大小大于10M的所有文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -size +10000000c</span></span><br></pre></td></tr></table></figure>
<p>-size：表示文件大小，＋表示大于某个数，－表示小于某个数。c表示单位是字节，你可以将c换成k,M,G。</p>
<h2 id="上述所有的find命令都是查找当前目录及其子目录。如果不想深入到子目录中，而是只查找当前一层目录，则可以：">上述所有的find命令都是查找当前目录及其子目录。如果不想深入到子目录中，而是只查找当前一层目录，则可以：</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># find . -maxdepth 1 -name "*.c"</span></span><br></pre></td></tr></table></figure>
<p></p><p></p><br>Contribution from ：<br><span class="exturl" data-url="aHR0cDovL3JvY2xpbnV4LmNuLz9wPTE4" title="http://roclinux.cn/?p=18">http://roclinux.cn/?p=18<i class="fa fa-external-link"></i></span><p></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/11/29/cut.html"/>

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes"/>
      <meta itemprop="description" content="hope bioinformatics blog"/>
      <meta itemprop="image" content="/images/author.jpg"/>
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope"/>
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">
                
                
                <a href="/2015/11/29/cut.html" class="post-title-link" itemprop="url">Linux常用命令之cut</a>
              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2015-11-29 14:55:55" itemprop="dateCreated datePublished" datetime="2015-11-29T14:55:55+08:00">2015-11-29</time>
            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/Linux/" itemprop="url" rel="index"><span itemprop="name">Linux</span></a></span>

                
                
              
            </span>
          

          
            
            
              
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
            
                <span class="post-meta-item-text">Comments: </span>
                <a href="/2015/11/29/cut.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2015/11/29/cut.html" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
            <span id="/2015/11/29/cut.html" class="leancloud_visitors" data-flag-title="Linux常用命令之cut">
              <span class="post-meta-divider">|</span>
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              
                <span class="post-meta-item-text">Views: </span>
              
                <span class="leancloud-visitors-count"></span>
                <span>℃</span>
            </span>
          

          

          
            <div class="post-symbolscount">
              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article: </span>
                
                <span title="Words count in article"></span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time"></span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        
          
            <h2 id="定位">定位</h2><p>cut命令主要是接受三个定位方法：</p>
<h3 id="第一，字节（bytes），用选项-b"><i class="fa fa-ship"></i>第一，字节（bytes），用选项<code>-b</code></h3><p>-b支持形如3-5的写法，而且多个定位之间用逗号隔开就成了,如：3-5，7<br>但有一点要注意，cut命令如果使用了-b选项，那么执行此命令时，cut会先把-b后面所有的定位进行从小到大排序，然后再提取<br>-3表示从第一个字节到第三个字节，而3-表示从第三个字节到行尾</p>
<h3 id="第二，字符（characters），用选项-c"><i class="fa fa-train"></i>第二，字符（characters），用选项<code>-c</code></h3><p>主要用于提取中文操作</p>
<h3 id="第三，域（fields），用选项-f"><i class="fa fa-fighter-jet"></i>第三，域（fields），用选项<code>-f</code></h3><p>为什么会有”域”的提取呢，因为刚才提到的-b和-c只能在固定格式的文档中提取信息，而对于非固定格式的信息则束手无策。这时候”域”就派上用场了。<br>-d  来设置间隔符<br><i class="fa fa-comments"></i>如果遇到空格和制表符时，怎么分辨呢？<br><figure class="highlight kotlin"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ cat tab_space.txt</span><br><span class="line"><span class="keyword">this</span> <span class="keyword">is</span> tab	finish.</span><br><span class="line"><span class="keyword">this</span> <span class="keyword">is</span> several space      finish.</span><br><span class="line">$ sed -n l tab_space.txt</span><br><span class="line"><span class="keyword">this</span> <span class="keyword">is</span> tab\tfinish.$</span><br><span class="line"><span class="keyword">this</span> <span class="keyword">is</span> several space      finish.$</span><br></pre></td></tr></table></figure></p>
<p>看到了吧，如果是制表符（TAB），那么会显示为\t符号，如果是空格，就会原样显示。<br>通过此方法即可以判断制表符和空格了。<br>注意，上面<code>sed -n</code>后面的字符是<code>L</code>的小写字母，不要看错。</p>
<h2 id="取反">取反</h2><p><code>--complement：补足被选择的字节、字符或字段；</code><br><figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ cut -f2 --complement dddd </span><br><span class="line"><span class="number">1</span>       <span class="number">3</span></span><br><span class="line">$  cat dddd </span><br><span class="line"><span class="number">1</span>       <span class="number">2</span>       <span class="number">3</span></span><br></pre></td></tr></table></figure></p>

          
        
      
    </div>

    

    
    
    

    

    

    
      
    
    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/9/"><i class="fa fa-angle-left" aria-label="Previous page"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/9/">9</a><span class="page-number current">10</span><a class="page-number" href="/page/11/">11</a><span class="space">&hellip;</span><a class="page-number" href="/page/15/">15</a><a class="extend next" rel="next" href="/page/11/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      

      <div class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <a href="/" class="site-author-image" rel="start" style="border:none">
              <img class="site-author-image" itemprop="image"
                src="/images/author.jpg"
                alt="tiramisutes"/>
              </a>
            
              <p class="site-author-name" itemprop="name">tiramisutes</p>
              <div class="site-description motion-element" itemprop="description">hope bioinformatics blog</div>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">148</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  
                    
                      <a href="/categories/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">17</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  
                    
                      <a href="/tags/">
                    
                  
                    
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">112</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly9naXRodWIuY29tL3RpcmFtaXN1dGVz" title="GitHub &rarr; https://github.com/tiramisutes"><i class="fa fa-fw fa-github"></i>GitHub</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="bWFpbHRvOnRpcmFtaXN1dGVze2F0fTEyNi5jb20=" title="E-Mail &rarr; mailto:tiramisutes{at}126.com"><i class="fa fa-fw fa-envelope"></i>E-Mail</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cDovL3dlaWJvLmNvbS90aXJhbWlzdXRlc1Rlcw==" title="Weibo &rarr; http://weibo.com/tiramisutesTes"><i class="fa fa-fw fa-globe"></i>Weibo</span>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <span class="exturl" data-url="aHR0cHM6Ly90d2l0dGVyLmNvbS9ob3BldG9neQ==" title="Twitter &rarr; https://twitter.com/hopetogy"><i class="fa fa-fw fa-twitter"></i>Twitter</span>
                </span>
              
            </div>
          

          
             <div class="cc-license motion-element" itemprop="license">
              
              
                
              
              
              
              <span class="exturl cc-opacity" data-url="aHR0cHM6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LW5jLXNhLzQuMC8="><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"/></span>
             </div>
          

          
          
            <div class="links-of-blogroll motion-element links-of-blogroll-inline">
              <div class="links-of-blogroll-title">
                <i class="fa  fa-fw fa-link"></i>
                Links
              </div>
              <ul class="links-of-blogroll-list">
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9paG9wZS5zaGlueWFwcHMuaW8vcVJULVBDUi1QaXBlbGluZS8=" title="https://ihope.shinyapps.io/qRT-PCR-Pipeline/">qRT-PCR-Pipeline</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly90aXJhbWlzdXRlcy5naXRib29rcy5pby9hd2svY29udGVudC8=" title="https://tiramisutes.gitbooks.io/awk/content/">GitBook《awk学习总结》</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2NvdHRvbi5oemF1LmVkdS5jbi8=" title="http://cotton.hzau.edu.cn/">棉花遗传改良团队</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW8taW5mby10cmFpbmVlLmNvbS8=" title="http://www.bio-info-trainee.com/">生信菜鸟团-深圳大学城</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5iaW90cmFpbmVlLmNvbS9mb3J1bS5waHA=" title="http://www.biotrainee.com/forum.php">生信技能树</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5pd2hnYW8uY29tLw==" title="http://www.iwhgao.com/">rabbit gao's blog</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9odWFucy5naXRodWIuaW8v" title="https://huans.github.io/">Huans(代谢+GWAS)</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5zZXF1ZW5jaW5nLnVpby5uby8=" title="http://www.sequencing.uio.no/">NSC-sequenceing</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2Jsb2cuam9iYm9sZS5jb20v" title="http://blog.jobbole.com/">伯乐在线</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9saW51eC5jbi8=" title="https://linux.cn/">Linux开源中文社区</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3d3dy5yLWJsb2dnZXJzLmNvbS8=" title="http://www.r-bloggers.com/">R-bloggers</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cHM6Ly9jcmFuLnItcHJvamVjdC5vcmcvb3RoZXItZG9jcy5odG1s" title="https://cran.r-project.org/other-docs.html">R-resource</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL3R1dG9yaWFscy5pcS5oYXJ2YXJkLmVkdS9SL1JncmFwaGljcy9SZ3JhcGhpY3MuaHRtbA==" title="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html">ggplot2</span>
                  </li>
                
                  <li class="links-of-blogroll-item">
                    <span class="exturl" data-url="aHR0cDovL2xpbGliZWkubmV0Lw==" title="http://lilibei.net/">pele' blog</span>
                  </li>
                
              </ul>
            </div>
          

          
            
          
          

        </div>
      </div>

      

      

    </div>
  </aside>
  


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; 2014 – <span itemprop="copyrightYear">2023</span>
  <span class="with-love" id="animate">
    <i class="fa fa-heart"></i>
  </span>
  <a class="theme-link" target="_blank" href="http://tiramisutes.github.io/about/"><span class="author" itemprop="copyrightHolder">tiramisutes</span></a>

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Symbols count total"></span>
  

  
    <span class="post-meta-item-icon">
    </span>
    <span title="Reading time total"></span>
  
</div>


  <div class="powered-by">Powered: <span class="exturl theme-link" data-url="aHR0cHM6Ly9oZXhvLmlv">Hexo</span></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">▽ – <span class="exturl theme-link" data-url="aHR0cHM6Ly9naXRodWIuY29tL25leHQtdGhlbWUvaGV4by10aGVtZS1uZXh0">NexT.Muse</span></div>




        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  

  

  
    <span class="post-meta-item-icon">
      <i class="fa fa-line-chart"></i>
    </span>
    <span class="site-pv" title="Total Views">
    <a style="line-height: 1.5;" href="https://tongji.baidu.com/web/welcome/login" target="_blank">The total visits
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>&nbsp;times</a>
    </span>
  
</div>









        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>












  



  
    
    
      
    
  
  <script color='0,0,255' opacity='0.5' zIndex='-1' count='20' src="//cdn.jsdelivr.net/gh/theme-next/theme-next-canvas-nest@1/canvas-nest-nomobile.min.js"></script>













  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="//cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.6/jquery.fancybox.min.js"></script>


  


  <script src="/js/src/utils.js?v=7.0.1"></script>

  <script src="/js/src/motion.js?v=7.0.1"></script>



  
  


  <script src="/js/src/schemes/muse.js?v=7.0.1"></script>



  

  


  <script src="/js/src/next-boot.js?v=7.0.1"></script>


  

  
  <script src="/js/src/exturl.js?v=7.0.1"></script>


  
  
  
    
  

  
    
      
    
  

  
  
  
    <script src="//cdn.jsdelivr.net/npm/quicklink@1/dist/quicklink.umd.js"></script>
    <script>
      
        window.addEventListener('load', () => {
      
        quicklink({
          timeout: 3000,
          priority: true,
          ignores: [uri => uri.includes('#'),uri => uri == 'http://tiramisutes.github.io/page/10/',]
        });
      
        });
      
    </script>
  


  
  

  

<script src="//cdn.jsdelivr.net/npm/leancloud-storage@3/dist/av-min.js"></script>



<script src="//unpkg.com/valine/dist/Valine.min.js"></script>

<script>
  var GUEST = ['nick', 'mail', 'link'];
  var guest = 'nick,mail,link';
  guest = guest.split(',').filter(function(item) {
    return GUEST.indexOf(item) > -1;
  });
  new Valine({
    el: '#comments',
    verify: false,
    notify: true,
    appId: 'c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz',
    appKey: 'iVJ6ST0e1O5xaiIhVwOyhbf3',
    placeholder: 'Just go go',
    avatar: 'mm',
    meta: guest,
    pageSize: '10' || 10,
    visitor: true
  });
</script>





  
  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  

  
  
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  

  


  
<style>
.pdfobject-container {
  position: relative;
  overflow: auto;
  width: 100%;
}
</style>
<script>
if ($('body').find('div.pdf').length) {
  $.ajax({
    type: 'GET',
    url: '//cdn.jsdelivr.net/npm/pdfobject@2.1.1/pdfobject.min.js',
    dataType: 'script',
    cache: true,
    success: function() {
      $('body').find('div.pdf').each(function(i, o) {
        PDFObject.embed($(o).attr('target'), $(o), {
          pdfOpenParams: {
            navpanes: 0,
            toolbar: 0,
            statusbar: 0,
            pagemode: 'thumbs',
            view: 'FitH'
          },
          PDFJS_URL: '/lib/pdf/web/viewer.html',
          height: $(o).attr('height') || '500px'
        });
      });
    },
  });
}
</script>


  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      bp.src = (curProtocol === 'https') ? 'https://zz.bdstatic.com/linksubmit/push.js' : 'http://push.zhanzhang.baidu.com/push.js';
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  
  
  
    
  
  <script src="//cdn.jsdelivr.net/gh/theme-next/theme-next-needmoreshare2@1/needsharebutton.min.js"></script>
  <script>
    
      pbOptions = {};
      
        pbOptions.iconStyle = "box";
      
        pbOptions.boxForm = "horizontal";
      
        pbOptions.position = "bottomCenter";
      
        pbOptions.networks = "Weibo,Wechat,Douban,QQZone,Twitter,Evernote,Facebook";
      
      new needShareButton('#needsharebutton-postbottom', pbOptions);
    
    
  </script>


  

  

  

  

</body>
</html>
